// @vds-core/tabs/dist/esm/configs/QuadrupleInverted.js
import { colors } from '@vds-marketing/theme';
var titleText = 'Lorem ipsum dolor sit ame';
var bodyText = 'Lorem ipsum dolor sit amet consectetur adipisicing eli';

var onClickItem = function onClickItem() {
  return alert('You clicked an item to go somewhere!');
};

var QuadrupleInvertedData = {
  desktop: {
    paddingRight: '20X',
    inverted: true,
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  desktopSmall: {
    paddingRight: '20X',
    inverted: true,
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  tablet: {
    paddingRight: '20X',
    inverted: true,
    stacked: true,
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true,
        width: '100%'
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '12X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '12X'
      },
      lastChild: true,
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        paddingRight: '20X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: true
      },
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          color: colors.white,
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  mobile: {
    inverted: true,
    data: [{
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '32X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }]
  },
  mobileSmall: {
    inverted: true,
    data: [{
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        backgroundColor: false,
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          color: colors.white,
          viewport: 'desktop',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '32X',
          color: colors.white,
          width: '72%'
        }]
      },
      onClick: onClickItem
    }]
  }
};
export { QuadrupleInvertedData };
