// @vds-core/tabs/dist/esm/configs/QuadrupleDefault.js
import { colors } from '@vds-marketing/theme';
var titleText = 'Lorem ipsum dolor sit ame';
var bodyText = 'Lorem ipsum dolor sit amet consectetur adipisicing eli';

var onClickItem = function onClickItem() {
  return alert('You clicked an item to go somewhere!');
};

var QuadrupleDefaultData = {
  desktop: {
    paddingRight: '20X',
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  desktopSmall: {
    paddingRight: '20X',
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'large',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  tablet: {
    paddingRight: '20X',
    stacked: true,
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X',
        width: '100%'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '12X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '12X'
      },
      lastChild: true,
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X'
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  mobile: {
    data: [{
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '32X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }]
  },
  mobileSmall: {
    data: [{
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'medium',
          text: titleText,
          viewport: 'desktop',
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '32X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }]
  }
};
export { QuadrupleDefaultData };
