// @vds/typography/dist/cjs/fonts/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Fonts = void 0;

var _webfontloader = _interopRequireDefault(require("webfontloader"));

var _typography = require("@vds-core/typography");

//****** To-do, test loading in the families first asynchronously from check whether to load the url if necessary ******//
var FontConfig = {
  loaded: false
}; //Turn on or off the font loader

function setFontLoaded(loaded) {
  FontConfig.loaded = loaded ? true : false;
  _typography.Fonts.setFontLoaded && _typography.Fonts.setFontLoaded(loaded);
}

function checkFontLoaded() {
  return FontConfig.loaded;
} // Define and Load the fonts


var customConfig = {
  families: ['Verizon-NHG-eTX', 'Verizon-NHG-eDS']
};
var fontUri1 = 'https://respframework.verizon.com/fonts/eText/verizon-nhg-etx-update.css';
var fontUri2 = 'respframework.verizon.com/fonts/eText/verizon-nhg-etx.css';

function finishLoading(customConfig) {
  var WebFontConfig = {
    custom: customConfig,
    timeout: 10,
    inactive: function inactive() {}
  };

  _webfontloader["default"].load(WebFontConfig);
}

function checkIfLoaded() {
  var sheetRefs = document.styleSheets;
  var sheetsArray = sheetRefs && Object.values(sheetRefs);
  return sheetsArray.find(function (styleSheet) {
    return styleSheet.href === fontUri1 || styleSheet.href === fontUri2;
  });
}

function loadFonts(loaded) {
  var headTag = document.documentElement && document.documentElement.outerHTML;
  var fontActive = (headTag.indexOf('verizonnhgetx') > -1 || headTag.indexOf('verizonnhgeds') > -1) && headTag.indexOf('wf-active') > -1;

  if (!fontActive) {
    //Double check in case the head tag is being wiped programmatically
    var _loaded = checkIfLoaded();

    if (!_loaded) customConfig.urls = [fontUri1];
    finishLoading(customConfig);
  }
}

if (!FontConfig.loaded) {
  _typography.Fonts.setFontLoaded && _typography.Fonts.setFontLoaded(true);
  loadFonts();
}

var Fonts = {
  VerizonNHGeDS: 'Verizon-NHG-eDS',
  // Display
  VerizonNHGeTX: 'Verizon-NHG-eTX',
  // Text
  checkFontLoaded: checkFontLoaded,
  setFontLoaded: setFontLoaded,
  loadFonts: loadFonts
};
exports.Fonts = Fonts;
