// @vds-core/carousels/dist/cjs/components/CarouselWrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _icons = _interopRequireDefault(require("@vds-core/icons"));

var _theme = require("@vds-core/theme");

var _utilities = require("@vds-core/utilities");

var _ProgressIndicator = _interopRequireDefault(require("./ProgressIndicator"));

var _CarouselBars = _interopRequireDefault(require("./CarouselBars"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject7() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: absolute;\n  bottom: ", ";\n  right: ", ";\n  left: ", ";\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n  justify-content: space-between;\n  position: absolute;\n  top: 20px;\n  right: 20px;\n  z-index: 1;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  margin-right: ", "px;\n  background-color: transparent;\n  padding: 0px;\n  border: none;\n  cursor: pointer;\n  outline: none;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  flex-grow: 1;\n  display: flex;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  transition: opacity 0.3s ease;\n  opacity: ", ";\n  display: flex;\n  order: ", ";\n  width: ", ";\n  flex: 1 0 100%;\n  flex-basis: 100%;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  width: 100%;\n  position: absolute;\n  bottom: 0;\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n  justify-content: center;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: relative;\n  overflow: hidden;\n  width: 100%;\n  user-select: none;\n  display: flex;\n  flex-direction: column;\n  justify-content: flex-end;\n  box-sizing: border-box;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  /**
   * Cards rendered in the Carousel.
   */
  children: _propTypes["default"].node.isRequired,

  /**
   * Viewport the Carousel will be rendered in.
   */
  viewport: _propTypes["default"].oneOf(['desktop', 'tablet', 'mobile']),

  /**
   * Specifies positioning of the ProgressIndicator on the Card.
   */
  autoplayIndicatorPosition: _propTypes["default"].oneOf(['left', 'right', 'hidden']),

  /**
   * Invert the color scheme of the Carousel.
   */
  inverted: _propTypes["default"].bool,

  /**
   *  Unique identifier value for 'aria-controls'.
   */
  uniqueId: _propTypes["default"].string.isRequired,

  /**
   * Allows Carousel to automatically move to next card after given interval.
   */
  autoplay: _propTypes["default"].bool,

  /**
   * Specifies how much of the next card up is shown in pixels or percentage.
   */
  peekOffset: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].number]),

  /**
   * Specifies the duration a Card will be shown before automatically switching to the next.
   */
  autoplayInterval: _propTypes["default"].number,

  /**
   * Opacity of slides that are inactive.
   */
  inactiveOpacity: _propTypes["default"].number,

  /**
   * Styling of the footer container.
   */
  paginationContainerStyle: _propTypes["default"].object,

  /**
   * function that returns the component that will be used in place of default arrow right.
   */
  renderNextControl: _propTypes["default"].func,

  /**
   * function that returns the component that will be used in place of default arrow left.
   */
  renderPreviousControl: _propTypes["default"].func,

  /**
   * function that returns the component that will be used in place of default carousel bars.
   */
  renderPagination: _propTypes["default"].func,

  /**
   * @ignore
   */
  dataLoc: _propTypes["default"].string
};
var defaultProps = {
  inverted: false,
  uniqueId: 'slide_carousel_bars_id',
  autoplay: false,
  viewport: 'desktop',
  autoplayIndicatorPosition: 'right',
  peekOffset: 0,
  inactiveOpacity: 0.5,
  autoplayInterval: 8000,
  paginationContainerStyle: {},
  renderNextControl: undefined,
  renderPreviousControl: undefined,
  renderPagination: undefined
};
var ARROW_LARGE = '46px';
var ARROW_SMALL = '23px';

var CarouselContainer = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "CarouselContainer",
  componentId: "ppes81-0"
})(_templateObject());

var CarouselFooter = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "CarouselFooter",
  componentId: "ppes81-1"
})(_templateObject2());

var CarouselSlot = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "CarouselSlot",
  componentId: "ppes81-2"
})(_templateObject3(), function (_ref) {
  var active = _ref.active,
      inactiveOpacity = _ref.inactiveOpacity,
      count = _ref.count;
  return active || count === 1 ? 1 : inactiveOpacity;
}, function (_ref2) {
  var order = _ref2.order;
  return order;
}, function (_ref3) {
  var peekOffset = _ref3.peekOffset;
  return "calc(100% - ".concat(peekOffset, ")");
});

var StaticSlot = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "StaticSlot",
  componentId: "ppes81-3"
})(_templateObject4());

var ButtonWrapper = /*#__PURE__*/_styledComponents["default"].button.withConfig({
  displayName: "ButtonWrapper",
  componentId: "ppes81-4"
})(_templateObject5(), function (_ref4) {
  var lastChild = _ref4.lastChild;
  return lastChild ? 0 : 10;
});

var ButtonGroup = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "ButtonGroup",
  componentId: "ppes81-5"
})(_templateObject6());

var IndicatorWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "IndicatorWrapper",
  componentId: "ppes81-6"
})(_templateObject7(), (0, _javascriptStyleHelpers.calculateRem)(8), function (_ref5) {
  var autoplayIndicatorPosition = _ref5.autoplayIndicatorPosition;
  return autoplayIndicatorPosition === 'right' ? (0, _javascriptStyleHelpers.calculateRem)(20) : undefined;
}, function (_ref6) {
  var autoplayIndicatorPosition = _ref6.autoplayIndicatorPosition;
  return autoplayIndicatorPosition === 'left' ? (0, _javascriptStyleHelpers.calculateRem)(20) : undefined;
}); // gets the correct clientX whether from touch or mouse event


var getClientX = function getClientX(e) {
  return e.targetTouches ? e.targetTouches[0].clientX : e.clientX;
};

var CarouselWrapper = /*#__PURE__*/function (_React$PureComponent) {
  (0, _inherits2["default"])(CarouselWrapper, _React$PureComponent);

  var _super = _createSuper(CarouselWrapper);

  function CarouselWrapper(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, CarouselWrapper);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "resize", function () {
      _this.slideWidth = _this.carouselRef.offsetWidth;
      var newTranslation = -_this.slideWidth;

      _this.setState({
        translation: newTranslation
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_startAutoplay", function () {
      var _this$props = _this.props,
          autoplayInterval = _this$props.autoplayInterval,
          autoplay = _this$props.autoplay;

      if (autoplayInterval && autoplay) {
        _this.autoplayTimer = setInterval(function () {
          _this._arrowToSlide(1, true)();
        }, Math.abs(autoplayInterval));
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_pauseAutoplay", function () {
      var _this$props2 = _this.props,
          autoplayInterval = _this$props2.autoplayInterval,
          autoplay = _this$props2.autoplay;
      if (!autoplayInterval || !autoplay) return;
      clearInterval(_this.autoplayTimer);
      clearTimeout(_this.pauseTimer);
      _this.pauseTimer = setTimeout(_this._startAutoplay, 0);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getSlideInvertedProp", function (previousIndex, index) {
      var _this$props3 = _this.props,
          children = _this$props3.children,
          inverted = _this$props3.inverted;
      var selectedChild = index === children.length - 1 ? _react["default"].Children.toArray(children)[0] : _react["default"].Children.toArray(children)[index + 1];
      var buttons = Array.prototype.slice.call(document.getElementsByTagName('button'));

      if (buttons.length > 2) {
        buttons.splice(0, 2);
        var indexButton = index === children.length - 1 ? buttons[0] : buttons[index + 1];
        var previousButton = previousIndex === children.length - 1 ? buttons[0] : buttons[previousIndex + 1];
        buttons.forEach(function (button) {
          if (button === indexButton) {
            button.setAttribute('tabindex', '0');
          }

          if (button === previousButton) {
            button.setAttribute('tabindex', '-1');
          }
        });
      }

      if ((selectedChild.props || {}).inverted !== undefined) {
        return selectedChild.props.inverted;
      }

      return inverted;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getOrder", function (index, position, children) {
      var numItems = children.length || 1;

      if (index - position < 0) {
        return numItems - Math.abs(index - position);
      }

      return index - position;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getActive", function (index, position, children) {
      var numItems = children.length - 1 || 1;

      if (index === 0) {
        return position === numItems;
      } else {
        if (index > _this.slideCount - 1) {
          return false;
        }

        return index === position + 1;
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getNextForwardIndex", function (index) {
      //index 2, direction forward
      if (index < _this.slideCount - 1) {
        return index += 1;
      } else if (index === _this.slideCount - 1) {
        return 0;
      }

      return index;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getNextBackwardIndex", function (index) {
      //index 0, direction forward
      if (index > 0) {
        return index -= 1;
      } else if (index === 0) {
        return _this.slideCount - 1;
      }

      return index;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_arrowToSlide", function (direction, fromAutoplay) {
      var _this$state = _this.state,
          position = _this$state.position,
          barPosition = _this$state.barPosition;
      var children = _this.props.children;
      if (!fromAutoplay) _this._pauseAutoplay();

      if (position !== barPosition) {
        var nextIndex = _this._getNextForwardIndex(barPosition);

        var previousIndex = _this._getNextBackwardIndex(barPosition);

        _this.setState({
          sliding: true,
          position: direction === 'forward' ? nextIndex : previousIndex,
          barPosition: direction === 'forward' ? nextIndex : previousIndex,
          translation: direction === 'forward' ? '0' : 2 * -_this.slideWidth,
          inverted: _this._getSlideInvertedProp(position, direction === 'forward' ? nextIndex : previousIndex)
        });
      } else {
        var _nextIndex = _this._getNextForwardIndex(position);

        var _previousIndex = _this._getNextBackwardIndex(position);

        _this.setState({
          sliding: true,
          position: direction === 'forward' ? _nextIndex : _previousIndex,
          barPosition: direction === 'forward' ? _nextIndex : _previousIndex,
          translation: direction === 'forward' ? '0' : 2 * -_this.slideWidth,
          inverted: _this._getSlideInvertedProp(position, direction === 'forward' ? _nextIndex : _previousIndex)
        });
      }

      setTimeout(function () {
        _this.setState({
          sliding: false,
          translation: -_this.slideWidth
        });
      }, 20);
      _this._resetIndicator && _this._resetIndicator();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_swipeToSlide", function (original, translation) {
      var position = _this.state.position;

      var nextIndex = _this._getNextForwardIndex(position);

      var previousIndex = _this._getNextBackwardIndex(position);

      if (!(Math.abs(original) <= 0.1 * _this.slideWidth)) {
        if (original >= 0) {
          _this.setState({
            sliding: true,
            position: nextIndex,
            barPosition: nextIndex,
            translation: -original,
            inverted: _this._getSlideInvertedProp(position, nextIndex)
          });
        } else {
          _this.setState({
            sliding: true,
            position: previousIndex,
            barPosition: previousIndex,
            translation: -_this.slideWidth - translation,
            inverted: _this._getSlideInvertedProp(position, previousIndex)
          });
        }
      }

      setTimeout(function () {
        _this.setState({
          sliding: false,
          translation: -_this.slideWidth
        });
      }, 20);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_moveToSlide", function (carouselBarIndex) {
      var _this$state2 = _this.state,
          position = _this$state2.position,
          barPosition = _this$state2.barPosition;
      var newIndex = carouselBarIndex - 1; //moving to first position

      if (position !== 1 && newIndex === 1) {
        if (position === 0) {
          _this._arrowToSlide('forward', false)();

          return;
        } else {
          _this.setState({
            position: 1,
            barPosition: newIndex,
            sliding: true,
            translation: -_this.slideWidth * position,
            inverted: _this._getSlideInvertedProp(barPosition, newIndex)
          });

          setTimeout(function () {
            _this.setState({
              sliding: false,
              translation: -_this.slideWidth
            });
          }, 20);
        } //moving to other positions

      } else {
        _this.setState({
          position: 1,
          barPosition: newIndex,
          sliding: false,
          translation: -_this.slideWidth * newIndex,
          inverted: _this._getSlideInvertedProp(barPosition, newIndex)
        });
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_snapToCard", function (swipeDistance) {
      return swipeDistance <= 0 ? -Math.ceil(-_this.slideWidth - swipeDistance) : -Math.floor(_this.slideWidth - swipeDistance);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_handleCarouselSwipeStart", function (e) {
      var _this$state3 = _this.state,
          maxedViewport = _this$state3.maxedViewport,
          barPosition = _this$state3.barPosition;
      var children = _this.props.children;
      if (maxedViewport) return;
      var clientX = getClientX(e);
      _this.prevScrollPos = clientX;
      _this.swipeStartPosition = clientX;

      _this._pauseAutoplay();

      _this.setState({
        swiping: true,
        sliding: true,
        position: barPosition,
        translation: -_this.slideWidth
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_handleCarouselSwipeMove", function (e) {
      var _this$state4 = _this.state,
          maxedViewport = _this$state4.maxedViewport,
          swiping = _this$state4.swiping,
          translation = _this$state4.translation;
      if (!swiping || maxedViewport) return;
      var clientX = getClientX(e);
      var deltaFromPrevScroll = clientX - _this.prevScrollPos;
      var newTranslation = translation + deltaFromPrevScroll;
      _this.prevScrollPos = clientX;

      _this._pauseAutoplay();

      _this.setState({
        translation: newTranslation
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_handleCarouselSwipeEnd", function () {
      var _this$state5 = _this.state,
          maxedViewport = _this$state5.maxedViewport,
          swiping = _this$state5.swiping,
          translation = _this$state5.translation;
      if (!swiping || maxedViewport) return;
      var swipeDistance = _this.swipeStartPosition - _this.prevScrollPos;

      var snap = _this._snapToCard(swipeDistance);

      _this.setState({
        swiping: false
      });

      _this._swipeToSlide(swipeDistance, snap);

      _this._resetIndicator && _this._resetIndicator();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_renderChildren", function (children) {
      var _this$props4 = _this.props,
          inverted = _this$props4.inverted,
          inactiveOpacity = _this$props4.inactiveOpacity,
          peekOffset = _this$props4.peekOffset;
      var _this$state6 = _this.state,
          position = _this$state6.position,
          barPosition = _this$state6.barPosition;

      var childArr = _react["default"].Children.toArray(children);

      var newChildren = (0, _toConsumableArray2["default"])(childArr);

      if (children.length === 2) {
        newChildren.push(newChildren[position]);
        newChildren.push(newChildren[position === 0 ? 1 : 0]);
      }

      if (barPosition === _this.slideCount - 1) {
        newChildren.push(newChildren[1]);
      }

      return newChildren.map(function (child, index) {
        return /*#__PURE__*/_react["default"].createElement(CarouselSlot, {
          index: index,
          peekOffset: peekOffset,
          order: _this._getOrder(index, position, children),
          key: index,
          active: _this._getActive(index, barPosition, children),
          inverted: inverted,
          count: _this.slideCount,
          inactiveOpacity: inactiveOpacity
        }, child);
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_renderChildrenInOrder", function (children) {
      var _this$props5 = _this.props,
          showDividers = _this$props5.showDividers,
          inverted = _this$props5.inverted;
      var position = _this.state.position;
      return _this.childArray.map(function (child, index) {
        return /*#__PURE__*/_react["default"].createElement(StaticSlot, {
          key: index,
          border: showDividers,
          inverted: inverted
        }, _this.childArray[(position + index) % _this.slideCount]);
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_renderCarouselChildren", function (children) {
      if (children && _this.slideCount === 1) {
        return children;
      } else if (children) {
        return _this._renderChildrenInOrder(children);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_setIndicatorRef", function (ref) {
      var noop = function noop() {};

      if (!ref || ref === null || !ref._resetIndicator) return _this._resetIndicator = noop;
      _this._resetIndicator = ref._resetIndicator;
    });
    var _children = props.children;
    _this.slideWidth = 0;
    _this.swipeStartPosition = 0;
    _this.carouselRef = null;
    _this.prevScrollPos = 0;
    _this.fullRotation = 0;
    _this.autoplayTimer = null;
    _this.pauseTimer = null;
    _this.slideCount = _react["default"].Children.count(_children);
    _this.childArray = _react["default"].Children.toArray(_children);
    _this.state = {
      position: 0,
      barPosition: 0,
      translation: 0,
      sliding: true,
      animate: false,
      swiping: false,
      swipeDirection: null,
      maxedViewport: false,
      delayedTranslation: false,
      inverted: _this._getSlideInvertedProp(0, 0)
    };
    return _this;
  }

  (0, _createClass2["default"])(CarouselWrapper, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this._startAutoplay();

      this.slideWidth = this.carouselRef.offsetWidth;
      this.fullRotation = -this.slideWidth;
      window.addEventListener('resize', this.resize);
      this.resize();
      var buttons = Array.prototype.slice.call(document.getElementsByTagName('button'));

      if (buttons.length > 2) {
        buttons.splice(0, 2);
        buttons.forEach(function (button) {
          button.setAttribute('tabindex', '-1');
        });
        buttons[1].setAttribute('tabindex', '0');
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      window.removeEventListener('resize', this.resize);
      clearInterval(this.autoplayTimer);
      clearTimeout(this.pauseTimer);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props6 = this.props,
          children = _this$props6.children,
          viewport = _this$props6.viewport,
          uniqueId = _this$props6.uniqueId,
          minSlideWidth = _this$props6.minSlideWidth,
          peekOffset = _this$props6.peekOffset,
          autoplay = _this$props6.autoplay,
          autoplayIndicatorPosition = _this$props6.autoplayIndicatorPosition,
          autoplayInterval = _this$props6.autoplayInterval,
          paginationContainerStyle = _this$props6.paginationContainerStyle,
          renderNextControl = _this$props6.renderNextControl,
          renderPreviousControl = _this$props6.renderPreviousControl,
          renderPagination = _this$props6.renderPagination,
          dataLoc = _this$props6.dataLoc;
      var _this$state7 = this.state,
          inverted = _this$state7.inverted,
          position = _this$state7.position,
          barPosition = _this$state7.barPosition,
          translation = _this$state7.translation,
          sliding = _this$state7.sliding,
          animate = _this$state7.animate;
      var arrowSize = viewport === 'mobile' ? ARROW_SMALL : ARROW_LARGE,
          arrowStyle = {
        height: arrowSize,
        width: arrowSize,
        minHeight: arrowSize,
        minWidth: arrowSize
      },
          arrowColor = inverted ? _theme.colors.white : _theme.colors.black;
      var viewTranslation = this.slideCount !== 1 ? translation : 0;

      if (!sliding) {
        setTimeout(function () {
          _this2.setState({
            sliding: true
          });
        }, 20);
      }

      return /*#__PURE__*/_react["default"].createElement(CarouselContainer, {
        inverted: inverted,
        slideCount: this.slideCount,
        id: uniqueId,
        peekOffset: parseInt(peekOffset)
      }, /*#__PURE__*/_react["default"].createElement(ButtonGroup, null, /*#__PURE__*/_react["default"].createElement(ButtonWrapper, {
        "data-loc": dataLoc,
        tabIndex: 0,
        onClick: function onClick() {
          _this2._arrowToSlide('backward', false);
        }
      }, renderPreviousControl ? renderPreviousControl({
        inverted: inverted
      }) : /*#__PURE__*/_react["default"].createElement(_icons["default"], {
        name: "arrow-left",
        color: arrowColor,
        style: arrowStyle,
        tabIndex: -1
      })), /*#__PURE__*/_react["default"].createElement(ButtonWrapper, {
        "data-loc": dataLoc,
        lastChild: true,
        onClick: function onClick() {
          _this2._arrowToSlide('forward', false);
        },
        tabIndex: 0
      }, renderNextControl ? renderNextControl({
        inverted: inverted
      }) : /*#__PURE__*/_react["default"].createElement(_icons["default"], {
        name: "arrow-right",
        color: arrowColor,
        style: arrowStyle,
        tabIndex: -1
      }))), /*#__PURE__*/_react["default"].createElement("div", {
        // style here because translation changes frequently, styled-components are too slow
        style: {
          display: 'flex',
          height: '100%',
          width: "calc(100% - ".concat(peekOffset, ")"),
          transform: "translateX(".concat(viewTranslation, "px)"),
          transition: "transform ".concat(!sliding ? 0.3 : 0, "s ease")
        },
        onMouseDown: this._handleCarouselSwipeStart,
        onMouseMove: this._handleCarouselSwipeMove,
        onMouseLeave: this._handleCarouselSwipeEnd,
        onMouseUp: this._handleCarouselSwipeEnd,
        onTouchStart: this._handleCarouselSwipeStart,
        onTouchMove: this._handleCarouselSwipeMove,
        onTouchEnd: this._handleCarouselSwipeEnd,
        ref: function ref(el) {
          _this2.carouselRef = el;
        }
      }, !!children && this._renderChildren(children)), /*#__PURE__*/_react["default"].createElement(CarouselFooter, {
        style: paginationContainerStyle
      }, renderPagination ? renderPagination({
        goToSlide: this._moveToSlide,
        slideCount: this.slideCount,
        activeSlide: barPosition + 1,
        inverted: inverted
      }) : /*#__PURE__*/_react["default"].createElement(_CarouselBars["default"], {
        uniqueId: uniqueId,
        goToSlide: this._moveToSlide,
        slideCount: this.slideCount,
        activeSlide: barPosition + 1,
        inverted: inverted,
        ariaLabel: "Slide Carousel"
      })), autoplayIndicatorPosition !== 'hidden' && autoplayInterval && autoplay && /*#__PURE__*/_react["default"].createElement(IndicatorWrapper, {
        autoplayIndicatorPosition: autoplayIndicatorPosition
      }, /*#__PURE__*/_react["default"].createElement(_ProgressIndicator["default"], {
        ref: this._setIndicatorRef,
        interval: autoplayInterval,
        inverted: inverted
      })));
    }
  }]);
  return CarouselWrapper;
}(_react["default"].PureComponent);

CarouselWrapper.propTypes = propTypes;
CarouselWrapper.defaultProps = defaultProps;

var _default = (0, _utilities.withVDSManager)(CarouselWrapper);

exports["default"] = _default;
