// @vds-core/carousels/dist/cjs/components/Carousel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _CarouselWrapper = _interopRequireDefault(require("./CarouselWrapper"));

var propTypes = {
  /**
   * The items in the carousel.
   */
  data: _propTypes["default"].array.isRequired,

  /**
   * function that returns the item to be rendered in each slide.
   */
  renderItem: _propTypes["default"].func.isRequired,

  /**
   * @ignore
   */
  dataLoc: _propTypes["default"].string
};

var Carousel = function Carousel(props) {
  var data = props.data,
      renderItem = props.renderItem,
      rest = (0, _objectWithoutProperties2["default"])(props, ["data", "renderItem"]);
  var newData = (0, _toConsumableArray2["default"])(data);
  var popElement = newData.pop();
  newData.unshift(popElement);

  var _renderChildren = function _renderChildren() {
    return newData.map(function (item, index) {
      return renderItem({
        item: item,
        index: index
      });
    });
  };

  return /*#__PURE__*/_react["default"].createElement(_CarouselWrapper["default"], rest, _renderChildren());
};

Carousel.propTypes = propTypes;
var _default = Carousel;
exports["default"] = _default;
