// @vds-core/typography/dist/esm/components/Body/Body.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  &:active,\n  &:focus,\n  &:hover {\n    outline: none;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { colors } from '@vds-core/theme';
import TypographyConfig from '../../typographyConfig';
import fontColorPropType from '../../util/fontColorPropType';
import { Typography } from '../Typography';
import { withVDSManager } from '@vds-core/utilities';
import defaultConfig from './config';

function getBodyFontWeight(typescale, bold) {
  switch (typescale) {
    case 'MVP':
    case 'VDS':
      return bold ? '750' : '400';
    // 400 is mapped to match 55 Roman

    default:
      return bold ? 'bold' : 'normal';
  }
}

var propTypes = {
  /**
   * Primitive used for the component.
   */
  primitive: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span', 'p']),

  /**
   * Viewport the font styling is based on. Desktop and tablet are the same size.
   */
  viewport: PropTypes.oneOf(['desktop', 'tablet', 'mobile']),

  /**
   * Size of the typescale that will be used.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),

  /**
   * Content that will be passed to the component.
   */
  children: PropTypes.node,

  /**
   * The color of the text.
   */
  color: fontColorPropType,

  /**
   * Adds a fontWeight of bold.
   */
  bold: PropTypes.bool,

  /**
   * Allows for specific typescale to be passed to override global config.
   */
  typescale: PropTypes.oneOf(['VDS', 'MVP']),

  /**
   * If the theme object contains the `typescale` key and no typescale prop was passed global typescale is overriden by theme.
   */
  theme: PropTypes.object,

  /**
   * The tabIndex of the text.
   */
  tabIndex: PropTypes.number,

  /**
   * @ignore
   */
  config: PropTypes.object,

  /**
   * @ignore
   */
  getfontWeight: PropTypes.func,

  /**
   * Assigns an Id to the component
   */
  id: PropTypes.string,

  /**
   * Adds an aria-labelledby attribute to the component
   */
  ariaLabelledBy: PropTypes.string
};
var defaultProps = {
  primitive: 'p',
  viewport: 'desktop',
  size: 'small',
  children: null,
  color: colors.black,
  bold: false,
  config: defaultConfig,
  getFontWeight: getBodyFontWeight
};
var StyledBody = /*#__PURE__*/styled(Typography).withConfig({
  displayName: "StyledBody",
  componentId: "sc-1s1yqd8-0"
})(_templateObject());

var Body = function Body(props) {
  var config = props.config,
      children = props.children,
      primitive = props.primitive,
      viewport = props.viewport,
      size = props.size,
      color = props.color,
      bold = props.bold,
      typescale = props.typescale,
      theme = props.theme,
      tabIndex = props.tabIndex,
      getFontWeight = props.getFontWeight,
      fontsConfigProp = props.fontsConfig,
      ariaLabel = props.ariaLabel,
      id = props.id,
      ariaLabelledBy = props.ariaLabelledBy;
  var // What typescale is being used for configuration
  typescaleToUse = typescale ? typescale : theme && theme.typescale ? theme.typescale : TypographyConfig.getTypescale(),
      viewportToUse = viewport === 'mobile' ? 'mobile' : 'desktop',
      // What is the specific font configuration we are using
  fontConfig = config[typescaleToUse][viewportToUse][size],
      fontWeight = getFontWeight(typescaleToUse, bold);
  return /*#__PURE__*/React.createElement(StyledBody, {
    fontSize: fontConfig.fontSize,
    fontWeight: fontWeight,
    fontFamily: fontConfig.fontFamily,
    lineheight: fontConfig.lineHeight,
    letterSpacing: fontConfig.letterSpacing && fontConfig.letterSpacing,
    primitive: primitive,
    color: color,
    tabIndex: tabIndex,
    "aria-hidden": tabIndex < 0,
    fontsconfig: fontsConfigProp,
    "aria-label": ariaLabel,
    id: id,
    "aria-labelledby": ariaLabelledBy
  }, children);
};

Body.propTypes = propTypes;
Body.defaultProps = defaultProps;
export default withVDSManager(Body);
