// @vds-core/tabs/dist/cjs/components/MarketingTabItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utilities = require("@vds/utilities");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  margin-top: ", ";\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  /**
   * When passed in, will return rendered component.
   */
  children: _propTypes["default"].node,

  /**
   * Viewport the TabItem will be rendered in.
    <br> PropTypes.oneOf([ "desktop", "desktopSmall", "tablet", "mobile", "mobileSmall" ])
   */
  viewport: _propTypes["default"].oneOf(['desktop', 'desktopSmall', 'tablet', 'mobile', 'mobileSmall'])
};
var defaultProps = {
  renderMargin: false
};

var ContentWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "ContentWrapper",
  componentId: "sc-1m9ibue-0"
})(_templateObject(), function (_ref) {
  var renderMargin = _ref.renderMargin;
  return renderMargin ? "32px" : 0;
});

var MarketingTabItem = function MarketingTabItem(props) {
  var children = props.children,
      viewport = props.viewport,
      renderMargin = props.renderMargin;

  function _cloneAndAddProps(children) {
    if (!children) return null;

    if (children && children.type && children.type.displayName && children.type.displayName.includes("TileNavigation")) {
      return /*#__PURE__*/_react["default"].createElement(ContentWrapper, {
        renderMargin: renderMargin
      }, /*#__PURE__*/_react["default"].cloneElement(children, _objectSpread(_objectSpread(_objectSpread({}, children.props[viewport]), children.props), {}, {
        tabsChild: true
      })));
    }

    return /*#__PURE__*/_react["default"].cloneElement(children, _objectSpread(_objectSpread(_objectSpread({}, children.props[viewport]), children.props), {}, {
      tabsChild: true
    }));
  }

  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, _cloneAndAddProps(children));
};

MarketingTabItem.propTypes = propTypes;
MarketingTabItem.defaultProps = defaultProps;

var _default = (0, _utilities.withVDSManager)(MarketingTabItem);

exports["default"] = _default;
