// @vds-core/tabs/dist/cjs/components/TabPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: none;\n  outline: none;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  /**
   * @ignore
   */
  children: _propTypes["default"].node,

  /**
   * @ignore
   */
  label: _propTypes["default"].string.isRequired,

  /**
   * @ignore
   */
  isHidden: _propTypes["default"].bool,

  /**
   * @ignore
   */
  uniqueId: _propTypes["default"].string,

  /**
   * @ignore
   */
  className: _propTypes["default"].string
};
var defaultProps = {
  isHidden: false,
  uniqueId: undefined,
  className: null
};

var StyledPanel = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "StyledPanel",
  componentId: "yzzvp2-0"
})(_templateObject());
/**
 * @ignore
 */


var TabPanel = function TabPanel(_ref) {
  var isHidden = _ref.isHidden,
      label = _ref.label,
      uniqueId = _ref.uniqueId,
      children = _ref.children,
      className = _ref.className;
  var styles = isHidden ? {
    display: 'none'
  } : {
    display: 'block'
  };
  return /*#__PURE__*/_react["default"].createElement(StyledPanel, {
    id: "tabPanel_".concat(uniqueId),
    "aria-hidden": isHidden,
    "aria-labelledby": "".concat(uniqueId),
    role: "tabpanel",
    style: styles,
    tabIndex: "-1",
    className: className
  }, children);
};

TabPanel.propTypes = propTypes;
TabPanel.defaultProps = defaultProps;
var _default = TabPanel;
exports["default"] = _default;
