// @vds-core/typography/dist/cjs/components/Subtitle/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _fonts = require("../../fonts");

function getStyles(typescale, viewport) {
  return SubtitleConfig[typescale][viewport];
}

var SubtitleConfig = {
  VDS: {
    mobile: {
      fontSize: 20,
      lineHeight: 24,
      fontWeight: 'normal',
      fontFamily: _fonts.Fonts.NeueHaasGroteskText
    },
    desktop: {
      fontSize: 20,
      lineHeight: 24,
      fontWeight: 'normal',
      fontFamily: _fonts.Fonts.NeueHaasGroteskText
    }
  },
  MVP: {
    mobile: {
      fontSize: 16,
      lineHeight: 19.2,
      fontWeight: '400',
      // 55 Roman replacement
      fontFamily: _fonts.Fonts.NeueHaasGroteskDisplay
    },
    desktop: {
      fontSize: 20,
      lineHeight: 24,
      fontWeight: '400',
      // 55 Roman replacement
      fontFamily: _fonts.Fonts.NeueHaasGroteskDisplay
    }
  },
  Marketing: {
    mobile: {
      fontSize: 16,
      lineHeight: 16,
      fontWeight: '400',
      // 55 Roman replacement
      fontFamily: _fonts.Fonts.NeueHaasGroteskDisplay
    },
    desktop: {
      fontSize: 20,
      lineHeight: 20,
      fontWeight: '400',
      // 55 Roman replacement
      fontFamily: _fonts.Fonts.NeueHaasGroteskDisplay
    }
  },
  getStyles: getStyles
};
var _default = SubtitleConfig;
exports["default"] = _default;
