// @vds-core/typography/dist/cjs/components/Subtitle/Subtitle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _theme = require("@vds-core/theme");

var _typographyConfig = _interopRequireDefault(require("../../typographyConfig"));

var _config = _interopRequireDefault(require("./config"));

var _fontColorPropType = _interopRequireDefault(require("../../util/fontColorPropType"));

var _Typography = require("../Typography");

var _utilities = require("@vds-core/utilities");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  &:active,\n  &:focus,\n  &:hover {\n    outline: none;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  /**
   * Primitive used for the component.
   */
  primitive: _propTypes["default"].oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span', 'p']),

  /**
   * Viewport the font styling is based on. Desktop and tablet are the same size.
   */
  viewport: _propTypes["default"].oneOf(['desktop', 'tablet', 'mobile']),

  /**
   * Content that will be passed to the component.
   */
  children: _propTypes["default"].node,

  /**
   * Color of the text.
   */
  color: _fontColorPropType["default"],

  /**
   * Allows for specific typescale to be passed to override global config.
   */
  typescale: _propTypes["default"].oneOf(['VDS', 'MVP', 'Marketing']),

  /**
   * If the theme object contains the `typescale` key and no typescale prop was passed global typescale is overriden by theme.
   */
  theme: _propTypes["default"].object,

  /**
   * The tabIndex of the text.
   */
  tabIndex: _propTypes["default"].number,

  /**
   * @ignore
   */
  config: _propTypes["default"].object,

  /**
   * Assigns an Id to the component
   */
  id: _propTypes["default"].string,

  /**
   * Adds an aria-labelledby attribute to the component
   */
  ariaLabelledBy: _propTypes["default"].string
};
var defaultProps = {
  primitive: 'p',
  viewport: 'desktop',
  children: null,
  color: _theme.colors.black,
  config: _config["default"]
};
var StyledSubtitle = /*#__PURE__*/(0, _styledComponents["default"])(_Typography.Typography).withConfig({
  displayName: "StyledSubtitle",
  componentId: "ufzqxp-0"
})(_templateObject());

var Subtitle = function Subtitle(props) {
  var children = props.children,
      primitive = props.primitive,
      viewport = props.viewport,
      color = props.color,
      typescale = props.typescale,
      theme = props.theme,
      tabIndex = props.tabIndex,
      config = props.config,
      fontsConfigProp = props.fontsConfig,
      ariaLabel = props.ariaLabel,
      id = props.id,
      ariaLabelledBy = props.ariaLabelledBy;
  var // What typescale is being used for configuration
  typescaleToUse = typescale ? typescale : theme && theme.typescale ? theme.typescale : _typographyConfig["default"].getTypescale(),
      viewportToUse = viewport === 'mobile' ? 'mobile' : 'desktop',
      // What is the specific font configuration we are using
  fontConfig = config[typescaleToUse][viewportToUse];
  return /*#__PURE__*/_react["default"].createElement(StyledSubtitle, {
    fontSize: fontConfig.fontSize,
    fontWeight: fontConfig.fontWeight,
    fontFamily: fontConfig.fontFamily,
    lineheight: fontConfig.lineHeight,
    primitive: primitive,
    color: color,
    tabIndex: tabIndex,
    "aria-hidden": tabIndex < 0,
    fontsconfig: fontsConfigProp,
    "aria-label": ariaLabel,
    id: id,
    "aria-labelledby": ariaLabelledBy
  }, children);
};

Subtitle.propTypes = propTypes;
Subtitle.defaultProps = defaultProps;

var _default = (0, _utilities.withVDSManager)(Subtitle);

exports["default"] = _default;
