// @vds/tabs/dist/cjs/components/VDSTab.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _tabs = require("@vds-core/tabs");

var _typography = require("@vds/typography");

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var typescale = _typography.TypographyConfig.getTypescale();

var Tab = function Tab(props) {
  return /*#__PURE__*/_react["default"].createElement(_tabs.CoreTab, (0, _extends2["default"])({}, props, {
    focusState: true,
    minHeight: true,
    tabSpacing: (0, _javascriptStyleHelpers.calculateRem)(0, 24, 0, 0),
    titleConfig: _typography.TitleConfig,
    typescale: typescale
  }));
};

var _default = Tab;
exports["default"] = _default;
