// @vds/theme/dist/esm/verizonTheme/font.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { calculateRem } from '@uie/javascript-style-helpers';
import colors from './colors';
var fontFamilies = {
  fontFamily: '"Verizon-NHG-eTX", Helvetica, Arial, sans-serif',
  monospace: '"SFMono-Regular", "RobotoMono-Regular", monospace',
  sans: '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
  serif: 'serif',
  iconFontUrl: 'https://respframework.verizon.com/fonts/verizon-icons/fonts.css',
  iconFont: 'verizon-icons'
};
var fontStyles = {
  fontColor: colors.black,
  fontColorAlt: colors.white,
  fontSize: calculateRem(16)
};
var fontWeights = {
  light: 300,
  normal: 400,
  medium: 500,
  bold: 700
};
export default _objectSpread(_objectSpread(_objectSpread({}, fontFamilies), fontStyles), {}, {
  fontWeights: fontWeights,
  fontHeader: fontFamilies.fontFamily,
  fontBody: fontFamilies.sans,
  fontWeight: fontWeights
});
