// @vds-core/carousels/dist/esm/components/Carousel.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { calculateRem } from '@uie/javascript-style-helpers';
import CarouselWrapper from './CarouselWrapper';
var propTypes = {
  /**
   * The items in the carousel.
   */
  data: PropTypes.array.isRequired,

  /**
   * function that returns the item to be rendered in each slide.
   */
  renderItem: PropTypes.func.isRequired,

  /**
   * @ignore
   */
  dataLoc: PropTypes.string
};

var Carousel = function Carousel(props) {
  var data = props.data,
      renderItem = props.renderItem,
      rest = _objectWithoutProperties(props, ["data", "renderItem"]);

  var newData = _toConsumableArray(data);

  var popElement = newData.pop();
  newData.unshift(popElement);

  var _renderChildren = function _renderChildren() {
    return newData.map(function (item, index) {
      return renderItem({
        item: item,
        index: index
      });
    });
  };

  return /*#__PURE__*/React.createElement(CarouselWrapper, rest, _renderChildren());
};

Carousel.propTypes = propTypes;
export default Carousel;
