// @vds/buttons/dist/esm/components/TextLinkCaret.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { TextLinkCaret as CoreTextLinkCaret } from '@vds-core/buttons';
import { BodyConfig, TypographyConfig } from '@vds/typography';
import { IconData } from '@vds/icons';
import { colors } from '@vds/theme';

var TextLinkCaret = function TextLinkCaret(props) {
  var _extends2;

  var typescale = TypographyConfig.getTypescale();
  return /*#__PURE__*/React.createElement(CoreTextLinkCaret, _extends({}, props, (_extends2 = {
    useIcon: true,
    size: "large",
    calculateFontWeight: 700,
    padding: "0",
    hitArea: true,
    hoverState: true,
    focusState: true,
    activeState: true
  }, _defineProperty(_extends2, "size", "large"), _defineProperty(_extends2, "typescale", typescale), _defineProperty(_extends2, "fontConfig", BodyConfig), _defineProperty(_extends2, "iconConfig", IconData), _defineProperty(_extends2, "colorConfig", colors), _extends2)));
};

export default TextLinkCaret;
