// @vds-core/tooltips/dist/cjs/components/ScrollView.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactCustomScrollbar = _interopRequireDefault(require("@uie/react-custom-scrollbar"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _theme = require("@vds-core/theme");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ScrollView = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(ScrollView, _Component);

  var _super = _createSuper(ScrollView);

  function ScrollView(_props) {
    var _this;

    (0, _classCallCheck2["default"])(this, ScrollView);
    _this = _super.call(this, _props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_handleUpdate", function (values) {
      var top = values.top,
          scrollHeight = values.scrollHeight,
          clientHeight = values.clientHeight;

      _this.setState({
        top: top,
        scrollHeight: scrollHeight,
        clientHeight: clientHeight
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "View", function (_ref) {
      var style = _ref.style,
          props = (0, _objectWithoutProperties2["default"])(_ref, ["style"]);
      var viewStyle = {
        marginRight: -15,
        paddingRight: 25,
        overflowX: 'hidden',
        height: '100%',
        outline: 'none'
      };
      return /*#__PURE__*/_react["default"].createElement("span", (0, _extends2["default"])({
        id: "tolltip-scrollbarview",
        className: "VDS__",
        tabIndex: 1,
        style: _objectSpread(_objectSpread({}, style), viewStyle)
      }, props));
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "Track", function (_ref2) {
      var style = _ref2.style,
          props = (0, _objectWithoutProperties2["default"])(_ref2, ["style"]);
      var _this$state = _this.state,
          clientHeight = _this$state.clientHeight,
          scrollHeight = _this$state.scrollHeight;
      var hideScroll = scrollHeight <= clientHeight;
      var trackStyle = {
        display: hideScroll ? 'none' : 'block',
        backgroundColor: _theme.colors.coolGray3,
        top: 0,
        right: 0,
        bottom: 0,
        width: 4,
        cursor: 'pointer'
      };
      return /*#__PURE__*/_react["default"].createElement("span", (0, _extends2["default"])({
        id: "tooltip-scrollbar-track",
        className: "VDS__",
        style: _objectSpread(_objectSpread({}, style), trackStyle) // onMouseOver={this._changeBackground}
        // onMouseOut={this._changeBackground}

      }, props));
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "Thumb", function (_ref3) {
      var style = _ref3.style,
          props = (0, _objectWithoutProperties2["default"])(_ref3, ["style"]);
      var thumbStyle = {
        backgroundColor: _theme.colors.coolGray6,
        height: '100%',
        width: 4
      };
      return /*#__PURE__*/_react["default"].createElement("span", (0, _extends2["default"])({
        id: "tooltip-scrollbarthumb",
        className: "VDS__",
        style: _objectSpread(_objectSpread({}, style), thumbStyle)
      }, props));
    });
    _this.state = {
      clientHeight: null,
      scrollHeight: null,
      hover: false,
      top: null,
      scroll: false
    };
    _this.scrollRef = null;
    return _this;
  }

  (0, _createClass2["default"])(ScrollView, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.scrollRef.view.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var scrollStyle = {
        height: 210 - 24,
        width: "".concat((0, _javascriptStyleHelpers.calculateRem)(206))
      };
      return /*#__PURE__*/_react["default"].createElement(_reactCustomScrollbar["default"], (0, _extends2["default"])({
        id: "Tooltip-scrollview",
        className: "VDS__",
        ref: function ref(el) {
          return _this2.scrollRef = el;
        },
        renderView: this.View,
        renderTrackVertical: this.Track,
        renderThumbVertical: this.Thumb,
        renderTrackHorizontal: function renderTrackHorizontal(_ref4) {
          var style = _ref4.style,
              props = (0, _objectWithoutProperties2["default"])(_ref4, ["style"]);
          return /*#__PURE__*/_react["default"].createElement("span", {
            className: "VDS__"
          });
        },
        renderThumbHorizontal: function renderThumbHorizontal(_ref5) {
          var style = _ref5.style,
              props = (0, _objectWithoutProperties2["default"])(_ref5, ["style"]);
          return /*#__PURE__*/_react["default"].createElement("span", {
            className: "VDS__"
          });
        },
        onUpdate: this._handleUpdate,
        style: scrollStyle
      }, this.props), this.props.children);
    }
  }]);
  return ScrollView;
}(_react.Component);

var _default = ScrollView;
exports["default"] = _default;
