// @vds-core/tabs/dist/esm/components/Tab.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  position: relative;\n  display: inline-block;\n  outline: none;\n  margin: ", ";\n  min-width: ", ";\n  :last-child {\n    margin: 0;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import StyledButton from './TabButton';
import { calculateRem } from '@uie/javascript-style-helpers';
import { colors } from '@vds-core/theme';
import { TitleConfig, TypographyConfig } from '@vds-core/typography';
var propTypes = {
  /**
   * The text label of the tab.
   */
  label: PropTypes.string.isRequired,

  /**
   * An optional callback that is called when this Tab is clicked. Passes parameters (event, tabIndex).
   */
  onClick: PropTypes.func,

  /**
   * If provided, it will set this Tab to the Active Tab on render.<br><br><strong>Note:</strong> <br>This will override whatever was set on the Tab <code>selectedIndex</code> Prop.<br>Also, it is "Last One Wins", ie: if this is set on multiple Tab elements, the last element with it set will be the Active Tab.
   */
  active: PropTypes.bool,

  /**
   * @ignore
   */
  hide: PropTypes.oneOf([true, false, 0, 1, 2]),

  /**
   * Minimum Width of this Tab.
   */
  minWidth: PropTypes.string,

  /**
   * Option to make Accent Border 100% width of this Tab.
   */
  indicatorFillTab: PropTypes.bool,

  /**
   * Sets the Position of the Active/Hover Border Accent on this Tab.
   */
  indicatorPosition: PropTypes.oneOf(['bottom', 'top', 'none']),

  /**
   * If provided, this will be used instead of the random uniqueId generated by the Tabs Component that is used to associate the Tab <code>li</code> and <code>button</code> with the TabPanel <code>div</code> components.
   */
  uniqueId: PropTypes.string,

  /**
   * @ignore
   */
  className: PropTypes.string,

  /**
   * @ignore
   */
  handleVisibilityChange: PropTypes.func,

  /**
   * @ignore
   */
  onVisibilityChange: PropTypes.func,

  /**
   * Supplies a set height on the tab component.
   */
  tabHeight: PropTypes.string,

  /**
   * @deprecated
   * Minimum Width of this Tab.
   */
  minTabWidth: PropTypes.string,
  //DEPRECATED

  /**
   * @deprecated
   * Option to make Accent Border 100% width of this Tab.
   */
  accentFill: PropTypes.bool,
  // DEPRECATED

  /**
   * @deprecated
   * Sets the Position of the Active/Hover Border Accent on this Tab.
   */
  accentPosition: PropTypes.oneOf(['bottom', 'top', 'none']),
  //DEPRECATED

  /**
   * @deprecated
   * If provided, it will set this Tab to the Active Tab on render.<br><br><strong>Note:</strong> <br>This will override whatever was set on the Tab <code>activeTabIndex</code> Prop.<br>Also, it is "Last One Wins", ie: if this is set on multiple Tab elements, the last element with it set will be the Active Tab.
   */
  isActive: PropTypes.bool,
  // DEPRECATED

  /**
   * @ignore
   */
  focusState: PropTypes.bool,

  /**
   * @ignore
   */
  minHeight: PropTypes.bool,

  /**
   * @ignore
   */
  typescale: PropTypes.string,

  /**
   * @ignore
   */
  titleConfig: PropTypes.object,

  /**
   * @ignore
   */
  tabSpacing: PropTypes.string,

  /**
   * @ignore
   */
  colorConfig: PropTypes.object,

  /**
   * Allows a string to be provided for analytics.
   */
  'data-analyticstrack': PropTypes.string,

  /**
   * Allows a string to be provided for analytics.
   */
  'data-track': PropTypes.string,

  /**
   * Allows a string to be provided for click stream.
   */
  'data-clickstream': PropTypes.string
};
var defaultProps = {
  hide: false,
  minWidth: undefined,
  className: null,
  typescale: TypographyConfig.getTypescale(),
  titleConfig: TitleConfig,
  tabSpacing: calculateRem(0, 40, 0, 0),
  colorConfig: colors
};
var StyledTab = /*#__PURE__*/styled.li.withConfig({
  displayName: "StyledTab",
  componentId: "eyuyma-0"
})(_templateObject(), function (_ref) {
  var tabSpacing = _ref.tabSpacing;
  return tabSpacing;
}, function (_ref2) {
  var minWidth = _ref2.minWidth,
      minTabWidth = _ref2.minTabWidth;
  return minWidth ? minWidth : minTabWidth;
});
export var Tab = /*#__PURE__*/function (_Component) {
  _inherits(Tab, _Component);

  var _super = _createSuper(Tab);

  function Tab(props) {
    var _this;

    _classCallCheck(this, Tab);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "_onClick", function (e) {
      if (e.detail === 0) e.preventDefault();

      _this.props.onClick(e);
    });

    _this.handleVisibilityChange = props.handleVisibilityChange || props.onVisibilityChange;
    _this.state = {
      selected: props.isActive || props.active
    };
    return _this;
  }

  _createClass(Tab, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(oldProps) {
      if (!this.props.handleVisibilityChange) {
        return;
      }

      if (this.props.isActive !== oldProps.isActive || this.props.selected !== oldProps.selected) {
        var active = this.props.isActive || this.props.selected;
        this.setState({
          selected: active
        });
        this.handleVisibilityChange('isActive', active);
      }

      if (this.props.hide !== oldProps.hide) {
        this.handleVisibilityChange('hide', this.props.hide);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          accentFill = _this$props.accentFill,
          indicatorFillTab = _this$props.indicatorFillTab,
          accentPosition = _this$props.accentPosition,
          indicatorPosition = _this$props.indicatorPosition,
          hide = _this$props.hide,
          label = _this$props.label,
          uniqueId = _this$props.uniqueId,
          minWidth = _this$props.minWidth,
          minTabWidth = _this$props.minTabWidth,
          onClick = _this$props.onClick,
          className = _this$props.className,
          tabHeight = _this$props.tabHeight,
          focusState = _this$props.focusState,
          minHeight = _this$props.minHeight,
          height = _this$props.height,
          typescale = _this$props.typescale,
          titleConfig = _this$props.titleConfig,
          tabSpacing = _this$props.tabSpacing,
          colorConfig = _this$props.colorConfig,
          analyticsTrack = _this$props.analyticsTrack,
          track = _this$props.track,
          clickStream = _this$props.clickStream;

      if (hide) {
        return null;
      }

      return /*#__PURE__*/React.createElement(StyledTab, {
        tabSpacing: tabSpacing,
        id: "tab_".concat(uniqueId),
        minWidth: minWidth,
        minTabWidth: minTabWidth,
        onClick: this._onClick,
        role: "presentation",
        className: className
      }, /*#__PURE__*/React.createElement(StyledButton, {
        accentFill: accentFill,
        indicatorFillTab: indicatorFillTab,
        accentPosition: accentPosition,
        indicatorPosition: indicatorPosition,
        selected: this.state.selected,
        label: label,
        uniqueId: uniqueId,
        tabHeight: tabHeight,
        focusState: focusState,
        minHeight: minHeight,
        typescale: typescale,
        titleConfig: titleConfig,
        colorConfig: colorConfig,
        clickStream: clickStream,
        analyticsTrack: analyticsTrack,
        track: track
      }, label));
    }
  }]);

  return Tab;
}(Component);
Tab.propTypes = propTypes;
Tab.defaultProps = defaultProps;
export default Tab;
