// @vds-core/typography/src/components/Body/Body.jsx
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { colors } from '@vds-core/theme';
import TypographyConfig from '../../typographyConfig';
import fontColorPropType from '../../util/fontColorPropType';
import { Typography } from '../Typography';
import { withVDSManager } from '@vds-core/utilities';
import defaultConfig from './config';

function getBodyFontWeight(typescale, bold) {
  switch (typescale) {
    case 'MVP':
    case 'VDS':
      return bold ? '750' : '400'; // 400 is mapped to match 55 Roman
    default:
      return bold ? 'bold' : 'normal';
  }
}
const propTypes = {
  /**
   * Primitive used for the component.
   */
  primitive: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span', 'p']),
  /**
   * Viewport the font styling is based on. Desktop and tablet are the same size.
   */
  viewport: PropTypes.oneOf(['desktop', 'tablet', 'mobile']),
  /**
   * Size of the typescale that will be used.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Content that will be passed to the component.
   */
  children: PropTypes.node,
  /**
   * The color of the text.
   */
  color: fontColorPropType,
  /**
   * Adds a fontWeight of bold.
   */
  bold: PropTypes.bool,
  /**
   * Allows for specific typescale to be passed to override global config.
   */
  typescale: PropTypes.oneOf(['VDS', 'MVP']),
  /**
   * If the theme object contains the `typescale` key and no typescale prop was passed global typescale is overriden by theme.
   */
  theme: PropTypes.object,
  /**
   * The tabIndex of the text.
   */
  tabIndex: PropTypes.number,
  /**
   * @ignore
   */
  config: PropTypes.object,
  /**
   * @ignore
   */
  getfontWeight: PropTypes.func,
  /**
   * Assigns an Id to the component
   */
  id: PropTypes.string,
  /**
   * Adds an aria-labelledby attribute to the component
   */
  ariaLabelledBy: PropTypes.string,
};

const defaultProps = {
  primitive: 'p',
  viewport: 'desktop',
  size: 'small',
  children: null,
  color: colors.black,
  bold: false,
  config: defaultConfig,
  getFontWeight: getBodyFontWeight,
};

const StyledBody = styled(Typography)`
  &:active,
  &:focus,
  &:hover {
    outline: none;
  }
`;

const Body = props => {
  const {
    config,
    children,
    primitive,
    viewport,
    size,
    color,
    bold,
    typescale,
    theme,
    tabIndex,
    getFontWeight,
    fontsConfig: fontsConfigProp,
    ariaLabel,
    id,
    ariaLabelledBy,
  } = props;
  let // What typescale is being used for configuration
    typescaleToUse = typescale
      ? typescale
      : theme && theme.typescale
      ? theme.typescale
      : TypographyConfig.getTypescale(),
    viewportToUse = viewport === 'mobile' ? 'mobile' : 'desktop',
    // What is the specific font configuration we are using
    fontConfig = config[typescaleToUse][viewportToUse][size],
    fontWeight = getFontWeight(typescaleToUse, bold);

  return (
    <StyledBody
      fontSize={fontConfig.fontSize}
      fontWeight={fontWeight}
      fontFamily={fontConfig.fontFamily}
      lineheight={fontConfig.lineHeight}
      letterSpacing={fontConfig.letterSpacing && fontConfig.letterSpacing}
      primitive={primitive}
      color={color}
      tabIndex={tabIndex}
      aria-hidden={tabIndex < 0}
      fontsconfig={fontsConfigProp}
      aria-label={ariaLabel}
      id={id}
      aria-labelledby={ariaLabelledBy}
    >
      {children}
    </StyledBody>
  );
};

Body.propTypes = propTypes;
Body.defaultProps = defaultProps;

export default withVDSManager(Body);
