// @vds/typography/dist/cjs/components/Title/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _fonts = require("../../fonts");

function getStyles(typescale, viewport, typeSize) {
  return TitleConfig[typescale][viewport][typeSize];
}

var TitleConfig = {
  VDS: {
    desktop: {
      '2XLarge': {
        fontSize: 64,
        lineHeight: 64,
        fontWeight: 700,
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      XLarge: {
        fontSize: 48,
        lineHeight: 48,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      large: {
        fontSize: 32,
        lineHeight: 32,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      medium: {
        fontSize: 24,
        lineHeight: 24,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      small: {
        fontSize: 20,
        lineHeight: 20,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      XSmall: {
        fontSize: 14,
        lineHeight: 16,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      }
    },
    mobile: {
      '2XLarge': {
        fontSize: 36,
        lineHeight: 36,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      XLarge: {
        fontSize: 32,
        lineHeight: 32,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      large: {
        fontSize: 24,
        lineHeight: 24,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      medium: {
        fontSize: 20,
        lineHeight: 20,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      small: {
        fontSize: 16,
        lineHeight: 16,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      XSmall: {
        fontSize: 14,
        lineHeight: 16,
        fontWeight: 'bold',
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      }
    }
  },
  getStyles: getStyles
};
var _default = TitleConfig;
exports["default"] = _default;
