// @vds/selects/dist/cjs/components/DropdownSelectMarketing.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _theme = require("@vds/theme");

var _icons = require("@vds/icons");

var _selects = require("@vds-core/selects");

var _typography = require("@vds/typography");

var propTypes = {
  /**
   * @ignore
   */
  color: _propTypes["default"].object,

  /**
   * @ignore
   */
  typescale: _propTypes["default"].string,
  // string that sets the typescale for the typography (example: typescale === 'VDS')

  /**
   * @ignore
   */
  bodyConfig: _propTypes["default"].object,
  // config for body typography

  /**
   * @ignore
   */
  titleConfig: _propTypes["default"].object,
  // config for title typography

  /**
   * @ignore
   */
  iconActiveState: _propTypes["default"].bool,
  // if true, the caret-down icon changes to caret-up if this.state.open is also true

  /**
   * @ignore
   */
  calculateErrorLabelSpacing: _propTypes["default"].func,
  // calculates the margin on top of the error text

  /**
   * @ignore
   */
  calculateIconSpacing: _propTypes["default"].func,
  // calculates the padding next to the icon

  /**
   * @ignore
   */
  hitArea: _propTypes["default"].bool,
  // adds additional hit area to the trigger

  /**
   * @ignore
   */
  triggerPaddingBottom: _propTypes["default"].string,
  // used to set padding underneath the trigger

  /**
   * @ignore
   */
  calculateListItemHeight: _propTypes["default"].func,
  // function that calculates the height of the drop down list item

  /**
   * @ignore
   */
  calculateListItemPadding: _propTypes["default"].func,
  // function that calculates the padding inside of each list item

  /**
   * @ignore
   */
  hoverState: _propTypes["default"].bool,
  // boolean that, if true, changes the hover states

  /**
   * @ignore
   */
  focusState: _propTypes["default"].bool,
  // boolean that, if true, changes the focus states

  /**
   * @ignore
   */
  iconConfig: _propTypes["default"].object // passes through data for icons

};
var defaultProps = {};

var DropdownSelectMarketing = function DropdownSelectMarketing(props) {
  function _calculateListItemHeight(size) {
    if (size === 'small') {
      return (0, _javascriptStyleHelpers.calculateRem)(44);
    } else return (0, _javascriptStyleHelpers.calculateRem)(56);
  }

  function _calculateIconSpacing() {
    return (0, _theme.calculateSpacing)('4X');
  }

  function _calculateListItemPadding(size, viewport) {
    if (viewport === 'mobile') {
      if (size === 'small') {
        return (0, _javascriptStyleHelpers.calculateRem)(14, 12);
      } else return (0, _javascriptStyleHelpers.calculateRem)(16, 12);
    } else return (0, _javascriptStyleHelpers.calculateRem)(11, 12);
  }

  function _calculateErrorLabelSpacing() {
    return (0, _javascriptStyleHelpers.calculateRem)(4);
  }

  return /*#__PURE__*/_react["default"].createElement(_selects.DropdownSelectMarketing, (0, _extends2["default"])({}, props, {
    hitArea: true,
    hoverState: true,
    focusState: true,
    iconActiveState: true,
    typescale: _typography.TypographyConfig.getTypescale(),
    iconConfig: _icons.IconData,
    color: _theme.colors,
    bodyConfig: _typography.BodyConfig,
    titleConfig: _typography.TitleConfig,
    triggerPaddingBottom: (0, _theme.calculateSpacing)('1X'),
    calculateErrorLabelSpacing: _calculateErrorLabelSpacing,
    calculateIconSpacing: _calculateIconSpacing,
    calculateListItemHeight: _calculateListItemHeight,
    calculateListItemPadding: _calculateListItemPadding
  }));
};

DropdownSelectMarketing.propTypes = propTypes;
DropdownSelectMarketing.defaultProps = defaultProps;
var _default = DropdownSelectMarketing;
exports["default"] = _default;
