// @vds-core/tabs/dist/cjs/configs/CatNavBoxed.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCatNavConfig = void 0;

var onClickItem = function onClickItem() {
  console.log('clicked!');
};

var getCatNavConfig = function getCatNavConfig(items) {
  if (items === 2) return BoxedDouble;
  return BoxedTriple;
};

exports.getCatNavConfig = getCatNavConfig;
var BoxedTriple = {
  desktop: {
    inverted: false,
    layout: 'triple',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }]
  },
  desktopSmall: {
    inverted: false,
    layout: 'triple',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }]
  },
  tablet: {
    inverted: false,
    layout: 'triple',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'medium',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'medium',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'medium',
        inverted: false
      },
      onClick: onClickItem
    }]
  },
  mobile: {
    inverted: false,
    layout: 'triple',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }]
  },
  mobileSmall: {
    inverted: false,
    layout: 'triple',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }]
  }
};
var BoxedDouble = {
  desktop: {
    inverted: false,
    layout: 'double',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        size: 'XLarge',
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'extraLarge',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        size: 'XLarge',
        text: 'Home internet',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'extraLarge',
        inverted: false
      },
      onClick: onClickItem
    }]
  },
  desktopSmall: {
    inverted: false,
    layout: 'double',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        size: 'XLarge',
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'extraLarge',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        size: 'XLarge',
        text: 'Home internet',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'extraLarge',
        inverted: false
      },
      onClick: onClickItem
    }]
  },
  tablet: {
    inverted: false,
    layout: 'double',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        size: 'large',
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'medium',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        size: 'large',
        text: 'Home internet',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'medium',
        inverted: false
      },
      onClick: onClickItem
    }]
  },
  mobile: {
    inverted: false,
    layout: 'double',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        size: 'large',
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        size: 'large',
        text: 'Home internet',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  mobileSmall: {
    inverted: false,
    layout: 'double',
    layoutStyle: 'boxed',
    data: [{
      title: {
        inverted: false,
        size: 'large',
        text: 'Lorem',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }, {
      title: {
        inverted: false,
        size: 'large',
        text: 'Home internet',
        width: '100%',
        color: '#000000'
      },
      arrow: {
        size: 'large',
        inverted: false
      },
      onClick: onClickItem
    }]
  }
};
