// @vds-core/tabs/dist/cjs/configs/FeatureCardConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FeatureCardConfig = void 0;
var mobileSizes = 'mobile-small';
var mobileSmallImages = mobileSizes === 'mobile-small' ? 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F5x6?wid=479' : 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F5x7?wid=479';
var sources = [{
  srcset: mobileSmallImages,
  type: 'image/webp',
  maxViewport: 'mobile'
}, {
  srcset: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F16x9?wid=767',
  type: 'image/webp',
  maxViewport: 'tablet'
}, {
  srcset: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F16x9?wid=991',
  type: 'image/webp',
  maxViewport: 'desktopSmall'
}, {
  srcset: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F9x4?wid=1079',
  type: 'image/webp',
  maxViewport: 'desktop'
}, {
  srcset: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F9x4?wid=1272',
  type: 'image/webp',
  maxViewport: 'desktop'
}];
var FeatureCardConfig = {
  desktop: {
    alignment: 'left',
    verticalAlignment: 'center',
    aspectRatio: '16:9',
    contentInset: true,
    paddingLeft: '20px',
    paddingRight: '20px',
    inverted: true,
    backgroundImage: {
      url: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F9x4?wid=1272',
      sources: sources
    },
    title: {
      text: 'Marketing and Customer Experience',
      width: '428px',
      marginBottom: '16px',
      size: 'XLarge'
    },
    subtitle: {
      text: 'Small businesses can cover a lot of ground. Get mobile solutions that keep you working everywhere you do business.',
      textComponent: 'body',
      size: 'large',
      width: '428px',
      marginBottom: '32px'
    },
    CTAGroup: {
      alignment: 'left',
      flexDirection: 'row',
      data: [{
        children: 'See all 5G products',
        onClick: function onClick() {
          alert('clicked link');
        },
        ctaType: 'textLinkCaret'
      }],
      linkPaddingBottom: '0px'
    }
  },
  desktopSmall: {
    alignment: 'left',
    verticalAlignment: 'center',
    inverted: true,
    contentInset: true,
    paddingLeft: '20px',
    paddingRight: '20px',
    aspectRatio: '16:9',
    backgroundImage: {
      url: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F9x4?wid=1272',
      sources: sources
    },
    title: {
      text: 'Marketing and Customer Experience',
      width: '428px',
      marginBottom: '16px',
      size: 'XLarge'
    },
    subtitle: {
      text: 'Small businesses can cover a lot of ground. Get mobile solutions that keep you working everywhere you do business.',
      textComponent: 'body',
      size: 'large',
      width: '428px',
      marginBottom: '32px'
    },
    CTAGroup: {
      alignment: 'left',
      flexDirection: 'row',
      data: [{
        children: 'See all 5G products',
        onClick: function onClick() {
          alert('clicked link');
        },
        ctaType: 'textLinkCaret'
      }],
      linkPaddingBottom: '0px'
    }
  },
  tablet: {
    alignment: 'left',
    verticalAlignment: 'center',
    aspectRatio: '16:9',
    contentInset: true,
    paddingLeft: '20px',
    paddingRight: '20px',
    inverted: true,
    backgroundImage: {
      url: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F9x4?wid=1272',
      sources: sources
    },
    title: {
      text: 'Marketing and Customer Experience',
      width: '328px',
      marginBottom: '16px',
      size: 'XLarge'
    },
    subtitle: {
      text: 'Small businesses can cover a lot of ground. Get mobile solutions that keep you working everywhere you do business.',
      textComponent: 'body',
      size: 'large',
      width: '328px',
      marginBottom: '32px'
    },
    CTAGroup: {
      alignment: 'left',
      flexDirection: 'row',
      data: [{
        children: 'See all 5G products',
        onClick: function onClick() {
          alert('clicked link');
        },
        ctaType: 'textLinkCaret'
      }],
      linkPaddingBottom: '0px'
    }
  },
  mobile: {
    alignment: 'left',
    verticalAlignment: 'center',
    inverted: true,
    aspectRatioMobile: '16:9',
    backgroundImage: {
      url: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F9x4?wid=1272',
      sources: sources
    },
    title: {
      text: 'Marketing and Customer Experience',
      width: '328px',
      marginBottom: '12px',
      size: 'XLarge'
    },
    subtitle: {
      text: 'Small businesses can cover a lot of ground. Get mobile solutions that keep you working everywhere you do business.',
      textComponent: 'body',
      size: 'large',
      width: '328px',
      marginBottom: '24px'
    },
    CTAGroup: {
      alignment: 'left',
      flexDirection: 'row',
      data: [{
        children: 'See all 5G products',
        onClick: function onClick() {
          alert('clicked link');
        },
        ctaType: 'textLinkCaret'
      }],
      linkPaddingBottom: '0px'
    }
  },
  mobileSmall: {
    alignment: 'left',
    aspectRatioMobile: '5:6',
    verticalAlignment: 'center',
    inverted: true,
    backgroundImage: {
      url: 'https://ss7.vzw.com/is/image/VerizonWireless/feature%5Fcard%5F9x4?wid=1272',
      sources: sources
    },
    title: {
      text: 'Marketing and Customer Experience',
      width: '240px',
      marginBottom: '12px',
      size: 'XLarge'
    },
    subtitle: {
      text: 'Small businesses can cover a lot of ground. Get mobile solutions that keep you working everywhere you do business.',
      textComponent: 'body',
      size: 'large',
      width: '240px',
      marginBottom: '24px'
    },
    CTAGroup: {
      alignment: 'left',
      flexDirection: 'row',
      data: [{
        children: 'See all 5G products',
        onClick: function onClick() {
          alert('clicked link');
        },
        ctaType: 'textLinkCaret'
      }],
      linkPaddingBottom: '0px'
    }
  }
};
exports.FeatureCardConfig = FeatureCardConfig;
