// @vds-core/tabs/dist/cjs/configs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "QuadrupleDefaultData", {
  enumerable: true,
  get: function get() {
    return _QuadrupleDefault.QuadrupleDefaultData;
  }
});
Object.defineProperty(exports, "TripleDefaultData", {
  enumerable: true,
  get: function get() {
    return _TripleDefault.TripleDefaultData;
  }
});
Object.defineProperty(exports, "DoubleDefaultData", {
  enumerable: true,
  get: function get() {
    return _DoubleDefault.DoubleDefaultData;
  }
});
Object.defineProperty(exports, "QuadrupleInvertedData", {
  enumerable: true,
  get: function get() {
    return _QuadrupleInverted.QuadrupleInvertedData;
  }
});
Object.defineProperty(exports, "FeatureCardConfig", {
  enumerable: true,
  get: function get() {
    return _FeatureCardConfig.FeatureCardConfig;
  }
});
Object.defineProperty(exports, "TileTextConfig", {
  enumerable: true,
  get: function get() {
    return _TileTextConfig.TileTextConfig;
  }
});
Object.defineProperty(exports, "TileTextDoubleConfig", {
  enumerable: true,
  get: function get() {
    return _TileTextConfig.TileTextDoubleConfig;
  }
});
Object.defineProperty(exports, "getCatNavConfig", {
  enumerable: true,
  get: function get() {
    return _CatNavBoxed.getCatNavConfig;
  }
});
Object.defineProperty(exports, "getCTAGroup", {
  enumerable: true,
  get: function get() {
    return _getCTAGroup.getCTAGroup;
  }
});

var _QuadrupleDefault = require("./QuadrupleDefault");

var _TripleDefault = require("./TripleDefault");

var _DoubleDefault = require("./DoubleDefault");

var _QuadrupleInverted = require("./QuadrupleInverted");

var _FeatureCardConfig = require("./FeatureCardConfig");

var _TileTextConfig = require("./TileTextConfig");

var _CatNavBoxed = require("./CatNavBoxed");

var _getCTAGroup = require("./getCTAGroup");
