// @vds-core/tabs/dist/cjs/configs/TileTextConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TileTextDoubleConfig = exports.TileTextConfig = void 0;
var TileTextConfig = {
  desktop: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: 'Inseego 5G MiFi',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  desktopSmall: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: '5G Ultra Wideband',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: 'Inseego 5G MiFi',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  tablet: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: '5G Ultra Wideband',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: 'Inseego 5G MiFi',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  mobile: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '68px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  mobileSmall: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '68px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  }
};
exports.TileTextConfig = TileTextConfig;
var TileTextDoubleConfig = {
  desktop: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  desktopSmall: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: '5G Ultra Wideband',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  tablet: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: '5G Ultra Wideband',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'large',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '16px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '32px'
      },
      textLinkCaret: {
        text: 'Lorem ipsum',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  mobile: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  },
  mobileSmall: {
    layoutStyle: 'boxed',
    boxColor: '#f6f6f6',
    layout: 'triple',
    data: [{
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: '5G Ultra Wideband',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }, {
      textWidth: '100%',
      title: {
        viewport: 'mobile',
        size: 'XLarge',
        text: 'Samsung Galaxy S20 Ultra 5G',
        paddingBottom: '12px'
      },
      subtitle: {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        size: 'large',
        viewport: 'desktop',
        paddingBottom: '24px'
      },
      textLinkCaret: {
        text: 'Learn more',
        viewport: 'desktop',
        onClick: function onClick() {
          return console.log();
        }
      }
    }]
  }
};
exports.TileTextDoubleConfig = TileTextDoubleConfig;
