// @vds-core/tabs/dist/cjs/configs/TripleDefault.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TripleDefaultData = void 0;

var _theme = require("@vds-marketing/theme");

var titleText = 'Lorem ipsum dolor sit ame';
var bodyText = 'Lorem ipsum dolor sit amet consectetur adipisicing eli';

var onClickItem = function onClickItem() {
  return alert('You clicked an item to go somewhere!');
};

var TripleDefaultData = {
  desktop: {
    paddingRight: '20X',
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: false,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: false,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        color: _theme.colors.black,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  desktopSmall: {
    paddingRight: '20X',
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: false,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: false,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        color: _theme.colors.black,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  tablet: {
    paddingRight: '20X',
    data: [{
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: false,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        inverted: false,
        type: 'arrow',
        paddingBottom: '32X'
      },
      onClick: onClickItem
    }, {
      line: {
        type: 'light',
        paddingBottom: '12X',
        inverted: false
      },
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          text: titleText,
          paddingBottom: '8X',
          width: '76%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '76%'
        }]
      },
      arrow: {
        name: 'arrow-right',
        size: 'large',
        medium: true,
        color: _theme.colors.black,
        type: 'arrow',
        paddingBottom: '32X'
      },
      lastChild: true,
      onClick: onClickItem
    }]
  },
  mobile: {
    iconSize: 'extraLarge',
    data: [{
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          viewport: 'desktop',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          viewport: 'desktop',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          viewport: 'desktop',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '32X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }]
  },
  mobileSmall: {
    iconSize: 'extraLarge',
    data: [{
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          viewport: 'desktop',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          viewport: 'desktop',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '16X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }, {
      sectionTitle: {
        data: [{
          type: 'title',
          size: 'XLarge',
          viewport: 'desktop',
          text: titleText,
          paddingTop: '16X',
          paddingBottom: '6X',
          width: '72%'
        }, {
          type: 'subtitle',
          text: bodyText,
          paddingBottom: '32X',
          width: '72%'
        }]
      },
      onClick: onClickItem
    }]
  }
};
exports.TripleDefaultData = TripleDefaultData;
