// @vds/typography/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TitleConfig", {
  enumerable: true,
  get: function get() {
    return _Title.TitleConfig;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function get() {
    return _Title.Title;
  }
});
Object.defineProperty(exports, "SubtitleConfig", {
  enumerable: true,
  get: function get() {
    return _Subtitle.SubtitleConfig;
  }
});
Object.defineProperty(exports, "Subtitle", {
  enumerable: true,
  get: function get() {
    return _Subtitle.Subtitle;
  }
});
Object.defineProperty(exports, "BodyConfig", {
  enumerable: true,
  get: function get() {
    return _Body.BodyConfig;
  }
});
Object.defineProperty(exports, "Body", {
  enumerable: true,
  get: function get() {
    return _Body.Body;
  }
});
Object.defineProperty(exports, "MicroConfig", {
  enumerable: true,
  get: function get() {
    return _Micro.MicroConfig;
  }
});
Object.defineProperty(exports, "Micro", {
  enumerable: true,
  get: function get() {
    return _Micro.Micro;
  }
});
Object.defineProperty(exports, "Typography", {
  enumerable: true,
  get: function get() {
    return _Typography.Typography;
  }
});
Object.defineProperty(exports, "Fonts", {
  enumerable: true,
  get: function get() {
    return _fonts.Fonts;
  }
});
Object.defineProperty(exports, "TypographyConfig", {
  enumerable: true,
  get: function get() {
    return _typographyConfig["default"];
  }
});
Object.defineProperty(exports, "fontColorPropType", {
  enumerable: true,
  get: function get() {
    return _fontColorPropType["default"];
  }
});

var _Title = require("./components/Title");

var _Subtitle = require("./components/Subtitle");

var _Body = require("./components/Body");

var _Micro = require("./components/Micro");

var _Typography = require("./components/Typography");

var _fonts = require("./fonts");

var _typographyConfig = _interopRequireDefault(require("./typographyConfig"));

var _fontColorPropType = _interopRequireDefault(require("./util/fontColorPropType"));
