// @vds/buttons/dist/cjs/components/ButtonGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _buttons = require("@vds-core/buttons");

var _Button = _interopRequireDefault(require("./Button"));

var ButtonGroup = function ButtonGroup(props) {
  function _onClick(e, onClick) {
    if (e.detail !== 0) {
      onClick && onClick(e);
      e.target.blur();
    } else if (e.type === 'click') {
      onClick && onClick(e);
    }
  }

  function _renderButton(props) {
    var type = props.type,
        children = props.children,
        size = props.size,
        width = props.width,
        _onClick2 = props.onClick;

    if (type === 'primary') {
      return /*#__PURE__*/_react["default"].createElement(_Button["default"], (0, _extends2["default"])({}, props, {
        primary: true,
        style: {
          maxWidth: '100%'
        },
        ariaLabel: children,
        children: children,
        size: size,
        width: size !== 'small' && '100%',
        onClick: function onClick(e) {
          return _onClick(e, _onClick2);
        }
      }));
    } else if (type === 'secondary') {
      return /*#__PURE__*/_react["default"].createElement(_Button["default"], (0, _extends2["default"])({}, props, {
        secondary: true,
        style: {
          maxWidth: '100%'
        },
        ariaLabel: children,
        children: children,
        size: size,
        width: size !== 'small' && '100%',
        onClick: function onClick(e) {
          return _onClick(e, _onClick2);
        }
      }));
    }
  }

  return /*#__PURE__*/_react["default"].createElement(_buttons.ButtonGroup, (0, _extends2["default"])({}, props, {
    renderButton: _renderButton,
    maxWidth: "596px",
    smallButtonSpacing: "4X"
  }));
};

var _default = ButtonGroup;
exports["default"] = _default;
