// @vds/buttons/dist/cjs/components/TextLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _buttons = require("@vds-core/buttons");

var _typography = require("@vds/typography");

var _theme = require("@vds/theme");

var propTypes = {
  size: _propTypes["default"].oneOf(['large', 'small'])
};

var TextLink = function TextLink(props) {
  var size = props.size;
  return /*#__PURE__*/_react["default"].createElement(_buttons.TextLink, (0, _extends2["default"])({}, props, {
    focusState: true,
    hitArea: true,
    typescale: _typography.TypographyConfig.getTypescale(),
    size: size,
    colorConfig: _theme.colors,
    bodyConfig: _typography.BodyConfig,
    activeState: false
  }));
};

TextLink.propTypes = propTypes;
var _default = TextLink;
exports["default"] = _default;
