// @vds/buttons/dist/cjs/components/TextLinkCaret.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _buttons = require("@vds-core/buttons");

var _typography = require("@vds/typography");

var _icons = require("@vds/icons");

var _theme = require("@vds/theme");

var TextLinkCaret = function TextLinkCaret(props) {
  var _extends2;

  var typescale = _typography.TypographyConfig.getTypescale();

  return /*#__PURE__*/_react["default"].createElement(_buttons.TextLinkCaret, (0, _extends3["default"])({}, props, (_extends2 = {
    useIcon: true,
    size: "large",
    calculateFontWeight: 700,
    padding: "0",
    hitArea: true,
    hoverState: true,
    focusState: true,
    activeState: true
  }, (0, _defineProperty2["default"])(_extends2, "size", "large"), (0, _defineProperty2["default"])(_extends2, "typescale", typescale), (0, _defineProperty2["default"])(_extends2, "fontConfig", _typography.BodyConfig), (0, _defineProperty2["default"])(_extends2, "iconConfig", _icons.IconData), (0, _defineProperty2["default"])(_extends2, "colorConfig", _theme.colors), _extends2)));
};

var _default = TextLinkCaret;
exports["default"] = _default;
