// @vds-core/grids/src/components/Grid.jsx
/* eslint-disable no-console */
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { calculateRem } from '@uie/javascript-style-helpers';
import { breakpoints } from '@vds-core/theme';
import { Grid as FlexboxGrid } from 'react-styled-flexboxgrid';
import { withVDSManager } from '@vds-core/utilities';

// ---------------------------------------------
// Props
// ---------------------------------------------

const propTypes = {
  /**
   * Children for the grid. This should be the VDS Row components,
   * containing the VDS Col components. The page content should be inside the Col components.
   */
  children: PropTypes.node.isRequired,
  /**
   * Set a class on the Grid container
   */
  className: PropTypes.string,
  /**
   * Set a fixed height for the grid container
   */
  height: PropTypes.string,
  /**
   * @Ignore
   */
  vertical: PropTypes.bool,
  /**
   * @Ignore
   */
  align: PropTypes.oneOf([
    'flex-start',
    'space-between',
    'space-around',
    'flex-end',
    'center',
    'space-evenly',
  ]),
  /**
   * When provided, will set global sizes for all Col components for each viewport [mobile, tablet, desktop]
   */
  colSizes: PropTypes.object,
  /**
   * Set the styling of the component to the desktop, tablet, or mobile size based on passed prop.
   * Desktop and tablet are the same size.
   */
  viewport: PropTypes.oneOf(['desktop', 'tablet', 'mobile']),
  /**
   * @Ignore
   */
  colGutter: PropTypes.bool,
  /**
   * When passed, will set the global spacing between all Row components inside the grid
   */
  rowGutter: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

const defaultProps = {
  className: null,
  colGutter: true,
  rowGutter: 0,
  height: '100%',
  vertical: false,
  colSizes: { mobile: 1, tablet: 1, desktop: 1 },
  viewport: undefined,
  rowGutter: undefined,
  align: 'flex-start',
};

const baseEm = 16;

const MOBILE_GUTTER = 32;
const DESKTOP_GUTTER = 40;

function _emNumberFromBreakPoint(breakpoint) {
  let parsedBreak = Number(breakpoint.split('px', 1)),
    number = parsedBreak / baseEm;
  return number;
}

const StyledGridContainer = styled.div`
  width: 100%;
  max-width: 1272px;
  margin: 0 auto;
  ${({ height }) =>
    height &&
    `
    height: ${height};
  `};
  ${({ align, vertical }) =>
    vertical &&
    `
    justify-content: ${align};
    padding-top: ${({ gutterWidth }) => calculateRem(gutterWidth / 2)};
    flex-flow: column nowrap;
  `};
`;

function getColumnSizes(child, parentSizes, viewport) {
  const childSizes =
    (child.props.colSizes && _mapColSizes(child.props.colSizes, viewport)) ||
    {};
  // remove undefined sizes
  Object.keys(childSizes).forEach(
    key => childSizes[key] === undefined && delete childSizes[key]
  );

  if (Object.keys(childSizes).length === 4) return childSizes;
  else return { ...parentSizes, ...childSizes };
}

function _mapColSizes(colSizes, viewport) {
  let xs = viewport ? colSizes[viewport] : colSizes.mobile,
    sm = viewport ? colSizes[viewport] : colSizes.mobile,
    md = viewport ? colSizes[viewport] : colSizes.tablet,
    lg = viewport ? colSizes[viewport] : colSizes.desktop;

  return {
    xs: xs,
    sm: sm,
    md: md,
    lg: lg,
  };
}

const Grid = props => {
  const {
    children,
    colGutter,
    rowGutter,
    vertical,
    viewport,
    colSizes,
    align,
    className,
    height,
  } = props;

  const innerWidth = window.innerWidth;
  const gutterWidth =
    viewport === 'mobile' || (innerWidth && innerWidth < 768)
      ? MOBILE_GUTTER
      : DESKTOP_GUTTER;
  const gutterWidthRem = calculateRem(gutterWidth);

  let mappedColSizes = _mapColSizes(colSizes, viewport);

  // sm: '544px',
  // md: '768px',
  // lg: '991px',
  // xl: '1272px',
  // tablet is sm -> md
  // desktop is md+
  const theme = {
    flexboxgrid: {
      gridSize: viewport === 'mobile' ? 4 : 12,
      gutterWidth: colGutter ? gutterWidthRem.replace('rem', '') : 0,
      mediaQuery: 'only screen',
      breakpoints: {
        xs: _emNumberFromBreakPoint('0px'), //0
        sm: _emNumberFromBreakPoint(breakpoints.md), //0 -> 767px mobile
        md: _emNumberFromBreakPoint(breakpoints.lg), //768 px -> 1024px tablet
        lg: _emNumberFromBreakPoint(breakpoints.xl), //1024 - 1272 desktop
      },
    },
  };

  return (
    <FlexboxGrid
      theme={theme}
      fluid={true}
      style={{
        paddingRight: 0,
        paddingLeft: 0,
        marginRight: 0,
        marginLeft: 0,
        maxWidth: '100%',
        fontSize: '16px',
        flex: 1,
      }}
    >
      <StyledGridContainer {...props} theme={theme} gutterWidth={gutterWidth}>
        {React.Children.map(children, (child, index) => {
          const styles = {
            flexDirection: vertical ? 'column' : 'row',
            justifyContent: align,
            marginRight: 0,
            marginLeft: 0,
            paddingRight: 0,
            paddingLeft: 0,
          };

          const gutteredChildren =
            child && child.props && child.props.children
              ? React.Children.map(
                  child.props.children,
                  (child, childIndex) => {
                    const sizes = getColumnSizes(
                      child,
                      mappedColSizes,
                      viewport
                    );
                    const gutter = {
                      marginBottom:
                        rowGutter && children.length - 1 !== index
                          ? rowGutter
                          : 0,
                    };
                    return React.cloneElement(child, {
                      key: childIndex + 1,
                      style: gutter,
                      theme: theme,
                      ...sizes,
                    });
                  }
                )
              : [];

          return React.cloneElement(
            child,
            { key: index + 1, style: styles, theme: theme },
            gutteredChildren
          );
        })}
      </StyledGridContainer>
    </FlexboxGrid>
  );
};

Grid.defaultProps = defaultProps;
Grid.propTypes = propTypes;

export default withVDSManager(Grid);
