// @vds-core/buttons/dist/cjs/components/ButtonGroupItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _theme = require("@vds-core/theme");

var _Button = _interopRequireDefault(require("./Button"));

var _TextLink = _interopRequireDefault(require("./TextLink"));

var _TextLinkCaret = _interopRequireDefault(require("./TextLinkCaret"));

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n  box-sizing: border-box;\n  width: ", ";\n  flex-shrink: 1;\n  outline: none;\n  ", "\n  &:last-child {\n    padding-bottom: ", "px;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _onClick(e, onClick) {
  if (e.detail !== 0) {
    onClick && onClick(e);
    e.target.blur();
  } else if (e.type === 'click') {
    onClick && onClick(e);
  }
}

function _renderButton(props) {
  var type = props.type,
      children = props.children,
      size = props.size,
      width = props.width,
      _onClick2 = props.onClick,
      dataLoc = props.dataLoc;

  if (type === 'primary') {
    return /*#__PURE__*/_react["default"].createElement(_Button["default"], (0, _extends2["default"])({}, props, {
      primary: true,
      style: {
        maxWidth: '100%'
      },
      ariaLabel: children,
      children: children,
      size: size,
      width: size !== 'small' && '100%',
      onClick: function onClick(e) {
        return _onClick(e, _onClick2);
      },
      dataLoc: dataLoc
    }));
  } else if (type === 'secondary') {
    return /*#__PURE__*/_react["default"].createElement(_Button["default"], (0, _extends2["default"])({}, props, {
      secondary: true,
      style: {
        maxWidth: '100%'
      },
      ariaLabel: children,
      children: children,
      size: size,
      width: size !== 'small' && '100%',
      onClick: function onClick(e) {
        return _onClick(e, _onClick2);
      },
      dataLoc: dataLoc
    }));
  }
}

var propTypes = {
  children: _propTypes["default"].string,
  rowQuantity: _propTypes["default"].number,
  size: _propTypes["default"].oneOf(['large', 'small']),
  type: _propTypes["default"].oneOf(['primary', 'secondary', 'textlink', 'textlinkcaret']),
  width: _propTypes["default"].string,
  index: _propTypes["default"].number,
  noVerticalPadding: _propTypes["default"].bool,
  smallButtonSpacing: _propTypes["default"].string,
  renderButton: _propTypes["default"].func
};
var defaultProps = {
  rowQuantity: 3,
  size: 'large',
  type: 'primary',
  noVerticalPadding: false,
  renderButton: _renderButton
};
var BUTTON_GAP_LG = '3X';
var BUTTON_GAP_SM = '2X';
var TEXTLINK_HORIZONTAL_GAP_LG = 24;
var TEXTLINK_HORIZONTAL_GAP_SM = 16;

function _calculateSpacing(size, type, rowQuantity, index) {
  if (rowQuantity === 1) {
    return 0;
  }

  if (size === 'large' && (type === 'primary' || type === 'secondary')) {
    return (0, _theme.calculateSpacing)(BUTTON_GAP_LG, 'px');
  } else if (size === 'large') {
    return (0, _javascriptStyleHelpers.calculateRem)(TEXTLINK_HORIZONTAL_GAP_LG);
  }

  if (size === 'small' && (type === 'primary' || type === 'secondary')) {
    return (0, _theme.calculateSpacing)(BUTTON_GAP_SM, 'px');
  } else if (size === 'large') {
    return (0, _javascriptStyleHelpers.calculateRem)(TEXTLINK_HORIZONTAL_GAP_SM);
  }
}

var createPaddingStyles = function createPaddingStyles(props) {
  var size = props.size,
      type = props.type,
      firstChild = props.firstChild,
      rowQuantity = props.rowQuantity,
      index = props.index;
  var paddingBottom;
  var paddingRight;

  switch (size) {
    case 'large':
      paddingBottom = (0, _theme.calculateSpacing)(BUTTON_GAP_LG, 'px');
      paddingRight = _calculateSpacing('large', type, rowQuantity, index);
      break;

    case 'small':
      paddingBottom = (0, _theme.calculateSpacing)(props.smallButtonSpacing, 'px');
      paddingRight = _calculateSpacing('small', type, rowQuantity, index);
      break;

    default:
      paddingBottom = 0;
      paddingRight = 0;
      break;
  }

  var paddingStyles = "\n    padding-bottom: ".concat(paddingBottom, ";\n    padding-right: ").concat(paddingRight, ";\n  ");
  return paddingStyles;
};

var ButtonWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "ButtonWrapper",
  componentId: "mtl1r3-0"
})(_templateObject(), function (props) {
  return createPaddingStyles(props);
}, function (props) {
  return props.size === 'small' ? 'auto' : props.flexBasis;
}, function (_ref) {
  var noVerticalPadding = _ref.noVerticalPadding;
  return noVerticalPadding && "\n    padding-bottom: 0px;\n  ";
}, function (_ref2) {
  var rowQuantity = _ref2.rowQuantity;
  return rowQuantity === 1 && 0;
});

var ButtonGroupItem = function ButtonGroupItem(props) {
  var children = props.children,
      rowQuantityProp = props.rowQuantity,
      size = props.size,
      type = props.type,
      width = props.width,
      onClick = props.onClick,
      index = props.index,
      noVerticalPadding = props.noVerticalPadding,
      smallButtonSpacing = props.smallButtonSpacing,
      renderButton = props.renderButton;
  var buttonsPerRow = rowQuantityProp;
  var flexBasis = buttonsPerRow ? 100 / buttonsPerRow + '%' : '100%';
  return /*#__PURE__*/_react["default"].createElement(ButtonWrapper, {
    index: index,
    rowQuantity: buttonsPerRow,
    flexBasis: flexBasis,
    size: size,
    type: type,
    width: width,
    noVerticalPadding: noVerticalPadding,
    smallButtonSpacing: smallButtonSpacing
  }, renderButton(props), type === 'textlink' && /*#__PURE__*/_react["default"].createElement(_TextLink["default"], (0, _extends2["default"])({}, props, {
    style: {
      maxWidth: '100%'
    },
    ariaLabel: children,
    children: children,
    size: size,
    type: "standAlone",
    width: width,
    onClick: _onClick
  })), type === 'textlinkcaret' && /*#__PURE__*/_react["default"].createElement(_TextLinkCaret["default"], (0, _extends2["default"])({}, props, {
    style: {
      maxWidth: '100%'
    },
    ariaLabel: children,
    children: children,
    iconPosition: "right",
    size: size,
    width: width,
    onClick: _onClick
  })));
};

ButtonGroupItem.propTypes = propTypes;
ButtonGroupItem.defaultProps = defaultProps;
var _default = ButtonGroupItem;
exports["default"] = _default;
