// @vds-core/buttons/dist/cjs/components/ButtonGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var React = _interopRequireWildcard(require("react"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ButtonGroupItem = _interopRequireDefault(require("./ButtonGroupItem"));

var _utilities = require("@vds-core/utilities");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-wrap: wrap;\n  flex: 1;\n  max-width: ", ";\n  width: 100%;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  /**
   * If provided, Button components will be rendered based on this data
   */
  data: _propTypes["default"].array.isRequired,

  /**
   * If provided, width of Button components will be rendered based on this value. If omitted, default button widths are rendered.
   */
  childWidth: _propTypes["default"].string,

  /**
   * An object containing number of Button components per row, in each viewport
   */
  rowQuantity: _propTypes["default"].object,

  /**
   * Viewport the Buttons will be rendered in
   */
  viewport: _propTypes["default"].oneOf(['desktop', 'tablet', 'mobile']),

  /**
   * @ignore
   * No vertical padding
   */
  noVerticalPadding: _propTypes["default"].bool,

  /**
  * @ignore 
   If provided, the provided string will determine the spacing between small buttons
   */
  smallButtonSpacing: _propTypes["default"].string,

  /**
  * @ignore
   If provided a max width of the provided string will be rendered.
  */
  maxWidth: _propTypes["default"].string,

  /**
  * @ignore
   If provided a new Button component will render
  */
  renderButton: _propTypes["default"].func,

  /**
   * @ignore
   */
  dataLoc: _propTypes["default"].string
};
var defaultProps = {
  childWidth: '100%',
  viewport: 'desktop',
  noVerticalPadding: false,
  maxWidth: undefined,
  smallButtonSpacing: '2X'
};

var ButtonGroupWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "ButtonGroupWrapper",
  componentId: "lt36ab-0"
})(_templateObject(), function (_ref) {
  var width = _ref.width,
      rowQuantity = _ref.rowQuantity;
  return rowQuantity !== 1 && width;
});

function _calculateWidth(childWidth, maxWidth) {
  var width;
  var maximum = parseInt(maxWidth);

  if (childWidth[childWidth.length - 1] === '%') {
    width = parseInt(childWidth) / 100;
    width = (0, _javascriptStyleHelpers.calculateRem)(596 * width);
  } else if (childWidth && parseInt(childWidth) < maximum) {
    width = (0, _javascriptStyleHelpers.calculateRem)(parseInt(childWidth));
  } else {
    width = (0, _javascriptStyleHelpers.calculateRem)(maximum);
  }

  return width;
}

var ButtonGroup = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(ButtonGroup, _Component);

  var _super = _createSuper(ButtonGroup);

  function ButtonGroup() {
    var _this;

    (0, _classCallCheck2["default"])(this, ButtonGroup);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_onClick", function (e, childClick) {
      childClick && childClick(e);
      _this.props.onClick && _this.props.onClick(e);
    });
    return _this;
  }

  (0, _createClass2["default"])(ButtonGroup, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          data = _this$props.data,
          childWidth = _this$props.childWidth,
          rowQuantityProp = _this$props.rowQuantity,
          viewport = _this$props.viewport,
          noVerticalPadding = _this$props.noVerticalPadding,
          smallButtonSpacing = _this$props.smallButtonSpacing,
          maxWidth = _this$props.maxWidth,
          renderButton = _this$props.renderButton,
          dataLoc = _this$props.dataLoc;
      var rowQuantityVal = rowQuantityProp ? rowQuantityProp[viewport] : data.length;
      return /*#__PURE__*/React.createElement(ButtonGroupWrapper, {
        childWidth: childWidth,
        viewport: viewport,
        maxWidth: maxWidth,
        width: _calculateWidth(childWidth, maxWidth),
        rowQuantity: rowQuantityVal
      }, data.map(function (child, index) {
        var children = child.children,
            size = child.size,
            type = child.type,
            width = child.width,
            noVerticalPadding = child.noVerticalPadding;
        var btnWidth = childWidth;
        btnWidth ? btnWidth : btnWidth = width;
        return /*#__PURE__*/React.createElement(React.Fragment, {
          key: index
        }, /*#__PURE__*/React.createElement(_ButtonGroupItem["default"], (0, _extends2["default"])({}, child, {
          children: children,
          rowQuantity: rowQuantityVal,
          size: size,
          type: type,
          width: btnWidth,
          onClick: function onClick(e) {
            return _this2._onClick(e, child.onClick);
          },
          index: index,
          noVerticalPadding: noVerticalPadding,
          smallButtonSpacing: smallButtonSpacing,
          renderButton: renderButton,
          dataLoc: dataLoc
        })));
      }));
    }
  }]);
  return ButtonGroup;
}(React.Component);

ButtonGroup.propTypes = propTypes;
ButtonGroup.defaultProps = defaultProps;
/** @component */

var _default = (0, _utilities.withVDSManager)(ButtonGroup);

exports["default"] = _default;
