// @vds-core/buttons/dist/cjs/components/Button.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _typography = require("@vds-core/typography");

var _theme = require("@vds-core/theme");

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  ", ";\n  width: ", ";\n  justify-content: center;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  ", ";\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 0;\n  border-radius: ", ";\n  border-width: ", ";\n  box-sizing: border-box;\n  cursor: pointer;\n  display: ", ";\n  flex-direction: row;\n  align-items: center;\n  justify-content: center;\n  font-family: ", ";\n  font-size: ", ";\n  font-weight: ", ";\n  line-height: ", ";\n  letter-spacing: ", ";\n  height: ", ";\n  position: relative;\n  text-align: center;\n  text-decoration: none;\n  touch-action: manipulation;\n  vertical-align: middle;\n  width: auto;\n  overflow: visible;\n\n  ", ";\n\n  background-color: ", ";\n  border: ", ";\n  color: ", ";\n  ", ";\n  ", ";\n  ", ";\n\n  ", ";\n\n  ", ";\n\n  ", ";\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: relative;\n  height: 100%;\n  width: 100%;\n  border-radius: ", ";\n  box-sizing: border-box;\n  display: flex;\n  align-items: center;\n  background: transparent;\n  justify-content: center;\n  text-overflow: ", ";\n  &:focus {\n    outline: none;\n  }\n  ", ";\n  &:hover {\n    outline: none;\n    border: none;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  height: ", ";\n  width: 100%;\n  left: 50%;\n  position: absolute;\n  transform: translate(-50%, -50%);\n  text-align: center;\n  top: 50%;\n  content: '';\n  display: inline-block;\n  &:hover {\n    cursor: pointer;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _calculateButtonHeight(size) {
  var height;

  if (size === 'small') {
    height = (0, _javascriptStyleHelpers.calculateRem)(24);
  } else {
    height = (0, _javascriptStyleHelpers.calculateRem)(42);
  }

  return height;
}

function _calculateSideMargin(size, width) {
  var sideMargin;

  if (size === 'small') {
    sideMargin = width === 'autoTight' ? 12 : 24;
  } else {
    sideMargin = width === 'autoTight' ? 21 : 42;
  }

  return (0, _javascriptStyleHelpers.calculateRem)(sideMargin);
}

var fontWeightStrong = 750;
var propTypes = {
  /**
   * Decides if the button is rendered as primary (will default as primary).
   */
  primary: _propTypes["default"].bool,

  /**
   * Decides if the button is rendered as secondary.
   */
  secondary: _propTypes["default"].bool,

  /** String value for inner button text. */
  children: _propTypes["default"].oneOfType([_propTypes["default"].node, _propTypes["default"].string]),

  /** Button will be disabled if disabled prop is passed. */
  disabled: _propTypes["default"].bool,

  /** Function for handling click events */
  onClick: _propTypes["default"].func,

  /** The size of the button. */
  size: _propTypes["default"].oneOf(['normal', 'tiny', 'small', 'large', 'block']),

  /** does the button have a fixed or variable width */
  width: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),

  /** DEPRECATED SIZES: normal, tiny, block **/
  display: _propTypes["default"].oneOf(['flex', 'inline-block', 'block']),
  ariaLabel: _propTypes["default"].string,

  /**
   * @ignore
  Object containg font styles, including font-size, line-height, and font-family */
  bodyConfig: _propTypes["default"].object,

  /**
   * @ignore
  Typescale of button text */
  typescale: _propTypes["default"].string,

  /**
   * @ignore
  Object containing colors in hex */
  colorConfig: _propTypes["default"].object,

  /**
   * @ignore
  Boolean that turns on new focus state */
  focusState: _propTypes["default"].bool,

  /**
   * @ignore
  Boolean that turns on new hover state */
  hoverState: _propTypes["default"].bool,

  /**
   * @ignore
  Boolean that turns on new active state */
  activeState: _propTypes["default"].bool,

  /**
   * @ignore
  Boolean that activates hit area */
  hitArea: _propTypes["default"].bool,

  /**
   * @ignore
  Prop that determines Button height */
  calculateButtonHeight: _propTypes["default"].func,

  /**
   * @ignore
  Prop that determines the margin size on each side of text */
  calculateSideMargin: _propTypes["default"].func,

  /**
   * @ignore
  Boolean that turns on/off vertical padding on small button text */
  smallButtonPadding: _propTypes["default"].bool,

  /**
   * @ignore
  If provided, text will overflow with ellipsis */
  overflowEllipsis: _propTypes["default"].bool,

  /**
   * @ignore
  If provided, the number given will determine the font weight */
  calculateFontWeight: _propTypes["default"].number,

  /**
   * Allows a string to be provided for analytics.
   */
  'data-analyticstrack': _propTypes["default"].string,

  /**
   * Allows a string to be provided for analytics.
   */
  'data-track': _propTypes["default"].string,

  /**
   * Allows a string to be provided for click stream.
   */
  'data-clickstream': _propTypes["default"].string,

  /**
   * @ignore
   */
  dataLoc: _propTypes["default"].string,

  /**
   * If provided, button primitive becomes an anchor, with the given href attribute.
   */
  href: _propTypes["default"].string
};
var defaultProps = {
  primary: false,
  secondary: false,
  children: null,
  disabled: false,
  size: 'large',
  display: 'flex',
  width: 'auto',
  onClick: undefined,
  bodyConfig: _typography.TitleConfig,
  colorConfig: _theme.colors,
  hitArea: false,
  calculateButtonHeight: _calculateButtonHeight,
  calculateSideMargin: _calculateSideMargin,
  hoverState: false,
  smallButtonPadding: true,
  overflowEllipsis: false,
  calculateFontWeight: fontWeightStrong,
  typescale: _typography.TypographyConfig.getTypescale()
};
/**
 * @component
 * */

/*
fontWeightStrong represents the font weight found in 
typography used when the component is bolded. Font weight is
not exported by typography configs due to the fact that 
typography components may have multiple font weight options
*/

var HitArea = /*#__PURE__*/_styledComponents["default"].span.withConfig({
  displayName: "HitArea",
  componentId: "bc3yhn-0"
})(_templateObject(), (0, _javascriptStyleHelpers.calculateRem)(44));

var StyledChildWrapper = /*#__PURE__*/_styledComponents["default"].span.withConfig({
  displayName: "StyledChildWrapper",
  componentId: "bc3yhn-1"
})(_templateObject2(), (0, _javascriptStyleHelpers.calculateRem)(100), function (_ref) {
  var overflowEllipsis = _ref.overflowEllipsis;
  return overflowEllipsis && 'ellipsis';
}, function (_ref2) {
  var size = _ref2.size,
      width = _ref2.width,
      calculateSideMargin = _ref2.calculateSideMargin,
      smallButtonPadding = _ref2.smallButtonPadding;
  if (size === 'block') return "margin: auto";

  switch (size) {
    case 'tiny':
      return "\n          margin: ".concat((0, _javascriptStyleHelpers.calculateRem)(0, 13), ";\n        ");

    case 'small':
      return "\n          padding-right: ".concat(calculateSideMargin(size, width), ";\n          padding-left: ").concat(calculateSideMargin(size, width), ";\n          padding: ").concat(function (_ref3) {
        var smallButtonPadding = _ref3.smallButtonPadding;
        return smallButtonPadding ? (0, _javascriptStyleHelpers.calculateRem)(6.5, 0) : 0;
      }, ";\n        ");

    case 'block':
      return "\n          display: block;\n          width: 100%;\n          padding-right: ".concat((0, _javascriptStyleHelpers.calculateRem)(width === 'autoTight' ? 21 : 42), ";\n          padding-left: ").concat((0, _javascriptStyleHelpers.calculateRem)(width === 'autoTight' ? 21 : 42), ";\n        ");

    default:
      return "\n          padding-right: ".concat(calculateSideMargin(size, width), ";\n          padding-left: ").concat(calculateSideMargin(size, width), ";");
  }
});

var findWidth = function findWidth(width) {
  if (typeof width === 'string') return width;
  return (0, _javascriptStyleHelpers.calculateRem)(width);
};

var buttonStyles = /*#__PURE__*/(0, _styledComponents.css)(_templateObject3(), (0, _javascriptStyleHelpers.calculateRem)(100), (0, _javascriptStyleHelpers.calculateRem)(1), function (_ref4) {
  var display = _ref4.display;
  return display;
}, function (_ref5) {
  var largeButtonText = _ref5.largeButtonText;
  return largeButtonText.fontFamily;
}, function (_ref6) {
  var largeButtonText = _ref6.largeButtonText;
  return (0, _javascriptStyleHelpers.calculateRem)(largeButtonText.fontSize);
}, function (_ref7) {
  var calculateFontWeight = _ref7.calculateFontWeight;
  return calculateFontWeight;
}, function (_ref8) {
  var largeButtonText = _ref8.largeButtonText;
  return (0, _javascriptStyleHelpers.calculateRem)(largeButtonText.lineHeight);
}, function (_ref9) {
  var largeButtonText = _ref9.largeButtonText,
      size = _ref9.size;
  return size === 'large' && (0, _javascriptStyleHelpers.calculateRem)(largeButtonText.letterSpacing);
}, function (_ref10) {
  var calculateButtonHeight = _ref10.calculateButtonHeight,
      size = _ref10.size;
  return calculateButtonHeight(size);
}, function (_ref11) {
  var size = _ref11.size,
      width = _ref11.width,
      smallButtonText = _ref11.smallButtonText;

  if (size === 'tiny') {
    return "\n        font-size: ".concat((0, _javascriptStyleHelpers.calculateRem)(smallButtonText.fontSize), ";\n        line-height: ").concat((0, _javascriptStyleHelpers.calculateRem)(smallButtonText.lineHeight), ";\n        font-family: ").concat(smallButtonText.fontFamily, ";\n      ");
  }

  if (size === 'small') {
    return "\n        font-size: ".concat((0, _javascriptStyleHelpers.calculateRem)(smallButtonText.fontSize), ";\n        line-height: ").concat((0, _javascriptStyleHelpers.calculateRem)(smallButtonText.lineHeight), ";\n        font-family: ").concat(smallButtonText.fontFamily, ";\n        width: ").concat(width ? findWidth(width) : 'auto', ";\n        white-space: nowrap;\n      ");
  }

  if (size === 'large') {
    return "\n        width: ".concat(width ? findWidth(width) : 'auto', ";\n        white-space: nowrap;\n      ");
  }

  if (size === 'block') {
    return "\n        display: block;\n        width: 100%;\n      ";
  }
}, function (_ref12) {
  var colorConfig = _ref12.colorConfig,
      secondary = _ref12.secondary;
  return secondary ? 'transparent' : colorConfig.black;
}, function (_ref13) {
  var colorConfig = _ref13.colorConfig;
  return "".concat((0, _javascriptStyleHelpers.calculateRem)(1), " solid ").concat(colorConfig.black);
}, function (_ref14) {
  var colorConfig = _ref14.colorConfig,
      secondary = _ref14.secondary;
  return secondary ? colorConfig.black : colorConfig.white;
}, function (_ref15) {
  var activeState = _ref15.activeState,
      disabled = _ref15.disabled,
      colorConfig = _ref15.colorConfig;
  return activeState && !disabled && "\n      &:active {\n        border-color: ".concat(colorConfig.coolGray6, ";\n        box-shadow: 0 0 0 ").concat((0, _javascriptStyleHelpers.calculateRem)(1), " ").concat(colorConfig.coolGray6, ";\n        background-color: ").concat(colorConfig.coolGray6, ";\n      }\n    ");
}, function (_ref16) {
  var focusState = _ref16.focusState,
      disabled = _ref16.disabled,
      inverted = _ref16.inverted,
      colorConfig = _ref16.colorConfig;
  return focusState && !disabled && "\n     &:focus:not(:hover) {\n      outline: none;\n      &::before {\n        border: ".concat((0, _javascriptStyleHelpers.calculateRem)(1), " dashed ").concat(inverted ? colorConfig.white : colorConfig.black, ";\n        border-radius: ").concat((0, _javascriptStyleHelpers.calculateRem)(100), ";\n        content: '';\n        height: calc(100% + ").concat((0, _javascriptStyleHelpers.calculateRem)(6), ");\n        left: 50%;\n        position: absolute;\n        top: 50%;\n        transform: translate(-50%, -50%);\n        width: calc(100% + ").concat((0, _javascriptStyleHelpers.calculateRem)(6), ");\n      }\n    }\n    ");
}, function (_ref17) {
  var hoverState = _ref17.hoverState,
      colorConfig = _ref17.colorConfig,
      inverted = _ref17.inverted,
      disabled = _ref17.disabled;
  return hoverState && !disabled && "\n    &:hover:not(:active) {\n      outline: none;\n      box-shadow: 0 0 0 ".concat((0, _javascriptStyleHelpers.calculateRem)(1), " ").concat(inverted ? colorConfig.white : colorConfig.black, ";\n      transition: ease-out 0.1s;\n    }\n    &:hover:not(:active) span {\n      border: none;\n    }\n  ");
}, function (_ref18) {
  var inverted = _ref18.inverted,
      disabled = _ref18.disabled,
      activeState = _ref18.activeState,
      colorConfig = _ref18.colorConfig,
      secondary = _ref18.secondary;
  return inverted && !disabled && "\n    background-color: ".concat(secondary ? 'transparent' : colorConfig.white, ";\n    border: ").concat((0, _javascriptStyleHelpers.calculateRem)(1), " solid ").concat(colorConfig.white, ";\n    color: ").concat(secondary ? colorConfig.white : colorConfig.black, ";\n\n  ");
}, function (_ref19) {
  var disabled = _ref19.disabled,
      colorConfig = _ref19.colorConfig,
      activeState = _ref19.activeState,
      secondary = _ref19.secondary;
  return disabled && "\n    background-color: ".concat(secondary ? 'transparent' : colorConfig.coolGray3, ";\n    border-color: ").concat(colorConfig.coolGray3, ";\n    color: ").concat(secondary ? colorConfig.coolGray3 : colorConfig.white, "\n    cursor: default;\n  ");
}, function (_ref20) {
  var disabled = _ref20.disabled,
      inverted = _ref20.inverted,
      colorConfig = _ref20.colorConfig,
      activeState = _ref20.activeState,
      secondary = _ref20.secondary;
  return disabled && inverted && "\n    background-color: ".concat(secondary ? 'transparent' : colorConfig.coolGray10, ";\n    border-color: ").concat(colorConfig.coolGray10, ";\n    color: ").concat(secondary ? colorConfig.coolGray10 : colorConfig.black, "\n  ");
});

var StyledButton = /*#__PURE__*/_styledComponents["default"].button.withConfig({
  displayName: "StyledButton",
  componentId: "bc3yhn-2"
})(_templateObject4(), buttonStyles);

var StyledAnchor = /*#__PURE__*/_styledComponents["default"].a.withConfig({
  displayName: "StyledAnchor",
  componentId: "bc3yhn-3"
})(_templateObject5(), buttonStyles, function (_ref21) {
  var width = _ref21.width,
      size = _ref21.size;
  return width !== 'auto' && width !== 'autoTight' && size === 'large' ? findWidth(width) : 'fit-content';
});
/**
 * @ignore
 */


var Button = function Button(props) {
  var children = props.children,
      size = props.size,
      width = props.width,
      disabled = props.disabled,
      ariaLabel = props.ariaLabel,
      onClick = props.onClick,
      display = props.display,
      colorConfig = props.colorConfig,
      bodyConfig = props.bodyConfig,
      hitArea = props.hitArea,
      typescale = props.typescale,
      calculateButtonHeight = props.calculateButtonHeight,
      calculateSideMargin = props.calculateSideMargin,
      smallButtonPadding = props.smallButtonPadding,
      calculateFontWeight = props.calculateFontWeight,
      overflowEllipsis = props.overflowEllipsis,
      dataLoc = props.dataLoc,
      secondary = props.secondary,
      href = props.href;
  var largeButtonText, smallButtonText;
  largeButtonText = bodyConfig.getStyles(typescale, 'desktop', typescale === 'VDS' ? 'large' : 'extraSmall');
  smallButtonText = bodyConfig.getStyles(typescale, 'mobile', typescale === 'VDS' ? 'small' : 'extraSmall');
  var newAriaLabel = ariaLabel ? ariaLabel : typeof children === 'string' ? children : 'Button';
  var ButtonBase = href ? StyledAnchor : StyledButton;
  return /*#__PURE__*/_react["default"].createElement(ButtonBase, (0, _extends2["default"])({}, props, {
    href: href,
    size: size,
    width: width,
    "aria-label": newAriaLabel,
    role: "button",
    "aria-disabled": props.disabled,
    tabIndex: 0,
    display: display,
    largeButtonText: largeButtonText,
    smallButtonText: smallButtonText,
    colorConfig: colorConfig,
    disabled: disabled,
    calculateButtonHeight: calculateButtonHeight,
    calculateSideMargin: calculateSideMargin,
    calculateFontWeight: calculateFontWeight,
    "data-loc": dataLoc,
    secondary: secondary
  }), hitArea && !disabled && /*#__PURE__*/_react["default"].createElement(HitArea, null), /*#__PURE__*/_react["default"].createElement(StyledChildWrapper, {
    tabIndex: -1,
    display: display,
    size: size,
    width: width,
    overflowEllipsis: overflowEllipsis,
    calculateSideMargin: calculateSideMargin,
    smallButtonPadding: smallButtonPadding
  }, children));
};

Button.defaultProps = defaultProps;
Button.propTypes = propTypes;
/** @component */

var _default = Button;
exports["default"] = _default;
