// @vds-core/buttons/dist/cjs/components/ButtonSecondary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _ButtonBase = _interopRequireDefault(require("./ButtonBase"));

var _theme = require("@vds-core/theme");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  background-color: transparent;\n  border: ", ";\n  span {\n    border: none;\n    background: transparent;\n  }\n  color: ", ";\n\n  ", ";\n  ", ";\n  ", ";\n  &.link:hover {\n    text-decoration: none;\n  }\n  ", ";\n\n  ", ";\n\n  ", ";\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  /** String value for inner button text. */
  children: _propTypes["default"].node,

  /** Button will be disabled if disabled prop is passed. */
  disabled: _propTypes["default"].bool,

  /** Button text color and background will be swapped. */
  inverted: _propTypes["default"].bool,

  /** Function to call if Button is clicked. */
  onClick: _propTypes["default"].func.isRequired,

  /** The size of the button. */
  size: _propTypes["default"].oneOf(['normal', 'tiny', 'small', 'large', 'block']),

  /**
   * Allows a string to be applied at data-analyticstrack
   */
  analyticsTrack: _propTypes["default"].string,

  /**
   * Allows a string to be applied at data-track
   */
  track: _propTypes["default"].string,

  /**
   * Allows a string to be applied at data-clickstream
   */
  clickStream: _propTypes["default"].string,
  width: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].number]),

  /**
   * @ignore
  Object containg font styles, including font-size, line-height, and font-family */
  bodyConfig: _propTypes["default"].object,

  /**
   * @ignore
  Typescale of button text */
  typescale: _propTypes["default"].string,

  /**
   * @ignore
  Object containing colors in hex */
  colorConfig: _propTypes["default"].object,

  /**
   * @ignore
  Boolean that turns on new focus state */
  focusState: _propTypes["default"].bool,

  /**
   * @ignore
  Boolean that turns on new hover state */
  hoverState: _propTypes["default"].bool,

  /**
   * @ignore
  Boolean that turns on new active state */
  activeState: _propTypes["default"].bool,

  /**
   * @ignore
   */
  dataLoc: _propTypes["default"].string
};
var defaultProps = {
  children: null,
  disabled: false,
  inverted: false,
  size: 'large',
  onClick: function onClick() {},
  colorConfig: _theme.colors,
  focusState: false,
  hoverState: false,
  activeState: false
};
/**
 * @ignore
 */

var ButtonSecondary = /*#__PURE__*/(0, _styledComponents["default"])(_ButtonBase["default"]).withConfig({
  displayName: "ButtonSecondary",
  componentId: "n2w9t7-0"
})(_templateObject(), function (_ref) {
  var colorConfig = _ref.colorConfig;
  return "".concat((0, _javascriptStyleHelpers.calculateRem)(1), " solid ").concat(colorConfig.black);
}, function (_ref2) {
  var colorConfig = _ref2.colorConfig;
  return colorConfig.black;
}, function (_ref3) {
  var activeState = _ref3.activeState,
      disabled = _ref3.disabled,
      colorConfig = _ref3.colorConfig;
  return activeState && !disabled && "\n      &:active {\n        border-color: ".concat(colorConfig.coolGray6, ";\n        box-shadow: 0 0 0 ").concat((0, _javascriptStyleHelpers.calculateRem)(1), " ").concat(colorConfig.coolGray6, ";\n        color: ").concat(colorConfig.coolGray6, ";\n      }\n    ");
}, function (_ref4) {
  var focusState = _ref4.focusState,
      disabled = _ref4.disabled,
      inverted = _ref4.inverted,
      colorConfig = _ref4.colorConfig;
  return focusState && !disabled && "\n    &::-moz-focus-inner {\n      border: 0;\n    }\n    &:focus:not(:hover):not(:active) {\n      outline: none;\n      &::before {\n        border: ".concat((0, _javascriptStyleHelpers.calculateRem)(1), " dashed ").concat(inverted ? colorConfig.white : colorConfig.black, ";\n        border-radius: ").concat((0, _javascriptStyleHelpers.calculateRem)(100), ";\n        content: '';\n        height: calc(100% + ").concat((0, _javascriptStyleHelpers.calculateRem)(6), ");\n        left: 50%;\n        position: absolute;\n        top: 50%;\n        transform: translate(-50%, -50%);\n        width: calc(100% + ").concat((0, _javascriptStyleHelpers.calculateRem)(6), ");\n      }\n    }\n    ");
}, function (_ref5) {
  var hoverState = _ref5.hoverState,
      colorConfig = _ref5.colorConfig,
      inverted = _ref5.inverted,
      disabled = _ref5.disabled;
  return hoverState && !disabled && "\n      &:hover:not(:active) {\n        outline: none;\n        box-shadow: 0 0 0 ".concat((0, _javascriptStyleHelpers.calculateRem)(1), " ").concat(inverted ? colorConfig.white : colorConfig.black, ";\n        transition: ease-out 0.1s;\n      }\n      &:hover:not(:active) span {\n        border: none;\n      }\n  ");
}, function (_ref6) {
  var inverted = _ref6.inverted,
      colorConfig = _ref6.colorConfig;
  return inverted && "\n    border-color: ".concat(colorConfig.white, ";\n    color: ").concat(colorConfig.white, ";\n  ");
}, function (_ref7) {
  var disabled = _ref7.disabled,
      colorConfig = _ref7.colorConfig;
  return disabled && "\n    background-color: transparent;\n    border-color: ".concat(colorConfig.coolGray3, ";\n    color: ").concat(colorConfig.coolGray3, ";\n    cursor: default;\n  ");
}, function (_ref8) {
  var disabled = _ref8.disabled,
      inverted = _ref8.inverted,
      colorConfig = _ref8.colorConfig;
  return disabled && inverted && "\n    border-color: ".concat(colorConfig.coolGray10, ";\n    color: ").concat(colorConfig.coolGray10, ";\n  ");
});
ButtonSecondary.defaultProps = defaultProps;
ButtonSecondary.propTypes = propTypes;
var _default = ButtonSecondary;
exports["default"] = _default;
