// @vds-core/carousels/src/util/AspectConfig.js
function getRatio(type, viewport, numItems) {
  return CarouselConfig[type][viewport][numItems];
}

const largePadding = {};

const CarouselConfig = {
  small: {
    mobile: {
      2: {
        aspectRatio: '56.25%',
        topBotPadding: 12,
        titlePadding: 12,
        barsPadding: 4,
        sidePadding: 12,
        iconSize: 'medium',
        minWidth: 272,
      },
      3: {
        aspectRatio: '56.25%',
        topBotPadding: 12,
        titlePadding: 12,
        barsPadding: 4,
        sidePadding: 12,
        iconSize: 'medium',
        minWidth: 272,
      },
      4: {
        aspectRatio: '56.25%',
        topBotPadding: 12,
        titlePadding: 12,
        barsPadding: 4,
        sidePadding: 12,
        iconSize: 'medium',
        minWidth: 272,
      },
    },
    largeMobile: {
      2: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      3: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    smallTablet: {
      2: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      3: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    tablet: {
      2: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      3: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    smallDesktop: {
      2: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      3: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    mediumDesktop: {
      2: {
        aspectRatio: '37.5%',
        topBotPadding: 20,
        titlePadding: 20,
        barsPadding: 4,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      3: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '37.5%',
        topBotPadding: 16,
        titlePadding: 16,
        barsPadding: 4,
        sidePadding: 16,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    desktop: {
      2: {
        aspectRatio: '37.5%',
        topBotPadding: 20,
        titlePadding: 20,
        barsPadding: 4,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      3: {
        aspectRatio: '37.5%',
        topBotPadding: 12,
        titlePadding: 12,
        barsPadding: 4,
        sidePadding: 12,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '37.5%',
        topBotPadding: 12,
        titlePadding: 12,
        barsPadding: 4,
        sidePadding: 12,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
  },
  medium: {
    mobile: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        titlePadding: 10,
        linePadding: 10,
        barsPadding: 20,
        barsPaddingChevron: 8,
        sidePadding: 16,
        iconSize: 'medium',
        minWidth: 272,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        titlePadding: 10,
        linePadding: 10,
        barsPadding: 20,
        barsPaddingChevron: 8,
        sidePadding: 16,
        iconSize: 'medium',
        minWidth: 272,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        titlePadding: 10,
        linePadding: 10,
        barsPadding: 20,
        barsPaddingChevron: 8,
        sidePadding: 16,
        iconSize: 'medium',
        minWidth: 272,
      },
    },
    largeMobile: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        titlePadding: 10,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 16,
        iconSize: 'medium',
        minWidth: 384,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    smallTablet: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        titlePadding: 10,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 16,
        iconSize: 'medium',
        minWidth: 384,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    tablet: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        titlePadding: 10,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 16,
        iconSize: 'medium',
        minWidth: 384,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    smallDesktop: {
      2: {
        aspectRatio: '50%',
        topBotPadding: 12.5,
        titlePadding: 10,
        linePadding: 10,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 12,
        iconSize: 'medium',
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    mediumDesktop: {
      2: {
        aspectRatio: '50%',
        topBotPadding: 12.5,
        titlePadding: 10,
        linePadding: 10,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 12,
        iconSize: 'medium',
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
    desktop: {
      2: {
        aspectRatio: '50%',
        topBotPadding: 20.5,
        titlePadding: 10,
        linePadding: 12,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'medium',
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 20,
        barsPaddingChevron: 20,
        sidePadding: 20,
        iconSize: 'extraLarge',
        minWidth: 424,
      },
    },
  },
  large: {
    mobile: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        eyebrowPadding: 10,
        titlePadding: 12,
        linePadding: 12,
        barsPadding: 26,
        sidePadding: 16,
        minWidth: 320,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        eyebrowPadding: 10,
        titlePadding: 12,
        linePadding: 12,
        barsPadding: 26,
        sidePadding: 16,
        minWidth: 320,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16.5,
        eyebrowPadding: 10,
        titlePadding: 12,
        linePadding: 12,
        barsPadding: 26,
        sidePadding: 16,
        minWidth: 320,
      },
    },
    largeMobile: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
    },
    smallTablet: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
    },
    tablet: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
    },
    smallDesktop: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
    },
    mediumDesktop: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
    },
    desktop: {
      2: {
        aspectRatio: '100%',
        topBotPadding: 20.5,
        eyebrowPadding: 16,
        titlePadding: 20,
        linePadding: 20,
        barsPadding: 28,
        sidePadding: 20,
        minWidth: 495.5,
      },
      3: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
      4: {
        aspectRatio: '100%',
        topBotPadding: 16,
        eyebrowPadding: 12,
        titlePadding: 16,
        linePadding: 16,
        barsPadding: 28,
        sidePadding: 16,
        minWidth: 495.5,
      },
    },
  },

  getRatio: getRatio,
};

export default CarouselConfig;
