// @vds-core/carousels/src/util/FixedConfig.js
function getFixed(type, viewport, numItems) {
  return FixedConfig[type][viewport][numItems];
}

const FixedConfig = {
  medium: {
    mobile: {
      2: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
      },
      3: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
      },
      4: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
      },
    },
    largeMobile: {
      2: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
      },
      3: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
      },
      4: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
      },
    },
    smallTablet: {
      2: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      3: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      4: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
    },
    tablet: {
      2: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      3: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      4: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
    },
    smallDesktop: {
      2: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      3: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      4: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
    },
    mediumDesktop: {
      2: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      3: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      4: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'small',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
    },
    desktop: {
      2: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'extraLarge',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      3: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'extraLarge',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
      4: {
        sidePadding: 16,
        topBotPadding: 16.5,
        linePadding: 12,
        iconSize: 'extraLarge',
        visibleSlides: 1,
        optionTwoVisible: 2,
      },
    },
  },
  large: {
    mobile: {
      2: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 1,
      },
      3: {
        linePadding: 12,
        eyebrowPadding: 12,
        topBotPadding: 12,
        sidePadding: 12,
        titlePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 1,
      },
      4: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 1,
      },
    },
    largeMobile: {
      2: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      3: {
        linePadding: 12,
        eyebrowPadding: 12,
        topBotPadding: 12,
        sidePadding: 12,
        titlePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      4: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
    },
    smallTablet: {
      2: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      3: {
        linePadding: 12,
        eyebrowPadding: 12,
        topBotPadding: 12,
        sidePadding: 12,
        titlePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      4: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
    },
    tablet: {
      2: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      3: {
        linePadding: 12,
        eyebrowPadding: 12,
        topBotPadding: 12,
        sidePadding: 12,
        titlePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      4: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
    },
    smallDesktop: {
      2: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      3: {
        linePadding: 16,
        eyebrowPadding: 12,
        topBotPadding: 16,
        sidePadding: 16,
        titlePadding: 16,
        buttonPadding: 16,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      4: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
    },
    mediumDesktop: {
      2: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      3: {
        linePadding: 16,
        eyebrowPadding: 12,
        topBotPadding: 16,
        sidePadding: 16,
        titlePadding: 16,
        buttonPadding: 16,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      4: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
    },
    desktop: {
      2: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      3: {
        linePadding: 16,
        eyebrowPadding: 12,
        topBotPadding: 16,
        sidePadding: 16,
        titlePadding: 16,
        buttonPadding: 16,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
      4: {
        topBotPadding: 12.5,
        sidePadding: 12,
        eyebrowPadding: 12,
        titlePadding: 12,
        linePadding: 12,
        buttonPadding: 10,
        iconSize: 'extraLarge',
        visibleSlides: 2,
      },
    },
  },

  getFixed: getFixed,
};

export default FixedConfig;
