// @vds-core/tooltips/dist/esm/components/Tooltip.js
import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject6() {
  var data = _taggedTemplateLiteral(["\n&&& {\n  transform: translate(-50%, -50%);\n  text-align: center;\n  cursor: pointer;\n  position: absolute;\n  top: 50%;\n  left: 50%;\n  bottom: 50%;\n  height: ", ";\n  width: ", ";\n  z-index: 1;\n}\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n&&& {\n  transform: translate(-50%, -50%);\n  text-align: center;\n  cursor: pointer;\n  position: absolute;\n  top: 50%;\n  left: 50%;\n  bottom: 50%;\n  height: ", ";\n  width: ", ";\n  z-index: 1;\n  outline: none;\n}\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n&&& {\n  position: absolute;\n  outline: none;\n}\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n&&& {\n  display: inline-block;\n  position: relative;\n}\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n&&& {\n  display: flex;\n  flex-direction: column;\n  outline: none;\n  z-index: ", ";\n}\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n&&& {\n  display: inline-flex;\n  align-items: center;\n  position: relative;\n  padding: 0;\n  outline: none;\n  ", ";\n}\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import UNSAFE_SetEnvRef from '@uie/set-env-ref';
import styled, { css } from 'styled-components';
import cuid from 'cuid';
import { calculateRem } from '@uie/javascript-style-helpers';
import AnchorIconBase from './AnchorIconBase';
import TooltipDialog from './TooltipDialog';
import AnchorIcon from '../utils/AnchorIcon';
import { IconData } from '@vds-core/icons';
import { BodyConfig, TypographyConfig } from '@vds-core/typography';
import { colors } from '@vds-core/theme';
import { withVDSManager } from '@vds-core/utilities';
var propTypes = {
  /**
   * @deprecated
   * Function that returns an HTML element or React component to be used as a custom anchor for the tooltip. The function is called with a React ref, and an object that will pass back aria-describedby and aria-expanded that has to be passed to the returned anchor element/component.
   */
  anchorElement: PropTypes.func,

  /**
   * String, React Component, or HTML to rendered as the body of the tooltip.
   */
  children: PropTypes.any,

  /**
   * @deprecated
   * String, React Component, or HTML to rendered as the header of the tooltip.
   */
  header: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

  /**
   * @deprecated
   * Boolean used to choose secondary version of icon element. <strong>Can only be used with <code>base icon element
   */
  secondary: PropTypes.bool,

  /**
   * Boolean used to invert icon element color.<strong>Can only be used with <code>base icon element</code></strong>
   */
  inverted: PropTypes.bool,

  /**
   * Size of icon
   */
  size: PropTypes.string,

  /**
   * Number used as a pixel value to offset space between tooltip and <code>anchorElement</code>. <strong>Can only be used with <code>anchorElement</code></strong>
   */
  offset: PropTypes.number,

  /**
   * String used to choose secondary or primary version of icon element. <strong>Can only be used with <code>base icon element</code></strong>
   */
  type: PropTypes.string,

  /**
   * String, React Component, or HTML to rendered as the header of the tooltip.
   */
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

  /**
   * A String label that is required for accessibility, the should be very terse description ot the tooltips content.
   */
  ariaLabel: PropTypes.string,

  /**
   * Function that returns an HTML element or React component to be used as a custom anchor for the tooltip. The function is called with a React ref, and an object that will pass back aria-describedby, aria-expanded, and aria-label that has to be passed to the returned anchor element/component.
   */
  renderAnchorElement: PropTypes.func,

  /**
   *@ignore
   *Icon package used to render info icon as anchor
   */
  iconConfig: PropTypes.object,

  /**
   *@ignore
   *Enables focus state with dashed lines
   */
  focusState: PropTypes.bool,
  //todo:

  /**
   *@ignore
   *Enables hover state with medium weight icon
   */
  hoverState: PropTypes.bool,

  /**
   *@ignore
   *Enables grey medium weight icon on active
   */
  activeState: PropTypes.bool,

  /**
   *@ignore
   *configuration for body component
   */
  bodyConfig: PropTypes.object,

  /**
   *@ignore
   *configuration for body component
   */
  colorConfig: PropTypes.object,

  /**
   *
   *@ignore// typescale for body config
   */
  typescale: PropTypes.string,

  /**
   *
   *@ignore// padding for inside of dialog
   */
  dialogPadding: PropTypes.string,

  /**
   * Allows a unique ID to be passed to the component.
   */
  id: PropTypes.string,

  /**
   * Boolean that activates disabled state
   */
  disabled: PropTypes.bool,

  /**
   * Allows a unique component ID to be passed to the component as a reference for positioning other than the window object.
   */
  containerId: PropTypes.string
};
var defaultProps = {
  anchorElement: null,
  header: null,
  secondary: false,
  inverted: false,
  size: 'large',
  offset: 0,
  //added after renaming
  type: null,
  title: null,
  ariaLabel: 'More information',
  renderAnchorElement: null,
  iconConfig: IconData,
  focusState: false,
  hoverState: false,
  activeState: false,
  bodyConfig: BodyConfig,
  typescale: TypographyConfig.getTypescale(),
  dialogPadding: calculateRem(12, 0, 16, 12),
  colorConfig: colors,
  disabled: false
};
var ARIA_LIVE_VALUE = 'polite';
var TooltipWrapper = /*#__PURE__*/styled.span.withConfig({
  componentId: "VDS__sc-8y4bkd-0"
})(_templateObject(), function (_ref) {
  var disabled = _ref.disabled;
  return disabled && "\n    pointer-events: none;\n    cursor: default;\n  ";
});
var TooltipIconWrapper = /*#__PURE__*/styled.span.withConfig({
  componentId: "VDS__sc-8y4bkd-1"
})(_templateObject2(), function (_ref2) {
  var clicked = _ref2.clicked,
      showDialog = _ref2.showDialog,
      hovered = _ref2.hovered;
  return clicked || hovered || showDialog ? 2 : 0;
});
var CustomAnchorElementWrapper = /*#__PURE__*/styled.span.withConfig({
  componentId: "VDS__sc-8y4bkd-2"
})(_templateObject3());
var IconPositionWrapper = /*#__PURE__*/styled.span.withConfig({
  componentId: "VDS__sc-8y4bkd-3"
})(_templateObject4());
var HitArea = /*#__PURE__*/styled.span.withConfig({
  componentId: "VDS__sc-8y4bkd-4"
})(_templateObject5(), calculateRem(44), calculateRem(44));
var HoverZone = /*#__PURE__*/styled.span.withConfig({
  componentId: "VDS__sc-8y4bkd-5"
})(_templateObject6(), calculateRem(44), calculateRem(228));

function _isiOS() {
  var isIPAD = navigator.platform === 'MacIntel' && navigator.maxTouchPoints > 1 && !window.MSStream;
  var isMobileDevice = /iPhone|iPad/i.test(navigator.userAgent);
  return isMobileDevice || isIPAD;
}

function _checkIfIe() {
  var ua = window.navigator.userAgent;
  var msie = ua.indexOf('MSIE ');

  if (msie > 0) {
    // IE 10 or older => return version number
    return parseInt(ua.substring(msie + 5, ua.indexOf('.', msie)), 10);
  }

  var trident = ua.indexOf('Trident/');

  if (trident > 0) {
    // IE 11 => return version number
    var rv = ua.indexOf('rv:');
    return parseInt(ua.substring(rv + 3, ua.indexOf('.', rv)), 10);
  }

  var edge = ua.indexOf('Edge/');

  if (edge > 0) {
    // Edge => return version number
    return parseInt(ua.substring(edge + 5, ua.indexOf('.', edge)), 10);
  } // other browser


  return false;
}

var Tooltip = /*#__PURE__*/function (_Component) {
  _inherits(Tooltip, _Component);

  var _super = _createSuper(Tooltip);

  function Tooltip() {
    var _this;

    _classCallCheck(this, Tooltip);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      hovered: false,
      keyboardFocused: false,
      showDialog: false,
      clicked: false,
      hoveredWhenFocused: false,
      active: false,
      customAnchorPresent: false,
      isIe: _checkIfIe(),
      isiOS: _isiOS(),
      scrollPosition: null,
      expanded: false,
      tooltipAnchorCoordinates: {}
    });

    _defineProperty(_assertThisInitialized(_this), "componentDidMount", function () {
      document.addEventListener('click', _this.handleClickOutside);
      window.addEventListener('scroll', _this.setAnchorElementPositionProperties());
      window.addEventListener('resize', _this.setAnchorElementPositionProperties()); // this check is to see if a custom element is being used as an anchor
      // this flag will activate different styles for the dialog based on the boolean

      if (_this.props.anchorElement || _this.props.renderAnchorElement) {
        _this.setState({
          customAnchorPresent: true
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "componentDidUpdate", function (prevProps, prevState) {
      if (prevState.showDialog !== _this.state.showDialog) {
        document.getElementById(_this.tooltipId).focus();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "componentWillUnmount", function () {
      document.removeEventListener('click', _this.handleClickOutside);
      window.removeEventListener('scroll', _this.setAnchorElementPositionProperties());
      window.removeEventListener('resize', _this.setAnchorElementPositionProperties());
    });

    _defineProperty(_assertThisInitialized(_this), "setAnchorElementPositionProperties", function () {
      // get the reference to the current tooltipAnchor
      var tooltipAnchor = _this.toolTipAnchorRef;
      var tooltipAnchorRef = tooltipAnchor;
      var tooltipAnchorCoordinates = tooltipAnchorRef.getBoundingClientRect(); // assign tooltipAnchorCoordinates the value of the current anchor ref boundingClientRect

      _this.setState({
        tooltipAnchorCoordinates: tooltipAnchorCoordinates
      });
    });

    _defineProperty(_assertThisInitialized(_this), "setDialogState", function () {
      return _this.setState({
        showDialog: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "setClickedState", function () {
      return _this.setState(function (prevState) {
        return {
          clicked: !prevState.clicked
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "deactivateDialogState", function () {
      return _this.setState({
        showDialog: false
      });
    });

    _defineProperty(_assertThisInitialized(_this), "deactivateClickedState", function () {
      return _this.setState({
        keyboardFocused: false,
        forceClose: false
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyboardInput", function (e, focused) {
      var showDialog = _this.state.showDialog;
      if (e.keyCode === 32) e.preventDefault(); //For the new passthrough icon focus

      if (_this.props.focusState && focused && !_this.state.keyboardFocused && !_this.state.clicked && !_this.state.showDialog) {
        _this.setState({
          keyboardFocused: !_this.state.keyboardFocused
        });
      } else if (_this.props.focusState && !focused && _this.state.keyboardFocused) {
        _this.setState({
          keyboardFocused: !_this.state.keyboardFocused
        });
      }

      if (e.keyCode === 13 || e.keyCode === 32 || // return key
      e.keyCode === 27 && showDialog // esc key
      ) {
          _this.setState({
            // toggle showDialog state based on keyboard interaction
            showDialog: !_this.state.showDialog,
            keyboardFocused: !_this.state.keyboardFocused,
            expanded: !_this.state.expanded
          });
        } else if (e.shiftKey && e.keyCode === 9) {
        _this.deactivateDialogState(); // deactivate dialog if user shift tabs

      } else if (e.keyCode === 9 && _this.showDialog) {
        _this.deactivateDialogState(); // deactivate dialog if use tabs away

      } else if (e.keyCode === 9) {
        _this.setState({
          keyboardFocused: !_this.state.keyboardFocused
        });
      }

      if (_this.anchorWithRef !== null && _this.anchorElementProps.props.onKeyDown) {
        _this.anchorElementProps.props.onKeyDown(e);
      }

      if (_this.anchorWithRef !== null && _this.anchorElementProps.props.onBlur) {
        _this.anchorElementProps.props.onBlur(e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleClickOutside", function (e) {
      // check if current event contains ref if not close dialog
      var toolTipAnchorRef = _this.toolTipAnchorRef;

      if (toolTipAnchorRef && !toolTipAnchorRef.contains(e.target)) {
        _this.deactivateDialogState();

        _this.deactivateClickedState();

        _this.setState({
          scrollPosition: 0
        });

        if (_this.state.clicked) {
          _this.setClickedState();
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleMouseEnter", function (e) {
      if (_this.state.keyboardFocused) {
        _this.setState({
          hoveredWhenFocused: true,
          keyboardFocused: false
        });
      }

      if (_this.state.showDialog && _this.state.clicked) {
        _this.setState({
          hovered: true
        });

        return;
      }

      _this.openTooltipTimeout = setTimeout(function () {
        _this.setState({
          showDialog: true,
          hovered: true
        });
      }, 200);

      if (_this.anchorWithRef !== null && _this.anchorElementProps.props.onMouseEnter) {
        _this.anchorElementProps.props.onMouseEnter(e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleMouseExit", function (e) {
      if (_this.state.clicked && _this.state.showDialog) {
        return;
      } // clear time out to prevent showDialog being set to true unintentionally


      clearTimeout(_this.openTooltipTimeout);

      _this.setState({
        hovered: false,
        showDialog: false,
        forceClose: false
      });

      if (_this.anchorWithRef !== null && _this.anchorElementProps.props.onMouseLeave) {
        _this.anchorElementProps.props.onMouseLeave(e);
      }

      if (_this.state.hoveredWhenFocused) {
        _this.setState({
          hoveredWhenFocused: false,
          keyboardFocused: true
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleMouseDown", function (e) {
      _this.setState({
        active: true
      });

      if (_this.state.clicked) {
        _this.setState({
          forceClose: true
        });
      }

      _this.deactivateDialogState();

      _this.setDialogState();

      _this.setClickedState();

      if (_this.anchorWithRef !== null && _this.anchorElementProps.props.onMouseDown) {
        _this.anchorElementProps.props.onMouseDown(e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleMouseUp", function (e) {
      _this.setState({
        active: false
      });

      if (_this.state.forceClose) {
        _this.deactivateDialogState();

        _this.deactivateClickedState();

        e.currentTarget.blur();
      }

      if (_this.anchorWithRef !== null && _this.anchorElementProps.props.onMouseUp) {
        _this.anchorElementProps.props.onMouseUp(e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_onFocus", function (e) {
      _this.handleKeyboardInput(e, true);
    });

    _defineProperty(_assertThisInitialized(_this), "setAnchorElement", function () {
      var computedProps = _objectSpread({
        ariaDescribedBy: _this.ariaId,
        ariaExpanded: _this.state.showDialog,
        ariaLabel: _this.state.ariaLabel,
        ariaLive: ARIA_LIVE_VALUE
      }, _defineProperty({}, _this.refKey, function (elem) {
        return _this.toolTipAnchorRef = elem;
      }));

      if (_this.props.anchorElement) {
        return _this.props.anchorElement(computedProps);
      } else if (_this.props.renderAnchorElement) {
        return _this.props.renderAnchorElement(computedProps);
      }

      return null;
    });

    _defineProperty(_assertThisInitialized(_this), "_setScrollPosition", function (y) {
      _this.setState({
        scrollPosition: y
      });
    });

    _defineProperty(_assertThisInitialized(_this), "refKey", UNSAFE_SetEnvRef());

    _defineProperty(_assertThisInitialized(_this), "ariaId", "tooltip-".concat(cuid()));

    _defineProperty(_assertThisInitialized(_this), "tooltipId", cuid());

    _defineProperty(_assertThisInitialized(_this), "anchorWithRef", _this.setAnchorElement());

    _defineProperty(_assertThisInitialized(_this), "anchorElementProps", _this.anchorWithRef);

    _defineProperty(_assertThisInitialized(_this), "renderTooltipDialog", function () {
      if (_this.state.showDialog && !_this.props.disabled) {
        return /*#__PURE__*/React.createElement(TooltipDialog, {
          scrollPosition: _this.state.scrollPosition,
          setScrollPosition: _this._setScrollPosition,
          colorConfig: _this.props.colorConfig,
          bodyConfig: _this.props.bodyConfig,
          typescale: _this.props.typescale,
          dialogPadding: _this.props.dialogPadding,
          clicked: _this.state.clicked,
          customAnchorElement: _this.toolTipAnchorRef,
          customAnchorPresent: _this.state.customAnchorPresent,
          content: _this.props.children,
          deactivateDialogState: _this.deactivateDialogState,
          header: _this.props.header,
          title: _this.props.title,
          ariaId: _this.ariaId,
          keyboardFocused: _this.state.keyboardFocused,
          keepDialogActiveOnMouseEnter: _this.setDialogState,
          toolTipAnchorRef: _this.toolTipAnchorRef,
          showDialog: _this.state.showDialog,
          offset: _this.props.offset,
          size: _this.props.size,
          inverted: _this.props.inverted,
          isIe: _this.state.isIe,
          containerId: _this.props.containerId,
          tooltipAnchorCoordinates: _this.state.tooltipAnchorCoordinates
        });
      }

      return null;
    });

    _defineProperty(_assertThisInitialized(_this), "renderIconAnchorElement", function () {
      var _this$props = _this.props,
          iconConfig = _this$props.iconConfig,
          colorConfig = _this$props.colorConfig,
          focusState = _this$props.focusState,
          disabled = _this$props.disabled;
      return /*#__PURE__*/React.createElement(TooltipIconWrapper, {
        tabIndex: -1,
        importantForAccessibility: "no",
        onMouseEnter: disabled ? undefined : _this.handleMouseEnter,
        onMouseDown: disabled ? undefined : _this.handleMouseDown,
        onMouseLeave: disabled ? undefined : _this.handleMouseExit,
        onMouseUp: disabled ? undefined : _this.handleMouseUp,
        clicked: _this.state.clicked,
        showDialog: _this.state.showDialog,
        hovered: _this.state.hovered
      }, /*#__PURE__*/React.createElement(AnchorIconBase, _extends({
        role: "button",
        "data-testid": "tooltip",
        id: _this.tooltipId,
        tabIndex: 0,
        colorConfig: colorConfig,
        focusState: focusState,
        "aria-describedby": _this.ariaId,
        "aria-labelledby": _this.state.isiOS ? _this.ariaId : null,
        "aria-expanded": _this.state.showDialog,
        "aria-label": _this.props.ariaLabel,
        hovered: _this.state.hovered,
        clicked: _this.state.clicked,
        focusable: "false" // to make IE11 happy
        ,
        onFocus: !_this.props.disabled ? _this._onFocus : undefined,
        onKeyDown: !_this.props.disabled ? _this.handleKeyboardInput : undefined,
        secondary: _this.props.secondary,
        type: _this.props.type,
        showDialog: _this.state.showDialog,
        size: _this.props.size,
        inverted: _this.props.inverted,
        keyboardFocused: _this.state.keyboardFocused,
        onClick: _this._onclick,
        disabled: _this.props.disabled
      }, _defineProperty({}, _this.refKey, function (elem) {
        return _this.toolTipAnchorRef = elem;
      })), /*#__PURE__*/React.createElement("span", {
        tabIndex: -1,
        style: {
          outline: 'none',
          display: 'inline-flex'
        }
      }, /*#__PURE__*/React.createElement(AnchorIcon, {
        tabIndex: -1,
        colorConfig: colorConfig,
        iconConfig: iconConfig,
        disabled: _this.props.disabled,
        onClick: _this._onclick,
        importantForAccessibility: "no",
        size: _this.props.size,
        secondary: _this.props.secondary,
        type: _this.props.type,
        showDialog: _this.state.showDialog,
        inverted: _this.props.inverted,
        clicked: _this.state.clicked,
        hovered: _this.state.hovered,
        activeState: _this.props.activeState,
        hoverState: _this.props.hoverState,
        keyboardFocused: _this.state.keyboardFocused,
        active: _this.state.active
      }), !_this.props.disabled && /*#__PURE__*/React.createElement(HitArea, {
        tabIndex: -1,
        importantForAccessibility: "no",
        "data-testid": "hit-area"
      }), _this.renderTooltipDialog())));
    });

    _defineProperty(_assertThisInitialized(_this), "renderCustomAnchorElement", function () {
      var computedProps = {
        onMouseEnter: _this.handleMouseEnter,
        onMouseDown: _this.handleMouseDown,
        onMouseLeave: _this.handleMouseExit,
        onMouseUp: _this.handleMouseUp,
        onKeyDown: _this.handleKeyboardInput,
        onBlur: _this.handleKeyboardInput
      };

      var anchorWithRef = _this.setAnchorElement();

      return /*#__PURE__*/React.createElement(CustomAnchorElementWrapper, _extends({}, _defineProperty({}, _this.refKey, function (elem) {
        return _this.toolTipAnchorRef = elem;
      }), {
        id: _this.tooltipId
      }, computedProps), /*#__PURE__*/React.cloneElement(anchorWithRef), _this.renderTooltipDialog());
    });

    return _this;
  }

  _createClass(Tooltip, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          id = _this$props2.id,
          className = _this$props2.className;
      return /*#__PURE__*/React.createElement(TooltipWrapper, {
        id: id,
        className: className,
        size: this.props.size,
        trailing: this.props.trailing
      }, this.props.anchorElement || this.props.renderAnchorElement ? this.renderCustomAnchorElement() : this.renderIconAnchorElement());
    }
  }]);

  return Tooltip;
}(Component);

Tooltip.defaultProps = defaultProps;
Tooltip.propTypes = propTypes;
export default withVDSManager(Tooltip);
