// @vds-core/selects/src/components/DropdownMarketingOption.jsx
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { verizonTheme as theme } from '@vds-core/theme';

const propTypes = {
  /**
   * Content in the dropdown item.
   */
  children: PropTypes.string,
  /**
   * Function run when option is selected.
   */
  onSelect: PropTypes.func,
  /**
   * The id for the selected item.
   */
  selected: PropTypes.string,
  /**
   * ignore
   */
  id: PropTypes.string,
  /**
   * ignore
   */
  selectionCallback: PropTypes.func,
};

const defaultProps = {
  children: null,
  onSelect: () => {},

  selected: undefined,
};

const Wrapper = styled.li`
  cursor: pointer;
`;

const DropdownMarketingOption = ({
  children,
  onSelect,
  selected,
  id,
  selectionCallback,
}) => {
  function _clicked(e) {
    // Manually construct event here =>
    // because changing to a-tag doesn't pass the same event
    let event = {
      target: {
        value: children,
      },
    };
    onSelect(event);
    selectionCallback(event);
    if (e.type === 'click' && e.clientX !== 0 && e.clientY !== 0) {
      // This is a real click. Do something here
      e.target.blur();
    }
  }

  function _onKeyDown(e) {
    let event = {
      target: {
        value: children,
      },
    };
    if (e.keyCode === 13 || e.keyCode === 32) {
      onSelect(event);
      selectionCallback(event);
    }
  }

  return (
    <Wrapper
      tabIndex={0}
      onKeyDown={_onKeyDown}
      id={id}
      value={children}
      onClick={_clicked}
      role="none"
    >
      {children}
    </Wrapper>
  );
};

DropdownMarketingOption.defaultProps = defaultProps;
DropdownMarketingOption.propTypes = propTypes;

export default DropdownMarketingOption;
