// @vds-core/theme/dist/esm/verizonTheme/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { calculateRem } from '@uie/javascript-style-helpers';
import colors from './colors';
import font from './font';
var general = {
  // General
  primary: colors.black,
  secondary: colors.white,
  accent: colors.red,
  background: colors.white,
  backgroundColorDisabled: colors.coolGray3,
  // Messaging Types
  success: colors.green,
  info: colors.blue,
  warning: colors.yellow,
  error: colors.orange,
  disabled: colors.coolGray3,
  textColorSuccess: colors.white,
  textColorInfo: colors.white,
  textColorWarning: colors.black,
  textColorError: colors.black,
  textColorDisabled: colors.white,
  // Borders
  borderWidth: calculateRem(1),
  borderStyle: 'solid',
  borderColor: colors.coolGray3,
  border: "".concat(calculateRem(1), " solid ").concat(colors.coolGray3),
  borderColorDisabled: colors.coolGray3,
  // Links
  linkColor: colors.black,
  linkColorHover: colors.blue,
  linkColorActive: colors.blue,
  linkDecoration: 'underline',
  // Transitions
  transitionDuration: '0.2s',
  // Horizontal Rule
  hrPrimary: "".concat(calculateRem(4), " solid ").concat(colors.charcoal),
  hrSecondary: "".concat(calculateRem(1), " solid ").concat(colors.charcoal),
  hrTertiary: "".concat(calculateRem(1), " solid ").concat(colors.coolGray3)
};
export default _objectSpread(_objectSpread(_objectSpread({
  name: 'Verizon Theme'
}, colors), font), general);
