// @vds/tooltips/dist/cjs/components/Tooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _tooltips = require("@vds-core/tooltips");

var _typography = require("@vds/typography");

var _icons = require("@vds/icons");

var _theme = require("@vds/theme");

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var Tooltip = function Tooltip(props) {
  return /*#__PURE__*/_react["default"].createElement(_tooltips.Tooltip, (0, _extends2["default"])({}, props, {
    focusState: true,
    hoverState: true,
    activeState: true,
    iconConfig: _icons.IconData,
    bodyConfig: _typography.BodyConfig,
    colorConfig: _theme.colors,
    typescale: _typography.TypographyConfig.getTypescale(),
    dialogPadding: (0, _javascriptStyleHelpers.calculateRem)(12, 0, 12, 12)
  }));
};

var _default = Tooltip;
exports["default"] = _default;
