// @vds/tooltips/dist/cjs/components/TrailingTooltip.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _typography = require("@vds/typography");

var _theme = require("@vds/theme");

var _icons = require("@vds/icons");

var _tooltips = require("@vds-core/tooltips");

var TrailingTooltip = function TrailingTooltip(props) {
  return /*#__PURE__*/_react["default"].createElement(_tooltips.TrailingTooltip, (0, _extends2["default"])({}, props, {
    Body: _typography.Body,
    Title: _typography.Title,
    Micro: _typography.Micro,
    Subtitle: _typography.Subtitle,
    colorConfig: _theme.colors,
    bodyConfig: _typography.BodyConfig,
    iconConfig: _icons.IconData
  }));
};

var _default = TrailingTooltip;
exports["default"] = _default;
