// @vds-core/theme/dist/cjs/layout/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "setSpacingFactor", {
  enumerable: true,
  get: function get() {
    return _spacing.setSpacingFactor;
  }
});
Object.defineProperty(exports, "getSpacingFactor", {
  enumerable: true,
  get: function get() {
    return _spacing.getSpacingFactor;
  }
});
Object.defineProperty(exports, "setSpacingScale", {
  enumerable: true,
  get: function get() {
    return _spacing.setSpacingScale;
  }
});
Object.defineProperty(exports, "getSpacingScale", {
  enumerable: true,
  get: function get() {
    return _spacing.getSpacingScale;
  }
});
Object.defineProperty(exports, "calculateSpacing", {
  enumerable: true,
  get: function get() {
    return _spacing.calculateSpacing;
  }
});
Object.defineProperty(exports, "vdsSpacingScale", {
  enumerable: true,
  get: function get() {
    return _spacing.vdsSpacingScale;
  }
});
Object.defineProperty(exports, "breakpoints", {
  enumerable: true,
  get: function get() {
    return _breakpoints["default"];
  }
});
Object.defineProperty(exports, "spacers", {
  enumerable: true,
  get: function get() {
    return _spacers["default"];
  }
});

var _spacing = require("./spacing");

var _breakpoints = _interopRequireDefault(require("./breakpoints"));

var _spacers = _interopRequireDefault(require("./spacers"));
