// @vds-core/tabs/src/components/TabPanel.jsx
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const propTypes = {
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * @ignore
   */
  label: PropTypes.string.isRequired,
  /**
   * @ignore
   */
  isHidden: PropTypes.bool,
  /**
   * @ignore
   */
  uniqueId: PropTypes.string,
  /**
   * @ignore
   */
  className: PropTypes.string,
};

const defaultProps = {
  isHidden: false,
  uniqueId: undefined,
  className: null,
};

const StyledPanel = styled.div`
  display: none;
  outline: none;
`;

/**
 * @ignore
 */
const TabPanel = ({ isHidden, label, uniqueId, children, className }) => {
  const styles = isHidden ? { display: 'none' } : { display: 'block' };

  return (
    <StyledPanel
      id={`tabPanel_${uniqueId}`}
      aria-hidden={isHidden}
      aria-labelledby={`${uniqueId}`}
      role="tabpanel"
      style={styles}
      tabIndex="-1"
      className={className}
    >
      {children}
    </StyledPanel>
  );
};

TabPanel.propTypes = propTypes;
TabPanel.defaultProps = defaultProps;

export default TabPanel;
