// @vds/typography/dist/esm/components/Title/config.js
import { Fonts } from '../../fonts';

function getStyles(typescale, viewport, typeSize) {
  return TitleConfig[typescale][viewport][typeSize];
}

var TitleConfig = {
  VDS: {
    desktop: {
      '2XLarge': {
        fontSize: 64,
        lineHeight: 64,
        fontWeight: 700,
        fontFamily: Fonts.VerizonNHGeDS
      },
      XLarge: {
        fontSize: 48,
        lineHeight: 48,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      large: {
        fontSize: 32,
        lineHeight: 32,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      medium: {
        fontSize: 24,
        lineHeight: 24,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      small: {
        fontSize: 20,
        lineHeight: 20,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      XSmall: {
        fontSize: 14,
        lineHeight: 16,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      }
    },
    mobile: {
      '2XLarge': {
        fontSize: 36,
        lineHeight: 36,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      XLarge: {
        fontSize: 32,
        lineHeight: 32,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      large: {
        fontSize: 24,
        lineHeight: 24,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      medium: {
        fontSize: 20,
        lineHeight: 20,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      small: {
        fontSize: 16,
        lineHeight: 16,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      },
      XSmall: {
        fontSize: 14,
        lineHeight: 16,
        fontWeight: 'bold',
        fontFamily: Fonts.VerizonNHGeDS
      }
    }
  },
  getStyles: getStyles
};
export default TitleConfig;
