// @vds-core/theme/dist/esm/layout/spacing/calculateSpacing.js
import _typeof from "@babel/runtime/helpers/typeof";
import { calculateRem } from '@uie/javascript-style-helpers';
import { vdsSpacingScale } from './spacingScale';
var SpacingConfig = {
  factor: 4,
  scale: vdsSpacingScale
};

function setSpacingFactor(factor) {
  if (typeof factor !== 'number') {
    return console.error('The factor passed is not a number!');
  }

  SpacingConfig.factor = factor;
}

function getSpacingFactor() {
  return SpacingConfig.factor;
}

function setSpacingScale(scale) {
  if (_typeof(scale) !== 'object') {
    return console.error('The config passed is not an object!');
  }

  SpacingConfig.scale = scale;
}

function getSpacingScale() {
  return SpacingConfig.scale;
}

function calculateSpacing(unitName, unit) {
  var scale = SpacingConfig.scale,
      factor = SpacingConfig.factor;
  var num = scale[unitName] * factor;

  if (!num) {
    return console.error('Size is not supported by spacing scale!');
  }

  return unit === 'rem' ? calculateRem(num) : "".concat(num, "px");
}

export { setSpacingFactor, getSpacingFactor, setSpacingScale, getSpacingScale, calculateSpacing };
