// @vds-core/theme/dist/cjs/verizonTheme/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _colors = _interopRequireDefault(require("./colors"));

var _font = _interopRequireDefault(require("./font"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var general = {
  // General
  primary: _colors["default"].black,
  secondary: _colors["default"].white,
  accent: _colors["default"].red,
  background: _colors["default"].white,
  backgroundColorDisabled: _colors["default"].coolGray3,
  // Messaging Types
  success: _colors["default"].green,
  info: _colors["default"].blue,
  warning: _colors["default"].yellow,
  error: _colors["default"].orange,
  disabled: _colors["default"].coolGray3,
  textColorSuccess: _colors["default"].white,
  textColorInfo: _colors["default"].white,
  textColorWarning: _colors["default"].black,
  textColorError: _colors["default"].black,
  textColorDisabled: _colors["default"].white,
  // Borders
  borderWidth: (0, _javascriptStyleHelpers.calculateRem)(1),
  borderStyle: 'solid',
  borderColor: _colors["default"].coolGray3,
  border: "".concat((0, _javascriptStyleHelpers.calculateRem)(1), " solid ").concat(_colors["default"].coolGray3),
  borderColorDisabled: _colors["default"].coolGray3,
  // Links
  linkColor: _colors["default"].black,
  linkColorHover: _colors["default"].blue,
  linkColorActive: _colors["default"].blue,
  linkDecoration: 'underline',
  // Transitions
  transitionDuration: '0.2s',
  // Horizontal Rule
  hrPrimary: "".concat((0, _javascriptStyleHelpers.calculateRem)(4), " solid ").concat(_colors["default"].charcoal),
  hrSecondary: "".concat((0, _javascriptStyleHelpers.calculateRem)(1), " solid ").concat(_colors["default"].charcoal),
  hrTertiary: "".concat((0, _javascriptStyleHelpers.calculateRem)(1), " solid ").concat(_colors["default"].coolGray3)
};

var _default = _objectSpread(_objectSpread(_objectSpread({
  name: 'Verizon Theme'
}, _colors["default"]), _font["default"]), general);

exports["default"] = _default;
