// @vds-core/tooltips/src/utils/vendorStyles.js
/*
 * Requires CSS:
 * https://www.npmjs.com/package/react-custom-scroll
 * https://unpkg.com/react-custom-scroll@4.0.0/dist/customScroll.css
 *
 * */

import { css } from 'styled-components';

const vendorStyles = css`
  .persistentScrollbar .rcs-custom-scroll::after {
    opacity: 1;
  }

  .persistentScrollbar .rcs-positioning {
    opacity: 1;
  }

  .rcs-custom-scroll {
    min-height: 0;
    min-width: 0;
  }

  .rcs-custom-scroll .rcs-outer-container {
    overflow: hidden;
  }

  .rcs-custom-scroll::after {
    content: '';
    position: absolute;
    right: 4px;
    width: 4px;
    background-color: #dad8d8;
    top: 12px;
    bottom: 12px;
    z-index: 30;
    opacity: 0;
    transition: 0.15s ease-out opacity;
  }

  .rcs-custom-scroll:hover::after {
    opacity: 1;
  }

  .rcs-positioning {
    z-index: 50;
    top: 6px;
    opacity: 0;
    position: relative;
    transition: 0.15s ease-out opacity;
  }

  .rcs-custom-scroll:hover .rcs-positioning {
    opacity: 1;
  }

  .rcs-custom-scroll .rcs-inner-container {
    overflow-x: hidden;
    overflow-y: scroll;
    -webkit-overflow-scrolling: touch;
    padding-right: 14px;
  }

  .rcs-custom-scroll .rcs-inner-container:after {
    content: '';
    position: absolute;
    top: 0;
    right: 0;
    left: 0;
    height: 0;
    background-image: linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0.2) 0%,
      rgba(0, 0, 0, 0.05) 60%,
      transparent 100%
    );
    pointer-events: none;
    transition: height 0.1s ease-in;
    will-change: height;
  }

  .rcs-custom-scroll .rcs-inner-container.rcs-content-scrolled:after {
    height: 5px;
    transition: height 0.15s ease-out;
  }

  .rcs-custom-scroll.rcs-scroll-handle-dragged .rcs-inner-container {
    user-select: none;
  }

  .rcs-custom-scroll .rcs-custom-scrollbar {
    position: absolute;
    height: 100%;
    width: 4px;
    right: 4px;
    opacity: 1;
    z-index: 1;
    transition: opacity 0.4s ease-out;
    padding: 6px 0;
    box-sizing: border-box;
    will-change: opacity;
    pointer-events: none;
  }

  .rcs-custom-scroll .rcs-custom-scrollbar.rcs-custom-scrollbar-rtl {
    right: auto;
    left: 3px;
  }

  .rcs-custom-scroll.rcs-scroll-handle-dragged .rcs-custom-scrollbar {
    opacity: 1;
  }

  .rcs-custom-scroll .rcs-custom-scroll-handle {
    position: absolute;
    width: 100%;
    top: 0;
  }

  .rcs-custom-scroll .rcs-inner-handle {
    height: calc(100% - 25px);
    margin-top: 6px;
    background-color: #747676;
  }
`;

export default vendorStyles;
