// @vds-core/tabs/dist/cjs/util/Animator.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _setEnvRef = _interopRequireDefault(require("@uie/set-env-ref"));

var _animationConfig = _interopRequireDefault(require("./animationConfig"));

var _utilities = require("@vds/utilities");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: relative;\n  ", ";\n\n  animation-play-state: ", ";\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n    animation: ", " ", " ", " ", ";\n  "]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["", ""]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

// This polyfill prevents Intersection Observer from breaking IE
require('intersection-observer');

var propTypes = {
  /**
   * @ignore
   * ref for component
   */
  // setAnimationRef: PropTypes.func,

  /**
  /**
   * Animation name
   */
  name: _propTypes["default"].oneOf(['slideUp', 'fadeInUp']),

  /**
   * Pass a boolean value to start the animation, if set to true, animation starts when component mounts.
   */
  startAnimation: _propTypes["default"].bool,

  /**
   *
   */
  animated: _propTypes["default"].bool
};
var defaultProps = {
  // setAnimationRef: () => {},
  name: 'slideUp',
  startAnimation: false,
  animated: true
};

var buildAnimation = function buildAnimation(animation) {
  if (animation === {} || !animation) {
    return null;
  }

  var duration = animation.duration,
      delay = animation.delay,
      timingFunction = animation.timingFunction,
      keyFrame = animation.keyFrame;
  var animationKeyframe = /*#__PURE__*/(0, _styledComponents.keyframes)(_templateObject(), keyFrame); // use css helper when using keyframes see https://styled-components.com/docs/basics#animations

  return (0, _styledComponents.css)(_templateObject2(), duration, timingFunction, delay, animationKeyframe);
};

var AnimatorWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "AnimatorWrapper",
  componentId: "sc-12xw7k2-0"
})(_templateObject3(), function (_ref) {
  var animation = _ref.animation,
      animated = _ref.animated;
  return animated && buildAnimation(animation);
}, function (_ref2) {
  var startAnimation = _ref2.startAnimation;
  return startAnimation ? 'running' : 'paused';
});

var Animator = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Animator, _Component);

  var _super = _createSuper(Animator);

  function Animator(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Animator);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "componentDidMount", function () {
      // this.props.setAnimationRef(this.animationElem);
      if (_this.props.animated) _this._setupObserver();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "componentDidUpdate", function (prevProps) {
      if (prevProps.animated !== _this.props.animated) {
        if (_this.props.animated) _this._setupObserver();
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_setupObserver", function () {
      var options = {
        // root: rootRef,
        rootMargin: '150px 0px 0px 0px',
        threshold: 0
      };
      _this.observer = new IntersectionObserver(_this._observationHandler, options);

      var elem = _reactDom["default"].findDOMNode(_this.animationElem);

      if (!_this.observer || !elem) return;

      _this.observer.observe(elem);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_observationHandler", function (entries, observer) {
      entries.forEach(function (entry) {
        _this.shouldAnimate = entry.isIntersecting;

        if (_this.shouldAnimate) {
          _this.setState({
            startAnimation: true
          });

          _this.observer.disconnect();
        }
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setAnimationRef", function (element) {
      _this.animationElem = element;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "refKey", (0, _setEnvRef["default"])());
    _this.state = {
      startAnimation: false
    };
    return _this;
  }

  (0, _createClass2["default"])(Animator, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          name = _this$props.name,
          animated = _this$props.animated;
      var startAnimation = this.state.startAnimation;
      var animation = _animationConfig["default"][name]; // uncomment bellow if custom keyframes are passed
      // if (keyFrames) {
      //   animation.keyFrame = `${keyFrames}`;
      // }

      return /*#__PURE__*/_react["default"].createElement(AnimatorWrapper, (0, _extends2["default"])({}, (0, _defineProperty2["default"])({}, this.refKey, function (ref) {
        return _this2.setAnimationRef(ref);
      }), {
        animated: animated,
        animation: animation,
        startAnimation: startAnimation
      }), this.props.children);
    }
  }]);
  return Animator;
}(_react.Component);

Animator.propTypes = propTypes;
Animator.defaultProps = defaultProps;

var _default = (0, _utilities.withVDSManager)(Animator);

exports["default"] = _default;
