// @vds-core/tabs/dist/cjs/util/PhantomLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _buttons = require("@vds-core/buttons");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  margin: 32px 20px 0;\n  display: ", ";\n  opacity: 1;\n  position: absolute;\n  top: ", ";\n  left: ", ";\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var PhantomLinkWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "PhantomLinkWrapper",
  componentId: "sc-1jkunwc-0"
})(_templateObject(), function (_ref) {
  var isActive = _ref.isActive;
  return isActive ? 'inherit' : 'none';
}, function (_ref2) {
  var position = _ref2.position;
  return position && "".concat(position.top, "px");
}, function (_ref3) {
  var position = _ref3.position;
  return position && "".concat(position.left, "px");
}); // Link used for accessibility purposes, will be active only on desktop
// and when the content is focused from the active tab item on the left.


var PhantomLink = function PhantomLink(_ref4) {
  var ctaLink = _ref4.ctaLink,
      isDesktop = _ref4.isDesktop,
      isActive = _ref4.isActive,
      getPosition = _ref4.getPosition,
      TextLinkCaret = _ref4.TextLinkCaret;
  var singleLinkPos = isActive && getPosition();
  return isDesktop && /*#__PURE__*/_react["default"].createElement(PhantomLinkWrapper, {
    position: singleLinkPos,
    isActive: isActive
  }, /*#__PURE__*/_react["default"].createElement(TextLinkCaret, (0, _extends2["default"])({
    ariaHidden: isActive,
    disabled: !isActive
  }, ctaLink)));
};

var _default = PhantomLink;
exports["default"] = _default;
