// @vds-core/theme/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "verizonTheme", {
  enumerable: true,
  get: function get() {
    return _verizonTheme["default"];
  }
});
Object.defineProperty(exports, "colors", {
  enumerable: true,
  get: function get() {
    return _verizonTheme["default"];
  }
});
Object.defineProperty(exports, "breakpoints", {
  enumerable: true,
  get: function get() {
    return _layout.breakpoints;
  }
});
Object.defineProperty(exports, "spacers", {
  enumerable: true,
  get: function get() {
    return _layout.spacers;
  }
});
Object.defineProperty(exports, "setSpacingFactor", {
  enumerable: true,
  get: function get() {
    return _layout.setSpacingFactor;
  }
});
Object.defineProperty(exports, "getSpacingFactor", {
  enumerable: true,
  get: function get() {
    return _layout.getSpacingFactor;
  }
});
Object.defineProperty(exports, "setSpacingScale", {
  enumerable: true,
  get: function get() {
    return _layout.setSpacingScale;
  }
});
Object.defineProperty(exports, "getSpacingScale", {
  enumerable: true,
  get: function get() {
    return _layout.getSpacingScale;
  }
});
Object.defineProperty(exports, "calculateSpacing", {
  enumerable: true,
  get: function get() {
    return _layout.calculateSpacing;
  }
});
Object.defineProperty(exports, "vdsSpacingScale", {
  enumerable: true,
  get: function get() {
    return _layout.vdsSpacingScale;
  }
});

var _verizonTheme = _interopRequireDefault(require("./verizonTheme"));

var _layout = require("./layout");
