// @vds-core/selects/dist/cjs/components/DropdownSelectMarketingTrigger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _setEnvRef = _interopRequireDefault(require("@uie/set-env-ref"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _typography = require("@vds-core/typography");

var _theme = require("@vds-core/theme");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  position: relative;\n  flex-direction: row;\n  align-items: center;\n  justify-content: space-between;\n  background: transparent;\n  border-bottom: ", ";\n  border-top: 0;\n  border-right: 0;\n  border-left: 0;\n  padding: 0;\n  margin: 0;\n  padding-bottom: ", ";\n  color: ", ";\n  cursor: ", ";\n  position: relative;\n  transition: border 300ms;\n  transition: color 300ms;\n  width: ", ";\n  overflow: ", ";\n  text-overflow: ellipsis;\n  z-index: 1;\n  white-space: nowrap;\n  &.focus-visible {\n    background: white;\n    color: ", ";\n    transition: color 300ms;\n  }\n\n  ", "\n\n  &:hover,\n  &:active {\n    outline: none;\n    outline-offset: 0;\n  }\n\n  font-weight: ", ";\n  font-family: ", ";\n  font-size: ", ";\n  line-height: ", ";\n\n  ", "\n\n  * {\n    outline: none;\n    outline-offset: 0;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

/*
fontWeightBold represents the font weight found in 
typography used when the component is bolded. Font weight is
not exported by typography configs due to the fact that 
typography components may have multiple font weight options
*/
var fontWeightBold = 'bold';
var propTypes = {
  /**
   * @ignore
   * ref for component
   */
  setTriggerRef: _propTypes["default"].func,

  /**
   * @ignore
   * config for colors
   */
  color: _propTypes["default"].object,

  /**
   * @ignore
   * boolean that, if true, changes the hover states
   */
  size: _propTypes["default"].oneOf(['large', 'normal', 'medium', 'small']),

  /**
   * @ignore
   * boolean that, if true, changes the hover states
   */
  viewport: _propTypes["default"].oneOf(['mobile', 'tablet', 'desktop']),

  /**
   * @ignore
   * prop that changes the padding between the text and the bottom line
   */
  triggerPaddingBottom: _propTypes["default"].string,

  /**
   * @ignore
   * string that changes which typescale to use for typography
   */
  typescale: _propTypes["default"].string,

  /**
   * @ignore
   * config for body typography
   */
  bodyConfig: _propTypes["default"].object,

  /**
   * @ignore
   * config for title typography
   */
  titleConfig: _propTypes["default"].object,

  /**
   * @ignore
   * boolean that, if true, adds an additional hit area
   */
  hitArea: _propTypes["default"].bool,

  /**
   * @ignore
   * boolean that, if true, changes the focus states
   */
  focusState: _propTypes["default"].bool
};
var defaultProps = {
  setTriggerRef: function setTriggerRef() {},
  size: 'large',
  viewport: 'desktop',
  color: _theme.colors,
  hitArea: false,
  focusState: false,
  triggerPaddingBottom: (0, _javascriptStyleHelpers.calculateRem)(2),
  typescale: _typography.TypographyConfig.getTypescale(),
  bodyConfig: _typography.BodyConfig,
  titleConfig: _typography.TitleConfig
};

var Trigger = /*#__PURE__*/_styledComponents["default"].button.withConfig({
  displayName: "Trigger",
  componentId: "sc-3o9m2k-0"
})(_templateObject(), function (_ref) {
  var error = _ref.error,
      disabled = _ref.disabled,
      color = _ref.color;
  var borderColor = error ? color.error : disabled ? color.coolGray3 : color.black;
  var borderHeight = (0, _javascriptStyleHelpers.calculateRem)(error ? 4 : 1);
  return "".concat(borderHeight, " solid ").concat(borderColor);
}, function (_ref2) {
  var triggerPaddingBottom = _ref2.triggerPaddingBottom;
  return triggerPaddingBottom;
}, function (props) {
  return props.disabled ? props.color.coolGray3 : props.selected ? props.color.black : props.color.coolGray6;
}, function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? 'default' : 'pointer';
}, function (_ref4) {
  var width = _ref4.width;
  return width;
}, function (_ref5) {
  var hitArea = _ref5.hitArea;
  return hitArea ? 'visible' : 'hidden';
}, function (_ref6) {
  var color = _ref6.color;
  return color.black;
}, function (_ref7) {
  var focusState = _ref7.focusState,
      color = _ref7.color;
  return focusState && "\n  &:focus {\n    outline: ".concat((0, _javascriptStyleHelpers.calculateRem)(1), " dashed ").concat(color.black, ";\n    outline-offset: ").concat((0, _javascriptStyleHelpers.calculateRem)(2), ";\n    color: ").concat(color.black, ";\n  }\n\n  &:focus-visible {\n    outline: ").concat((0, _javascriptStyleHelpers.calculateRem)(1), " dashed ").concat(color.black, ";\n    outline-offset: ").concat((0, _javascriptStyleHelpers.calculateRem)(2), ";\n    color: ").concat(color.black, ";\n  }\n\n  :-moz-focusring {\n    color: transparent;\n    text-shadow: 0 0 0 #000;\n  }\n\n");
}, fontWeightBold, function (_ref8) {
  var body = _ref8.body;
  return body.fontFamily;
}, function (_ref9) {
  var body = _ref9.body;
  return (0, _javascriptStyleHelpers.calculateRem)(body.fontSize);
}, function (_ref10) {
  var body = _ref10.body;
  return (0, _javascriptStyleHelpers.calculateRem)(body.lineHeight);
}, function (_ref11) {
  var disabled = _ref11.disabled,
      color = _ref11.color;
  return !disabled && "\n      &:hover {\n        transition: color 300ms;\n        color: ".concat(color.black, ";\n      }\n  ");
});
/**
 * @ignore
 */


var DropdownTrigger = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DropdownTrigger, _Component);

  var _super = _createSuper(DropdownTrigger);

  function DropdownTrigger() {
    var _this;

    (0, _classCallCheck2["default"])(this, DropdownTrigger);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "componentDidMount", function () {
      _this.props.setTriggerRef(_this.triggerElem);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "refKey", (0, _setEnvRef["default"])());
    return _this;
  }

  (0, _createClass2["default"])(DropdownTrigger, [{
    key: "render",
    value: function render() {
      var _this2 = this,
          _extends2;

      var _this$props = this.props,
          hitArea = _this$props.hitArea,
          color = _this$props.color,
          focusState = _this$props.focusState,
          size = _this$props.size,
          typescale = _this$props.typescale,
          viewport = _this$props.viewport,
          titleConfig = _this$props.titleConfig,
          bodyConfig = _this$props.bodyConfig;
      var fontScale;

      if (size === 'small' && typescale !== 'VDS') {
        fontScale = 'desktop';
      } else {
        fontScale = viewport === 'mobile' ? 'mobile' : 'desktop';
      }

      var fontSize;

      if (typescale === 'VDS') {
        fontSize = size === 'small' ? 'small' : 'large';
      } else {
        if (size === 'small') {
          fontSize = viewport === 'mobile' ? 'small' : 'medium';
        } else {
          fontSize = 'large';
        }
      }

      var typeConfig;

      if (size !== 'small' || typescale === 'VDS') {
        typeConfig = titleConfig;
      } else {
        typeConfig = bodyConfig;
      }

      var body = typeConfig.getStyles(typescale, fontScale, fontSize);
      return /*#__PURE__*/_react["default"].createElement(Trigger, (0, _extends3["default"])((_extends2 = {
        body: body,
        bodyConfig: bodyConfig,
        titleConfig: titleConfig,
        typescale: typescale,
        viewport: viewport
      }, (0, _defineProperty2["default"])(_extends2, "body", body), (0, _defineProperty2["default"])(_extends2, "size", size), (0, _defineProperty2["default"])(_extends2, "focusState", focusState), (0, _defineProperty2["default"])(_extends2, "hitArea", hitArea), (0, _defineProperty2["default"])(_extends2, "role", "listbox"), (0, _defineProperty2["default"])(_extends2, "aria-owns", "dropdown-list"), _extends2), this.props, (0, _defineProperty2["default"])({}, this.refKey, function (elem) {
        return _this2.triggerElem = elem;
      })));
    }
  }]);
  return DropdownTrigger;
}(_react.Component);

DropdownTrigger.defaultProps = defaultProps;
DropdownTrigger.propTypes = propTypes;
var _default = DropdownTrigger;
exports["default"] = _default;
