// @vds-core/selects/dist/cjs/components/DropdownSelectMarketingList.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _typography = require("@vds-core/typography");

var _theme = require("@vds-core/theme");

var _setEnvRef = _interopRequireDefault(require("@uie/set-env-ref"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  z-index: 2;\n  position: absolute;\n  left: 0;\n  top: 100%;\n  margin: 0;\n  padding: 0;\n  width: ", ";\n\n  .slide-down[data-menu-list] {\n  }\n\n  > [data-menu] {\n    display: block;\n  }\n\n  &[data-menu-list] {\n    animation: ", " 0.2s ease;\n    background: ", ";\n    padding-bottom: 0;\n    padding-top: 0;\n\n    transition: border 400ms, height 350ms;\n  }\n\n  > [data-menu-item],\n  li {\n    display: flex;\n    align-items: center;\n    background: none;\n    border-left: 1px solid ", ";\n    border-right: 1px solid ", ";\n    border-bottom: 1px solid ", ";\n    box-sizing: border-box;\n    min-height: ", ";\n  }\n  > [data-menu-item],\n  li {\n    overflow-wrap: anywhere;\n    width: 100%;\n    display: flex;\n    align-items: center;\n    color: ", ";\n    cursor: pointer;\n    text-align: left;\n    text-decoration: none;\n    margin-top: ", ";\n\n    padding: ", ";\n\n    font-weight: ", ";\n    font-family: ", ";\n    font-size: ", ";\n    line-height: ", ";\n\n    &:hover {\n      color: ", ";\n    }\n\n    &:focus {\n      outline: none;\n      color: ", ";\n      transition: color 300ms;\n    }\n\n    ", "\n\n    ", "    \n\n    &:first-of-type {\n      margin-top: 0;\n    }\n  }\n\n  > [data-menu-item][data-selected] {\n    background: transparent;\n    color: ", ";\n    transition: color 300ms;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  0% {\n    opacity: 0;\n    transform: translateY(-10px);\n  }\n  100% {\n    opacity: 1;\n    transform: translateY(0);\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

/*
fontWeightBold represents the font weight found in 
typography used when the component is bolded. Font weight is
not exported by typography configs due to the fact that 
typography components may have multiple font weight options
*/
var fontWeightBold = 'bold';

function _calculateListItemHeight(size) {
  if (size === 'small') {
    return (0, _javascriptStyleHelpers.calculateRem)(42);
  } else return (0, _javascriptStyleHelpers.calculateRem)(56);
}

function _calculateListItemPadding(size, viewport) {
  if (size === 'small' && viewport === 'mobile') {
    return (0, _javascriptStyleHelpers.calculateRem)(14, 14, 13.6, 14);
  } else if (size === 'small' && viewport !== 'mobile') {
    return (0, _javascriptStyleHelpers.calculateRem)(13, 14, 12.2, 14);
  } else if (size !== 'small' && viewport === 'mobile') {
    return (0, _javascriptStyleHelpers.calculateRem)(16, 12, 16, 12);
  } else {
    return (0, _javascriptStyleHelpers.calculateRem)(13, 12, 12.6, 12);
  }
}

var propTypes = {
  /**
   * @ignore
   */
  color: _propTypes["default"].object,

  /**
   * @ignore
   * config for passing though title typography
   */
  titleConfig: _propTypes["default"].object,

  /**
   * @ignore
   * config for passing though bible typography
   */
  bodyConfig: _propTypes["default"].object,

  /**
   * @ignore
   * passes through typescale via sting if there's more than one option [example 'VDS' instead of 'MVP']
   */
  typescale: _propTypes["default"].string,

  /**
   * @ignore
   * function that calculates the height of the drop down list item
   */
  calculateListItemHeight: _propTypes["default"].func,

  /**
   * @ignore
   * function that calculates the padding inside of each list item
   */
  calculateListItemPadding: _propTypes["default"].func,

  /**
   * @ignore
   * boolean that, if true, changes the hover states
   */
  hoverState: _propTypes["default"].bool,

  /**
   * @ignore
   * boolean that, if true, changes the focus states
   */
  focusState: _propTypes["default"].bool,

  /**
   * @ignore
   * togggles list when executed
   */
  toggleList: _propTypes["default"].func
};
var defaultProps = {
  color: _theme.colors,
  titleConfig: _typography.TitleConfig,
  bodyConfig: _typography.BodyConfig,
  typescale: _typography.TypographyConfig.getTypescale(),
  calculateListItemHeight: _calculateListItemHeight,
  calculateListItemPadding: _calculateListItemPadding,
  hoverState: false,
  focusState: false
};
var slideDown = /*#__PURE__*/(0, _styledComponents.keyframes)(_templateObject());

var List = /*#__PURE__*/_styledComponents["default"].ul.withConfig({
  displayName: "List",
  componentId: "sc-1fp3lpg-0"
})(_templateObject2(), function (_ref) {
  var width = _ref.width;
  return width;
}, slideDown, function (_ref2) {
  var color = _ref2.color;
  return color.white;
}, function (_ref3) {
  var color = _ref3.color;
  return color.coolGray3;
}, function (_ref4) {
  var color = _ref4.color;
  return color.coolGray3;
}, function (_ref5) {
  var color = _ref5.color;
  return color.coolGray3;
}, function (_ref6) {
  var calculateListItemHeight = _ref6.calculateListItemHeight,
      size = _ref6.size;
  return calculateListItemHeight(size);
}, function (_ref7) {
  var color = _ref7.color;
  return color.coolGray6;
}, (0, _javascriptStyleHelpers.calculateRem)(-1), function (_ref8) {
  var calculateListItemPadding = _ref8.calculateListItemPadding,
      size = _ref8.size,
      viewport = _ref8.viewport;
  return calculateListItemPadding(size, viewport);
}, fontWeightBold, function (_ref9) {
  var body = _ref9.body;
  return body.fontFamily;
}, function (_ref10) {
  var body = _ref10.body;
  return (0, _javascriptStyleHelpers.calculateRem)(body.fontSize);
}, function (_ref11) {
  var body = _ref11.body;
  return (0, _javascriptStyleHelpers.calculateRem)(body.lineHeight);
}, function (_ref12) {
  var color = _ref12.color;
  return color.black;
}, function (_ref13) {
  var focusState = _ref13.focusState,
      color = _ref13.color;
  return focusState ? null : color.black;
}, function (_ref14) {
  var hoverState = _ref14.hoverState,
      color = _ref14.color;
  return hoverState && "\n      &:hover {\n        color: ".concat(color.black, ";\n        transition: color 300ms;\n      }\n    ");
}, function (_ref15) {
  var focusState = _ref15.focusState,
      color = _ref15.color;
  return focusState && "\n      &:focus {\n        outline: ".concat((0, _javascriptStyleHelpers.calculateRem)(1), " dashed ").concat(color.black, ";\n        outline-offset: -").concat((0, _javascriptStyleHelpers.calculateRem)(8), ";\n      }\n\n      &:hover,\n      &:active {\n        outline: none;\n        outline-offset: 0;\n      }\n\n    ");
}, function (_ref16) {
  var color = _ref16.color;
  return color.black;
});
/**
 * @ignore
 */


var DropdownList = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DropdownList, _Component);

  var _super = _createSuper(DropdownList);

  function DropdownList(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, DropdownList);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "componentDidMount", function () {
      var topOfList = _this.props.itemId + 'option=0';
      _this.props.keyboardFocused && document.getElementById(topOfList).focus();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_onKeyDown", function (e) {
      var key = e.keyCode;
      var shift = e.shiftKey;
      var tab = key === 9;
      var currentId = e.target.id;
      var index = parseInt(currentId.split('=')[1]); //strips index from option id

      var itemPreface = _this.props.itemId + 'option=';
      var firstOption = itemPreface + '0';
      var lastOption = itemPreface + (_this.props.children.length - 1);
      var nextOption = itemPreface + (index + 1);
      var previousOption = itemPreface + (index - 1);
      e.preventDefault();
      if (shift && tab) key = 37; //if shift + tab, go backward

      switch (key) {
        case 37: //shift + tab, up or right moves position backwards in list

        case 38:
          if (currentId === firstOption) {
            document.getElementById(lastOption).focus();
          } else {
            document.getElementById(previousOption).focus();
          }

          break;

        case 9:
        case 39:
        case 40:
          //tab, down or left advances position in list
          if (currentId === lastOption) {
            document.getElementById(firstOption).focus();
          } else {
            document.getElementById(nextOption).focus();
          }

          break;

        case 27:
          //esc closes list
          _this.props.toggleList();

          _this.props.trigger.focus();

          break;
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_renderChildren", function (children, itemId) {
      return _react["default"].Children.map(children, function (child, i) {
        var id = _this.props.itemId + 'option=' + i;

        var el = /*#__PURE__*/_react["default"].cloneElement(child, {
          id: id,
          selectionCallback: _this.props.onSelect
        });

        return el;
      });
    });
    return _this;
  }

  (0, _createClass2["default"])(DropdownList, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          color = _this$props.color,
          size = _this$props.size,
          viewport = _this$props.viewport,
          calculateListItemHeight = _this$props.calculateListItemHeight,
          calculateListItemPadding = _this$props.calculateListItemPadding,
          titleConfig = _this$props.titleConfig,
          bodyConfig = _this$props.bodyConfig,
          typescale = _this$props.typescale,
          hoverState = _this$props.hoverState,
          focusState = _this$props.focusState,
          children = _this$props.children,
          width = _this$props.width;
      var fontScale;

      if (size === 'small' && typescale !== 'VDS') {
        fontScale = 'desktop';
      } else {
        fontScale = viewport === 'mobile' ? 'mobile' : 'desktop';
      }

      var fontSize;

      if (typescale === 'VDS') {
        fontSize = size === 'small' ? 'small' : 'large';
      } else {
        if (size === 'small') {
          fontSize = viewport === 'mobile' ? 'small' : 'medium';
        } else {
          fontSize = 'large';
        }
      }

      var typeConfig;

      if (size !== 'small' || typescale === 'VDS') {
        typeConfig = titleConfig;
      } else {
        typeConfig = bodyConfig;
      }

      var body = typeConfig.getStyles(typescale, fontScale, fontSize);
      return /*#__PURE__*/_react["default"].createElement(List, (0, _extends2["default"])({
        width: width,
        onKeyDown: this._onKeyDown,
        role: "listbox",
        id: "dropdown-list",
        body: body,
        size: size,
        focusState: focusState,
        hoverState: hoverState,
        viewport: viewport,
        titleConfig: titleConfig,
        bodyConfig: bodyConfig,
        typescale: typescale,
        calculateListItemHeight: calculateListItemHeight,
        calculateListItemPadding: calculateListItemPadding,
        color: color,
        "data-menu-list": true
      }, this.props), this._renderChildren(children));
    }
  }]);
  return DropdownList;
}(_react.Component);

DropdownList.defaultProps = defaultProps;
DropdownList.propTypes = propTypes;
var _default = DropdownList;
exports["default"] = _default;
