// @vds-core/selects/dist/cjs/components/DropdownSelectMarketing.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _setEnvRef = _interopRequireDefault(require("@uie/set-env-ref"));

var _javascriptStyleHelpers = require("@uie/javascript-style-helpers");

var _DropdownSelectMarketingTrigger = _interopRequireDefault(require("./DropdownSelectMarketingTrigger"));

var _DropdownSelectMarketingList = _interopRequireDefault(require("./DropdownSelectMarketingList"));

var _icons = _interopRequireWildcard(require("@vds-core/icons"));

var _theme = require("@vds-core/theme");

var _typography = require("@vds-core/typography");

var _utilities = require("@vds-core/utilities");

var _cuid = _interopRequireDefault(require("cuid"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  overflow: hidden;\n  text-overflow: ellipsis;\n  text-align: left;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  justify-content: center;\n  outline: none;\n  padding-left: ", ";\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: absolute;\n  top: ", ";\n  cursor: ", ";\n  display: inline-block;\n  height: ", ";\n  width: 100%;\n  z-index: 1;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: relative;\n  outline: none;\n  display: flex;\n  flex-direction: column;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  margin-top: ", ";\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _calculateIconSpacing(size) {
  if (size === 'small') {
    return (0, _theme.calculateSpacing)('2X');
  } else {
    return (0, _theme.calculateSpacing)('1X');
  }
}

function _calculateErrorLabelSpacing(size) {
  if (size === 'large') {
    return (0, _javascriptStyleHelpers.calculateRem)(3);
  } else if (size === 'small') {
    return (0, _javascriptStyleHelpers.calculateRem)(1);
  } else {
    return (0, _theme.calculateSpacing)('1X');
  }
}

var propTypes = {
  /**
   * Items for the Dropdown List. Should use DropdownSelectOption.
   */
  children: _propTypes["default"].node.isRequired,

  /**
   * @ignore
   */
  className: _propTypes["default"].string,

  /**
   * disables the Dropdown
   */
  disabled: _propTypes["default"].bool,

  /**
   * Notify the user if there is an error
   */
  error: _propTypes["default"].bool,

  /**
   * @deprecated
   * Message displayed when there is an error
   */
  errorMsg: _propTypes["default"].string,
  // DEPRECATED

  /**
   * Message displayed when there is an error
   */
  errorText: _propTypes["default"].string,

  /**
   * @deprecated
   * Text shown in the dropdown trigger.
   */
  label: _propTypes["default"].string,
  // DEPRECATED

  /**
   * Label for reading out meaningful text in aria-labels
   */
  hiddenLabel: _propTypes["default"].string,

  /**
   * Text shown in the dropdown trigger.
   */
  placeholder: _propTypes["default"].string,

  /**
   * The width of the dropdown. If no width is passed small dropdowns with be 156px and large dropdowns with be 145px wide.
   */
  width: _propTypes["default"].string,

  /**
   * The size of the entire dropdown.
   */
  size: _propTypes["default"].oneOf(['large', 'normal', 'medium', 'small']),

  /**
   * The viewport that the dropdown styling should conform to.
   */
  viewport: _propTypes["default"].oneOf(['mobile', 'tablet', 'desktop']),

  /**
   * @ignore
   */
  color: _propTypes["default"].object,

  /**
   * Selected item of the Dropdown List.
   */
  selected: _propTypes["default"].node,

  /**
   * @ignore
   * string that sets the typescale for the typography (example: typescale === 'VDS')
   */
  typescale: _propTypes["default"].string,

  /**
   * @ignore
   * config for body typography
   */
  bodyConfig: _propTypes["default"].object,

  /**
   * @ignore
   * config for title typography
   */
  titleConfig: _propTypes["default"].object,

  /**
   * @ignore
   * if true, the caret-down icon changes to caret-up if this.state.open is also true
   */
  iconActiveState: _propTypes["default"].bool,

  /**
   * @ignore
   * calculates the margin on top of the error text
   */
  calculateErrorLabelSpacing: _propTypes["default"].func,

  /**
   * @ignore
   * calculates the padding next to the icon
   */
  calculateIconSpacing: _propTypes["default"].func,

  /**
   * @ignore
   * adds additional hit area to the trigger
   */
  hitArea: _propTypes["default"].bool,

  /**
   * @ignore
   * used to set padding underneath the trigger
   */
  triggerPaddingBottom: _propTypes["default"].string,

  /**
   * @ignore
   * function that calculates the height of the drop down list item
   */
  calculateListItemHeight: _propTypes["default"].func,

  /**
   * @ignore
   * function that calculates the padding inside of each list item
   */
  calculateListItemPadding: _propTypes["default"].func,

  /**
   * @ignore
   * boolean that, if true, changes the hover states
   */
  hoverState: _propTypes["default"].bool,

  /**
   * @ignore
   * boolean that, if true, changes the focus states
   */
  focusState: _propTypes["default"].bool,

  /**
   * @ignore
   * passes through data for icons
   */
  iconConfig: _propTypes["default"].object,

  /**
   * Allows a string to be provided for analytics.
   */
  'data-analyticstrack': _propTypes["default"].string,

  /**
   * Allows a string to be provided for analytics.
   */
  'data-track': _propTypes["default"].string,

  /**
   * Allows a string to be provided for click stream.
   */
  'data-clickstream': _propTypes["default"].string,

  /**
   * Allows a unique ID to be passed to the component.
   */
  id: _propTypes["default"].string
};
var defaultProps = {
  className: null,
  disabled: false,
  error: false,
  size: 'normal',
  viewport: 'desktop',
  color: _theme.colors,
  selected: undefined,
  iconActiveState: false,
  calculateErrorLabelSpacing: _calculateErrorLabelSpacing,
  calculateIconSpacing: _calculateIconSpacing,
  hitArea: false,
  hoverState: false,
  focusState: false,
  typescale: _typography.TypographyConfig.getTypescale(),
  iconConfig: _icons.IconData,
  hiddenLabel: null
};
var selectWidth = 156;
var selectWidthSmallMobile = 166;

var DropdownErrorWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "DropdownErrorWrapper",
  componentId: "sc-1aqp0gl-0"
})(_templateObject(), function (_ref) {
  var calculateErrorLabelSpacing = _ref.calculateErrorLabelSpacing,
      size = _ref.size;
  return calculateErrorLabelSpacing(size);
});

var DropdownWrapper = /*#__PURE__*/_styledComponents["default"].span.withConfig({
  displayName: "DropdownWrapper",
  componentId: "sc-1aqp0gl-1"
})(_templateObject2());

var HitArea = /*#__PURE__*/_styledComponents["default"].span.withConfig({
  displayName: "HitArea",
  componentId: "sc-1aqp0gl-2"
})(_templateObject3(), function (_ref2) {
  var viewport = _ref2.viewport,
      size = _ref2.size;
  return viewport === 'mobile' && size === 'small' ? (0, _javascriptStyleHelpers.calculateRem)(-10) : viewport === 'mobile' && size === 'large' ? (0, _javascriptStyleHelpers.calculateRem)(-5) : viewport === 'desktop' && size === 'small' ? (0, _javascriptStyleHelpers.calculateRem)(-7.5) : 0;
}, function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? 'default' : 'pointer';
}, (0, _javascriptStyleHelpers.calculateRem)(44));

var IconWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "IconWrapper",
  componentId: "sc-1aqp0gl-3"
})(_templateObject4(), function (_ref4) {
  var calculateIconSpacing = _ref4.calculateIconSpacing,
      size = _ref4.size;
  return calculateIconSpacing(size);
});

var SelectedTextWrapper = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  displayName: "SelectedTextWrapper",
  componentId: "sc-1aqp0gl-4"
})(_templateObject5());

var Dropdown = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Dropdown, _Component);

  var _super = _createSuper(Dropdown);

  function Dropdown(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Dropdown);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "componentDidUpdate", function (prevProps, prevState) {
      if ((!prevProps.selected || prevProps.selected !== _this.props.selected) && _this.props.selected) {
        _this.setState({
          selectedOption: _this.props.selected,
          open: false
        });
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "componentDidMount", function () {
      document.addEventListener('click', _this.handleClickOutside);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "componentWillUnmount", function () {
      document.removeEventListener('click', _this.handleClickOutside);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleClickOutside", function (e) {
      if (_this.triggerElem && !_this.triggerElem.contains(e.target)) {
        _this.setState({
          open: false
        });
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "toggle", function (e) {
      _this.setState({
        open: !_this.state.open
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_onKeyDown", function (e) {
      var key = e.keyCode;
      var entered = key === 13 || key === 32;
      var arrowed = key === 38 || key === 40;

      if (entered || arrowed) {
        _this.setState({
          keyboardFocused: true
        });
      }

      if (arrowed) _this.toggle();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setTriggerRef", function (element) {
      _this.triggerElem = element;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getIconSize", function (size, viewport, iconActiveState) {
      if (!iconActiveState) {
        switch (size) {
          case 'normal':
            return 15;

          case 'large':
            return 'medium';

          case 'small':
          case 'medium':
          default:
            return 'extraSmall';
        }
      } else {
        if ((viewport === 'desktop' || viewport === 'tablet') && size === 'large') {
          return 'large';
        } else if (viewport === 'mobile' && size === 'large') {
          return 'medium';
        } else if ((viewport === 'desktop' || viewport === 'tablet') && size === 'small') {
          return 'small';
        } else return 'extraSmall';
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_onSelect", function (e) {
      _this.setState({
        open: false
      });

      _this.props.onSelect && _this.props.onSelect(e);

      _this.triggerElem.focus();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getWidth", function (size, width, viewport) {
      if (width) return width;
      return size !== 'small' ? (0, _javascriptStyleHelpers.calculateRem)(selectWidth) : viewport === 'mobile' ? (0, _javascriptStyleHelpers.calculateRem)(selectWidthSmallMobile) : (0, _javascriptStyleHelpers.calculateRem)(selectWidth);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_getAriaLabel", function () {
      if (_this.props.hiddenLabel && _this.state.selectedOption) {
        return _this.state.selectedOption + 'selected from ' + _this.props.hiddenLabel;
      } else if (!_this.props.hiddenLabel && _this.state.selectedOption) {
        return _this.state.selectedOption + 'selected from dropdown select';
      } else if (_this.props.hiddenLabel && !_this.state.selectedOption) {
        return _this.props.hiddenLabel;
      } else return 'dropdown select';
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "refKey", (0, _setEnvRef["default"])());

    _this.focusTrigger = function () {
      if (_this.triggerElem) _this.triggerElem.focus();
    };

    _this.state = {
      open: false,
      selectedOption: _this.props.selected || undefined,
      keyboardFocused: false
    };
    return _this;
  }

  (0, _createClass2["default"])(Dropdown, [{
    key: "render",
    value: function render() {
      var _React$createElement2;

      var _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          disabled = _this$props.disabled,
          error = _this$props.error,
          errorMsg = _this$props.errorMsg,
          errorText = _this$props.errorText,
          label = _this$props.label,
          placeholder = _this$props.placeholder,
          size = _this$props.size,
          color = _this$props.color,
          width = _this$props.width,
          viewport = _this$props.viewport,
          selected = _this$props.selected,
          iconConfig = _this$props.iconConfig,
          hitArea = _this$props.hitArea,
          hoverState = _this$props.hoverState,
          focusState = _this$props.focusState,
          typescale = _this$props.typescale,
          triggerPaddingBottom = _this$props.triggerPaddingBottom,
          calculateErrorLabelSpacing = _this$props.calculateErrorLabelSpacing,
          calculateIconSpacing = _this$props.calculateIconSpacing,
          calculateListItemHeight = _this$props.calculateListItemHeight,
          calculateListItemPadding = _this$props.calculateListItemPadding,
          iconActiveState = _this$props.iconActiveState,
          bodyConfig = _this$props.bodyConfig,
          titleConfig = _this$props.titleConfig,
          id = _this$props.id,
          analyticsTrack = _this$props['data-analyticstrack'],
          track = _this$props['data-track'],
          clickStream = _this$props['data-clickstream'],
          hiddenLabel = _this$props.hiddenLabel;
      var errorMessage = errorMsg ? errorMsg : errorText;
      var selectedOption = this.state.selectedOption || (label ? label : selected ? selected : placeholder ? placeholder : '');

      var iconSize = this._getIconSize(size, viewport, iconActiveState);

      var selectWidth = this._getWidth(size, width, viewport);

      var errorState = error;
      var LabelType;

      if (typescale === 'VDS') {
        LabelType = _typography.Body;
      } else {
        LabelType = _typography.Micro;
      }

      var itemId = (0, _cuid["default"])();
      error && !errorText && console.error('Error Text for DropdownSelectMarketing MUST be provided');
      return /*#__PURE__*/_react["default"].createElement(DropdownWrapper, {
        tabIndex: -1,
        id: id,
        className: className
      }, /*#__PURE__*/_react["default"].createElement(_DropdownSelectMarketingTrigger["default"], {
        triggerPaddingBottom: triggerPaddingBottom,
        hitArea: hitArea,
        focusState: focusState,
        bodyConfig: bodyConfig,
        titleConfig: titleConfig,
        typescale: typescale,
        onClick: this.toggle,
        onKeyDown: this._onKeyDown,
        disabled: disabled,
        error: error,
        size: size,
        viewport: viewport,
        selected: selected,
        setTriggerRef: this.setTriggerRef,
        "data-clickstream": clickStream,
        "data-analyticstrack": analyticsTrack,
        "data-track": track,
        tabIndex: 0,
        width: selectWidth,
        "aria-expanded": this.state.open,
        "aria-label": this._getAriaLabel()
      }, hitArea && /*#__PURE__*/_react["default"].createElement(HitArea, {
        disabled: disabled,
        size: size,
        viewport: viewport
      }), /*#__PURE__*/_react["default"].createElement(SelectedTextWrapper, null, selectedOption), /*#__PURE__*/_react["default"].createElement(IconWrapper, {
        size: size,
        viewport: viewport,
        calculateIconSpacing: calculateIconSpacing
      }, /*#__PURE__*/_react["default"].createElement(_icons["default"], (0, _defineProperty2["default"])({
        size: size,
        viewport: viewport,
        iconActiveState: iconActiveState,
        tabIndex: -1,
        data: iconConfig,
        name: this.state.open ? 'caret-up' : 'caret-down',
        "aria-hidden": true,
        color: disabled ? color.coolGray3 : color.black
      }, "size", iconSize)))), this.state.open && /*#__PURE__*/_react["default"].createElement(_DropdownSelectMarketingList["default"], (_React$createElement2 = {
        width: selectWidth,
        keyboardFocused: this.state.keyboardFocused,
        onSelect: this._onSelect,
        size: size,
        viewport: viewport,
        titleConfig: titleConfig,
        bodyConfig: bodyConfig,
        typescale: typescale,
        hoverState: hoverState,
        focusState: focusState,
        onClick: this.focusTrigger,
        calculateListItemHeight: calculateListItemHeight,
        calculateListItemPadding: calculateListItemPadding,
        color: color,
        error: error
      }, (0, _defineProperty2["default"])(_React$createElement2, "viewport", viewport), (0, _defineProperty2["default"])(_React$createElement2, "toggleList", this.toggle), (0, _defineProperty2["default"])(_React$createElement2, "itemId", itemId), (0, _defineProperty2["default"])(_React$createElement2, "trigger", this.triggerElem), _React$createElement2), children), errorState && errorMessage && !disabled && !this.state.open && /*#__PURE__*/_react["default"].createElement(DropdownErrorWrapper, {
        color: color,
        size: size,
        calculateErrorLabelSpacing: calculateErrorLabelSpacing
      }, /*#__PURE__*/_react["default"].createElement(LabelType, {
        color: disabled ? color.coolGray3 : color.black,
        config: typescale === 'VDS' ? bodyConfig : _typography.MicroConfig,
        typescale: typescale
      }, errorMessage)));
    }
  }]);
  return Dropdown;
}(_react.Component);

Dropdown.propTypes = propTypes;
Dropdown.defaultProps = defaultProps;

var _default = (0, _utilities.withVDSManager)(Dropdown);

exports["default"] = _default;
