// @vds/typography/dist/cjs/components/Body/Body.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _typography = require("@vds-core/typography");

var _index = require("../../index");

var _config = _interopRequireDefault(require("./config"));

var _fonts = require("../../fonts");

var propTypes = {
  size: _propTypes["default"].oneOf(['small', 'large'])
};
var defaultProps = {
  size: 'small'
};

var Body = function Body(props) {
  function getFontWeight(typescale, bold) {
    return bold ? '700' : '400';
  }

  return /*#__PURE__*/_react["default"].createElement(_typography.Body, (0, _extends2["default"])({}, props, {
    getFontWeight: getFontWeight,
    typescale: _index.TypographyConfig.getTypescale(),
    config: _config["default"],
    size: props.size,
    fontsConfig: _fonts.Fonts
  }));
};

Body.propTypes = propTypes;
Body.defaultProps = defaultProps;
var _default = Body;
exports["default"] = _default;
