// @vds/typography/dist/cjs/components/Body/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _fonts = require("../../fonts");

function getStyles(typescale, viewport, typeSize) {
  return BodyConfig[typescale][viewport][typeSize];
}

var BodyConfig = {
  VDS: {
    mobile: {
      large: {
        fontSize: 16,
        lineHeight: 20,
        letterSpacing: 0.5,
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      medium: {
        fontSize: 12,
        lineHeight: 16,
        fontFamily: _fonts.Fonts.VerizonNHGeTX
      },
      small: {
        fontSize: 12,
        lineHeight: 16,
        fontFamily: _fonts.Fonts.VerizonNHGeTX
      }
    },
    desktop: {
      large: {
        fontSize: 16,
        lineHeight: 20,
        letterSpacing: 0.5,
        fontFamily: _fonts.Fonts.VerizonNHGeDS
      },
      medium: {
        fontSize: 12,
        lineHeight: 16,
        fontFamily: _fonts.Fonts.VerizonNHGeTX
      },
      small: {
        fontSize: 12,
        lineHeight: 16,
        fontFamily: _fonts.Fonts.VerizonNHGeTX
      }
    }
  },
  getStyles: getStyles
};
var _default = BodyConfig;
exports["default"] = _default;
